---
title: "Maela genomes & antigens"
author: "Nicholas Croucher"
date: "2023-02-18"
always_allow_html: true
output:
  html_document: default
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE, warning = FALSE,
                      message = FALSE, cache.lazy = FALSE)
library(magrittr)
library(tidyverse)
library(ggrepel)
library(tsne)
library(ggforce)
library(SIMLR)
library(rstatix)
library(ggpubr)
library(e1071)
library(kableExtra)
library(ggbeeswarm)
library(ggpubr)
library(corrr)
library(cowplot)
library(archive)

firstup <- function(x) {
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  x
}

plot_tsne <- function(mat,df,perplexity = 40,match_by=NULL,colour_by=NULL,group_by=NULL,shape_by=NULL,group_min=10,max_iter=1000) {
  
  tsne_out<-tsne::tsne(mat, perplexity = perplexity, max_iter = max_iter)
  
  mat_row_order<-match(rownames(mat),df[[match_by]])
  df<-df[mat_row_order,]
  df[["x"]]<-tsne_out[,1]
  df[["y"]]<-tsne_out[,2]
  
  df %<>%
    dplyr::group_by(.data[[colour_by]]) %>%
    dplyr::mutate(group_count = n()) %>% 
    dplyr::ungroup()
  
  base_plot <- NULL
  if (is_null(shape_by)) {
    base_plot <- ggplot(df,
                         aes(x = x,
                             y = y,
                             colour = .data[[colour_by]],
                             group = .data[[group_by]]))
  } else {
    base_plot <- ggplot(df,
                         aes(x = x,
                             y = y,
                             colour = .data[[colour_by]],
                             shape = .data[[shape_by]],
                             group = .data[[group_by]]))
    
  }
  
  base_plot <-
    base_plot +
      geom_point()
  
  if (group_min < 4) {
    base_plot <-
      base_plot +
      ggforce::geom_mark_ellipse(aes(x0 = x, y0 = y),
                                 expand = unit (1.5,"mm"))
  } else {
    base_plot <-
      base_plot +
      stat_ellipse(data = df %>% dplyr::filter(group_count >= group_min),type = "t")
  }
    
  base_plot <-
    base_plot +
    xlab("t-SNE dimension 1") +
    ylab("t-SNE dimension 2") +
    theme_bw() +
    theme(legend.position = "bottom")

}

pairwise_model_comparisons <- function(m_list) {
  
  num_models <- length(m_list)
  
  i_vec <- NULL
  j_vec <- NULL
  i_aic_vec <- NULL
  j_aic_vec <- NULL
  i_bic_vec <- NULL
  j_bic_vec <- NULL
  p_vec <- NULL
  for (i in 1:num_models) {
    for (j in 1:num_models) {
      if (i != j & m_list[[i]]@devcomp$dims[1] == m_list[[j]]@devcomp$dims[1]) {
        m_comp <- anova(m_list[[i]],m_list[[j]])
        i_vec <- c(i_vec,i)
        j_vec <- c(j_vec,j)
        i_aic_vec <- c(i_aic_vec,AIC(m_list[[i]]))
        j_aic_vec <- c(j_aic_vec,AIC(m_list[[j]]))
        i_bic_vec <- c(i_bic_vec,BIC(m_list[[i]]))
        j_bic_vec <- c(j_bic_vec,BIC(m_list[[j]]))
        p_vec <- c(p_vec,log(m_comp$`Pr(>Chisq)`[2]+1,10))
      }
    }
  }
  
  model_comp.df <- data.frame(
    "i" = as.factor(i_vec),
    "j" = as.factor(j_vec),
    "AIC_i" = i_aic_vec,
    "BIC_i" = i_bic_vec,
    "AIC_j" = j_aic_vec,
    "BIC_j" = j_bic_vec,
    "p" = p_vec
  ) %>%
    dplyr::mutate(
      delta_aic = AIC_i - AIC_j,
      delta_bic = BIC_i - BIC_j
    )
  
  return(model_comp.df)
}

# Cluster names
cluster_names <-
  data.frame(
    "cluster" = c("1","2","3","4","5","6"),
    # "cluster_name" = c("Neonatal highest", # 1
    #                    "Neonatal lowest", # 2
    #                    "Infant low", # 3
    #                    "Neonatal infant-like", # 4
    #                    "Infant high", # 5
    #                    "Neonatal intermediate") # 6
    # "cluster_name" = c("Neonatal infant-like", # 1
    #                    "Neonatal lowest", # 2
    #                    "Neonatal intermediate", # 3
    #                    "Infant low",  # 4
    #                    "Neonatal highest", # 5
    #                    "Infant high") # 6
    # "cluster_name" = c("Neonatal highest", # 1
    #                    "Neonatal intermediate",  # 4
    #                    "Infant low", # 5
    #                    "Neonatal lowest", # 2
    #                    "Infant high", # 3
    #                    "Neonatal infant-like" # 6
    #                    ) 
    "cluster_name" = c("Infant high", # 3
                       "Neonatal lowest", # 2
                       "Neonatal infant-like", # 6
                       "Neonatal highest", # 1
                       "Infant low", # 5
                       "Neonatal intermediate"  # 4
                       ) 
  )

# Palettes
cluster_palette <-
  c("Neonatal highest" = "#003C67FF", # 1
    "Neonatal lowest" = "#7AA6DCFF", # 2
    "Infant low" = "#95C11FFF", # 3
    "Neonatal infant-like" = "#0073C2FF" , # 4
    "Infant high" = "#007B3Dff", # 5
    "Neonatal intermediate" = "#4A6990FF"
  )

type_palette <- c(
  "non-ABT" = "#ADB6B6FF",
  "Other ABT" = "#1B1919FF",
  "Degradative" = "#925E9FFF",
  "SBP" = "#EE4C97FF",
  "CellWall" = "#42B540FF",
  "PBP" = "#20854EFF",
  "CellWall (not PBP)" = "#42B540FF",
  "Adhesin" = "#E18727FF",
  "PclA" = "#FDAF91FF",
  "Adhesin (not PclA)" = "#E18727FF",
  "PspA" = "#00468BFF",
  "PspC" = "#0099B4FF",
  "ZmpA" = "#ED0000FF",
  "ZmpB" = "#AD002AFF"
)

sero_age_shapes <- c(
  "Cord" = 15,
  "Birth" = 16,
  "6 months" = 7,
  "12 months" = 3,
  "18 months" = 4,
  "24 months" = 11
)

```

# Process raw array data

Load data from antigen array analysis:

```{r Processing of raw data to generate foreground values}

# Get probes from vaccine array analysis
raw.igg.data<-read.table(file="wSP-VAC002_PanGenome_DataMatrix.txt",header=T,row.names=1,stringsAsFactors = F)
igg.data<-t(raw.igg.data)
# keep some SP probes with different names
colnames(igg.data)[grep("^SP_0071",colnames(igg.data))]<-sub("SP_0071","CLS01991",colnames(igg.data)[grep("^SP_0071",colnames(igg.data))])
colnames(igg.data)[grep("^SP_1175",colnames(igg.data))]<-sub("SP_1175","CLS02425",colnames(igg.data)[grep("^SP_1175",colnames(igg.data))])
colnames(igg.data)[grep("^SP_1693",colnames(igg.data))]<-sub("SP_1693","CLS01450",colnames(igg.data)[grep("^SP_1693",colnames(igg.data))])
#"SP_0071_s3_316","SP_0071_s2_317","SP_0071_642","SP_0071_s1_3728" ->zmpC, degradative, CLS01991
#"SP_1175_436" -> CLS02425, adhesin
igg.data<-igg.data[,grep("^SP_",colnames(igg.data),invert=T)]
igg.data<-igg.data[,!(grepl("CLS02728",colnames(igg.data)))]
pv.probes<-colnames(igg.data)

# Get equivalent probes for Maela analysis
ann.df <- 
  read.csv(archive::archive_read("probe_annotation.csv.tar.gz"),
           stringsAsFactors = FALSE,
           check.names = FALSE,
           row.names = 1) %>%
  tibble::rownames_to_column(var = "probe") %>%
  dplyr::mutate(probe = dplyr::case_when(
                    grepl("SP_0071",probe) ~ gsub("SP_0071","CLS01991",probe),
                    grepl("SP_1175",probe) ~ gsub("SP_1175","CLS02425",probe),
                    grepl("SP_1693",probe) ~ gsub("SP_1693","CLS01450",probe),
                    TRUE ~ probe
                  )
                ) %>%
  dplyr::mutate(Unique.ID = dplyr::case_when(
                    grepl("SP_0071",Unique.ID) ~ gsub("SP_0071","CLS01991",Unique.ID),
                    grepl("SP_1175",Unique.ID) ~ gsub("SP_1175","CLS02425",Unique.ID),
                    grepl("SP_1693",Unique.ID) ~ gsub("SP_1693","CLS01450",Unique.ID),
                    TRUE ~ Unique.ID
                  )
                ) %>%
  dplyr::mutate(Gene.ID = dplyr::case_when(
                    grepl("SP_0071",Gene.ID) ~ gsub("SP_0071","CLS01991",Gene.ID),
                    grepl("SP_1175",Gene.ID) ~ gsub("SP_1175","CLS02425",Gene.ID),
                    grepl("SP_1693",Gene.ID) ~ gsub("SP_1693","CLS01450",Gene.ID),
                    TRUE ~ Gene.ID
                  )
                ) %>%
  dplyr::mutate(Strain.ID = dplyr::if_else(
                    Gene.ID %in% c("CLS01991","CLS02425","CLS01450"),
                    "SPARC",
                    Strain.ID
                  )
                ) %>%
  dplyr::filter(is.na(Gene.Name) | Gene.Name != "CLS02728") #%>%
  #tibble::column_to_rownames(var = "probe")

ctrl.df <-
  ann.df %>%
    dplyr::filter(Spot.Type == "IVTT.CTRL") 

ann.df %<>%
  dplyr::filter(Strain.ID != "TIGR4" & Spot.Type == "IVTT.AG")

# Check probes set matches
if (sum(ann.df$probe %in% pv.probes) == length(pv.probes)) {
  message("PROBE SETS DO NOT MATCH")
}

# Read in raw data
pheno.df <- 
  read.csv(archive::archive_read("sample_characteristics.csv.tar.gz"), stringsAsFactors = FALSE, check.names = FALSE)
back.df <- 
  read.csv(archive::archive_read("probe_background.csv.tar.gz"), stringsAsFactors = FALSE, check.names = FALSE, row.names = 1)
fore.df <- 
  read.csv(archive::archive_read("probe_foreground.csv.tar.gz"), stringsAsFactors = FALSE, check.names = FALSE, row.names = 1)

# Set IVTT spots with foreground MFI < 100 to NA--it is likely a missing spot
back.df[fore.df < 100] <- NA
fore.df[fore.df < 100] <- NA

# Now produce measurements without background for probes
raw_probe.df <-
  tibble::as_tibble(fore.df - back.df,
                    rownames = "probe") %>%
  # Update annotation
  dplyr::mutate(probe = dplyr::case_when(
                    grepl("SP_0071",probe) ~ gsub("SP_0071","CLS01991",probe),
                    grepl("SP_1175",probe) ~ gsub("SP_1175","CLS02425",probe),
                    grepl("SP_1693",probe) ~ gsub("SP_1693","CLS01450",probe),
                    TRUE ~ probe
                  )
                ) %>%
  # Link to protein annotation
  tidyr::pivot_longer(cols = -probe,
                      names_to = "Sample.ID",
                      values_to = "mfi")

# IVTT background calculation
ivtt_background.df <-
  raw_probe.df %>%
    dplyr::right_join(ctrl.df, by=c("probe")) %>%
    # Group by protein and calculate mean measurement
    dplyr::group_by(Sample.ID) %>%
    dplyr::summarise(background = median(mfi)) %>%
    dplyr::ungroup() %>%
    # Calculate log2 transform
    dplyr::mutate(log2_background = log(background,2)) %>%
    # Link to phenotype information
    dplyr::left_join(pheno.df %>% dplyr::select(Sample.Type,Sample.ID), by=c("Sample.ID")) %>%
    dplyr::filter(!grepl("REF",Sample.Type)) %>%
    # Tidy up
    dplyr::select(Sample.ID,log2_background)

# Now summarise responses per protein
protein.df <-
  raw_probe.df %>%
    # Link to phenotype information
    dplyr::left_join(pheno.df, by=c("Sample.ID")) %>%
    dplyr::filter(!grepl("REF",Sample.Type)) %>%
    # Link to protein annotation
    dplyr::right_join(ann.df, by = c("probe")) %>%
    # Group by protein and calculate mean measurement
    dplyr::group_by(Gene.ID,Sample.ID) %>%
    dplyr::summarise(mfi = mean(mfi)) %>%
    dplyr::ungroup() %>%
    # Adjust negatives to 1--this will be the floor of the data
    dplyr::mutate(mfi = dplyr::if_else(mfi < 1,1,mfi)) %>%
    # Calculate log2 transform
    dplyr::mutate(log2_mfi = log(mfi,2)) %>%
    # Remove background level
    dplyr::left_join(ivtt_background.df,by=c("Sample.ID")) %>%
    dplyr::mutate(normalised_log2_mfi = log2_mfi - log2_background) %>%
    # Set floor of normalized data to -2, or 1/4th of the IVTT.CTRL medians
    dplyr::mutate(normalised_log2_mfi = dplyr::if_else(normalised_log2_mfi < -2,
                                                       -2,
                                                       normalised_log2_mfi)
                  )

# Get values to replace missing data
missing_values.df <-
  protein.df %>%
    dplyr::left_join(pheno.df, by=c("Sample.ID")) %>%
    dplyr::group_by(Time.Point,Gene.ID) %>%
    dplyr::summarise(missing_val = median(normalised_log2_mfi, na.rm = TRUE),
                     Sample.ID = Sample.ID) %>%
    dplyr::ungroup()

# Summarise the final IgG data
igg.data <-
  as.matrix(
    protein.df %>%
      # Filter to only immunoreactive proteins
      dplyr::group_by(Gene.ID) %>%
      dplyr::mutate(max_igg = max(normalised_log2_mfi, na.rm = TRUE)) %>%
      dplyr::ungroup() %>%
      dplyr::filter(max_igg >= 1) %>%
      # Replace missing values with median for protein for age
      dplyr::left_join(missing_values.df %>%
                         dplyr::select(Sample.ID,Gene.ID,missing_val),
                       by = c("Sample.ID","Gene.ID")) %>%
      dplyr::mutate(normalised_log2_mfi = dplyr::if_else(is.na(normalised_log2_mfi),
                                                         missing_val,
                                                         normalised_log2_mfi)
                    ) %>%
      dplyr::select(-missing_val) %>%
      # Filter to only immunoreactive proteins
      tidyr::pivot_wider(id_cols = Gene.ID,names_from = Sample.ID,values_from = normalised_log2_mfi) %>%
      # Rename antigens
      dplyr::mutate(Gene.ID = dplyr::case_when(
                    grepl("zmp",Gene.ID) ~ gsub("zmp","Zmp",Gene.ID),
                    Gene.ID == "psrP" ~ "PsrP",
                    Gene.ID == "lytA" ~ "LytA",
                    Gene.ID == "pblB" ~ "PblB",
                    TRUE ~ Gene.ID
                  )
                ) %>%
      # Set rownames for matrix
      tibble::column_to_rownames(var = "Gene.ID")
  )

# Produce data per probe for Zmp analysis
probe.df <-
  raw_probe.df %>%
    # Link to phenotype information
    dplyr::left_join(pheno.df, by=c("Sample.ID")) %>%
    dplyr::filter(!grepl("REF",Sample.Type)) %>%
    # Link to protein annotation
    dplyr::right_join(ann.df, by = c("probe")) %>%
    # Adjust negatives to 1--this will be the floor of the data
    dplyr::mutate(mfi = dplyr::if_else(mfi < 1,1,mfi)) %>%
    # Calculate log2 transform
    dplyr::mutate(log2_mfi = log(mfi,2)) %>%
    # Remove background level
    dplyr::left_join(ivtt_background.df,by=c("Sample.ID")) %>%
    dplyr::mutate(normalised_log2_mfi = log2_mfi - log2_background) %>%
    # Set floor of normalized data to -2, or 1/4th of the IVTT.CTRL medians
    dplyr::mutate(normalised_log2_mfi = dplyr::if_else(normalised_log2_mfi < -2,
                                                       -2,
                                                       normalised_log2_mfi)
                  )

# Get probe values to replace missing data
missing_probe_values.df <-
  probe.df %>%
    dplyr::group_by(Time.Point,probe) %>%
    dplyr::summarise(missing_val = median(normalised_log2_mfi, na.rm = TRUE),
                     Sample.ID = Sample.ID) %>%
    dplyr::ungroup()

# Summarise the final probe IgG data
probe.igg.data <-
  as.matrix(
    probe.df %>%
      # Replace missing values with median for protein for age
      dplyr::left_join(missing_probe_values.df %>%
                         dplyr::select(Sample.ID,probe,missing_val),
                       by = c("Sample.ID","probe")) %>%
      dplyr::mutate(normalised_log2_mfi = dplyr::if_else(is.na(normalised_log2_mfi),
                                                         missing_val,
                                                         normalised_log2_mfi)
                    ) %>%
      dplyr::select(-missing_val) %>%
      # Filter to only immunoreactive proteins
      tidyr::pivot_wider(id_cols = probe,names_from = Sample.ID,values_from = normalised_log2_mfi) %>%
      # Rename antigens
      dplyr::mutate(probe = dplyr::case_when(
                    grepl("zmp",probe) ~ gsub("zmp","Zmp",probe),
                    probe == "psrP" ~ "PsrP",
                    probe == "lytA" ~ "LytA",
                    probe == "pblB" ~ "PblB",
                    TRUE ~ probe
                  )
                ) %>%
      # Set rownames for matrix
      tibble::column_to_rownames(var = "probe")
  )

# Detailed sample information
translation.df <-
  as_tibble(
    read.csv(archive::archive_read("sample_timings.csv.tar.gz")) %>%
      dplyr::mutate(specdate = lubridate::dmy(specdate)) %>%
      dplyr::mutate(dob = lubridate::dmy(dob)) %>%
      dplyr::select(sero_specnum,codenum,category,sero_age,dob,specdate) %>%
      dplyr::mutate(category = factor(str_to_sentence(category),levels=c("Mother","Infant"))) %>%
      dplyr::mutate(sero_age = factor(dplyr::case_when(
                                          sero_age == "DELIVERY" ~ "Cord",
                                          sero_age == "0" ~ "Birth",
                                          sero_age == "6" ~ "6 months",
                                          sero_age == "12" ~ "12 months",
                                          sero_age == "18" ~ "18 months",
                                          sero_age == "24" ~ "24 months"
                                     ),
                                     levels = c("Cord","Birth","6 months","12 months","18 months","24 months")
                              )
                    ) %>%
      dplyr::arrange(codenum,category,specdate) %>%
      dplyr::group_by(codenum) %>%
      dplyr::mutate(previous_specdate = dplyr::lag(specdate, n = 1L)) %>%
      dplyr::mutate(first_specdate = min(specdate)) %>%
      dplyr::ungroup()
  )

```

# Prepare the genomic data

## Load epidemiological data

Load data from:

-   original serological study

-   MLST analyses, which are converted to GPSCs using the GPS data from Gladstone et al

-   single genome studies from PMEN14, Hilty et al and Chewapreecha et al (latter study filtered through NFDS paper)

-   deep sequencing data

```{r Load epidemiological data}

# Link samples to epi identifiers
np.sample.info <- read.csv(archive::archive_read("nasopharyngeal_sampling.csv.tar.gz"))

# Get updated LRTI information
updated.lrti.df <-
  read.csv(archive::archive_read("lrti_sampling.csv.tar.gz")) %>%
  dplyr::rename(Xray_diagnosis = cxr) %>%
  dplyr::select(specnum,diagnosis_clinical,Xray_diagnosis)

# Sample information
epi.sample.info <-
  np.sample.info %>%
      dplyr::select(specnum,codenum,specdate,dob,age.d,age.v,category,nps.type,pnc) %>%
      dplyr::distinct() %>%
      dplyr::mutate(specdate = lubridate::dmy(specdate)) %>%
      dplyr::mutate(dob = lubridate::dmy(dob)) %>%
      dplyr::left_join(updated.lrti.df,
                       by = "specnum") %>%
      dplyr::mutate(diagnosis_clinical = dplyr::case_when(
                                    is.na(diagnosis_clinical) & nps.type == "Pneumonia" ~ "Pneumonia",
                                    is.na(diagnosis_clinical) ~ "Healthy",
                                    TRUE ~  diagnosis_clinical
                                  )
                                ) %>%
      dplyr::mutate(Xray_diagnosis = dplyr::if_else(
                                     is.na(Xray_diagnosis),
                                     "Not done",
                                     Xray_diagnosis
                                  )
                                ) %>%
    dplyr::select(-nps.type)

# Get information by serotype
serotype.sample.info <-
  np.sample.info %>%
      dplyr::select(specnum,serotype1,serotype2,sanger.id1,sanger.id2) %>%
      tidyr::pivot_longer(cols = c(serotype1,serotype2), names_to = "serotype_index", values_to = "serotype") %>%
      tidyr::pivot_longer(cols = c(sanger.id1,sanger.id2), names_to = "sanger_index", values_to = "sangernum") %>%
      dplyr::filter(!is.na(serotype)) %>%
      dplyr::filter((serotype_index == "serotype1" & sanger_index == "sanger.id1") |
                      (serotype_index == "serotype2" & sanger_index == "sanger.id2")) %>%
      dplyr::select(specnum,serotype,sangernum)
      
# Combine epi and serotype information
combined.info <-
  epi.sample.info %>%
      dplyr::full_join(serotype.sample.info, by = c("specnum")) %>%
      dplyr::mutate(serotype = dplyr::if_else(serotype %in% c("15B","15C","15B/C","15B/15C"),
                                              "15B/C",
                                              serotype))

# Translate MLST to GPSC using Gladstone et al
gpsc_to_st.df <- 
  read.csv(archive::archive_read("gpsc_information.csv.tar.gz")) %>%
  dplyr::select(ST,GPSC) %>%
  dplyr::distinct() %>%
  dplyr::mutate(ST = as.integer(ST)) %>% dplyr::group_by(ST) %>%
  dplyr::arrange(GPSC) %>%
  dplyr::slice(1) %>%
  dplyr::ungroup()

# MLST information
mlst.sample.info <- 
  read.csv(archive::archive_read("mlst_information.csv.tar.gz")) %>%
  dplyr::filter(specnum %in% np.sample.info$specnum) %>%
  dplyr::select(specnum,serotype,sequence_type) %>%
      dplyr::mutate(serotype = dplyr::if_else(serotype %in% c("15B","15C","15B/C","15B/15C"),
                                              "15B/C",
                                              serotype)) %>%
  dplyr::left_join(gpsc_to_st.df, by = c("sequence_type" = "ST"))

# Combine serotype and MLST
combined.info %<>%
    dplyr::full_join(mlst.sample.info, by = c("specnum","serotype"))

# Add in single isolate sequences
isolate.sample.info <-
  dplyr::bind_rows(
    read.csv(archive::archive_read("nt_isolates.csv.tar.gz")) %>%
      dplyr::select(specnum,serotype,gpsc,lane,accession),
    read.csv(archive::archive_read("gpsc1_isolates.csv.tar.gz")) %>%
      dplyr::filter(Source == "Paul Turner") %>%
      dplyr::rename(lane = Fastq) %>%
      dplyr::rename(specnum = Strain) %>%
      dplyr::rename(serotype = Inferred.Serotype) %>%
      dplyr::rename(accession = ENA.Accession) %>%
      dplyr::mutate(gpsc = 1) %>%
      dplyr::select(specnum,serotype,gpsc,lane,accession),
    read.csv(archive::archive_read("nfds_isolates.csv.tar.gz")) %>%
      dplyr::select(specnum,serotype,gpsc,lane,accession)
  ) %>%
  dplyr::rename(GPSC = gpsc) %>%
  dplyr::rename(SingleIsolateLane = lane) %>%
  dplyr::rename(SingleIsolateAccession = accession) %>%
  dplyr::mutate(serotype = dplyr::if_else(serotype %in% c("15B","15C","15B/C","15B/15C"),
                                              "15B/C",
                                              serotype))

# Add single isolate information
combined.info %<>%
    dplyr::full_join(isolate.sample.info %>%
                       dplyr::filter(specnum %in% combined.info$specnum) %>%
                       dplyr::mutate(GPSC),
                     by = c("specnum","serotype","GPSC"))

# Now deep sequencing
deep_sequencing.sample.info <-
    xlsx::read.xlsx(file="deep_sequencing_samples.xlsx",1) %>%
    dplyr::filter(Sample.ID %in% np.sample.info$specnum) %>% # changed from epi_codes.df
    tidyr::separate_rows(c(serotype,`GPSC.s.`),sep = ",") %>% # superseded by separate_longer_delim()
    dplyr::rename(specnum = Sample.ID) %>%
    dplyr::rename(GPSC = `GPSC.s.`) %>%
    dplyr::select(specnum,serotype,GPSC,Lane.ID,ENA.accession) %>%
    dplyr::mutate(serotype = dplyr::case_when(
                      serotype == "15B/15C" ~ "15B/C",
                      serotype == "35A/35C/42" ~ "35C",
                      serotype == "10X" ~ "10F",
                      TRUE ~ serotype
                    )
                  ) %>%
      dplyr::mutate(serotype = dplyr::if_else(serotype %in% c("15B","15C","15B/C","15B/15C"),
                                              "15B/C",
                                              serotype)) %>%
    dplyr::rename(DeepSeqLane = Lane.ID) %>%
    dplyr::rename(DeepSeqAccession = ENA.accession) %>%
    dplyr::distinct() %>%
    dplyr::filter(GPSC != "unknown" | serotype != "unknown")


# or try
combined.info %<>%
    dplyr::full_join(deep_sequencing.sample.info %>%
                       dplyr::mutate(GPSC = as.integer(GPSC)),
                     by = c("specnum","serotype","GPSC")) %>%
    dplyr::mutate(serotype = dplyr::if_else(serotype=="unknown",
                                            NA_character_,
                                            serotype)) %>%
    # dplyr::group_by(specnum,serotype) %>%
    # tidyr::fill(c(GPSC,sangernum,SingleIsolateLane,SingleIsolateAccession,DeepSeqLane,DeepSeqAccession),
    #             .direction="updown") %>%
    # dplyr::ungroup() %>%
    dplyr::group_by(specnum,serotype,GPSC) %>%
    tidyr::fill(c(sangernum,SingleIsolateLane,SingleIsolateAccession,DeepSeqLane,DeepSeqAccession),
                .direction="updown") %>%
    dplyr::ungroup() %>%
    dplyr::group_by(specnum) %>%
    tidyr::fill(c(codenum,specdate,dob,age.d,age.v,category,diagnosis_clinical,Xray_diagnosis,pnc),
                .direction="updown") %>%
    dplyr::ungroup() %>%
    dplyr::distinct()

```

There is some imputation of missing data:

-   if GPSC or serotype is unknown, then it is imputed from GPSC & serotype combinations observed within the same mother-child pair

-   if it is still not known, then if there is a one-to-one relationship between serotype and GPSC in the single isolate Maela collection, then this is used to infer missing information

Where there is no direct genomic sampling of a given specimen, representative genomes were selected by:

-   taking a genome of the same GPSC & serotype from the same mother/child pair

-   taking a genome of the same GPSC and serotype from any mother/child pair in the immunology cohort

-   taking a genome of the same GPSC and serotype from the wider cohort

```{r Identify representatives where no sequence available}

##############
# IMPUTATION #
##############

# Infer GPSCs/serotypes from mother-child pairs
combined.info %<>%
  dplyr::mutate(InferredGPSC = dplyr::if_else(pnc == "Pneumococcus NOT isolated",
                                              "None",
                                              as.character(GPSC))
                ) %>%
  dplyr::mutate(InferredSerotype = dplyr::if_else(pnc == "Pneumococcus NOT isolated",
                                              "None",
                                              serotype)
                ) %>%
  dplyr::mutate(RepresentativeGenome = dplyr::case_when(
                    pnc == "Pneumococcus NOT isolated" ~ "None",
                    !is.na(SingleIsolateLane) ~ SingleIsolateLane,
                    !is.na(DeepSeqLane) ~ DeepSeqLane,
                    TRUE ~ NA_character_
                  )
                ) %>%
  # Set levels for selecting representatives
  dplyr::mutate(RepresentativeEvidence = factor(dplyr::case_when(
                      pnc == "Pneumococcus NOT isolated" ~ "No carriage",
                      !is.na(SingleIsolateLane) ~ "Single isolate data",
                      !is.na(DeepSeqLane) ~ "Deep sequencing data",
                      TRUE ~ NA_character_
                    ),
                    levels = c("No carriage",
                               "No representative",
                               "Deep sequencing data",
                               "Single isolate data"
                               )
                  )
                ) %>%
  dplyr::mutate(RepresentativeAccession = dplyr::case_when(
                  pnc == "Pneumococcus NOT isolated" ~ "None",
                  !is.na(SingleIsolateLane) ~ SingleIsolateAccession,
                  !is.na(DeepSeqLane) ~ DeepSeqAccession,
                  TRUE ~ NA_character_
                )
              ) %>%
  # Infer characteristics from repeated observation of isolate
  dplyr::group_by(codenum,GPSC) %>%
  tidyr::fill(InferredSerotype,.direction="updown") %>%
  dplyr::ungroup() %>%
  dplyr::group_by(codenum,serotype) %>%
  tidyr::fill(InferredGPSC,.direction="updown") %>%
  dplyr::ungroup()

# Fill in unambiguous GPSCs from single isolate data
# Identify unambiguous serotype-GPSC pairings from overall single isolate Maela cohort
serotype_gpsc_links.df <-
  isolate.sample.info %>%
    dplyr::group_by(serotype) %>%
    dplyr::mutate(serotype_count = n()) %>%
    dplyr::ungroup() %>%
    dplyr::group_by(serotype,GPSC) %>%
    dplyr::mutate(genotype_count = n()) %>%
    dplyr::ungroup() %>%
    dplyr::filter(genotype_count == serotype_count) %>%
    dplyr::group_by(serotype,GPSC) %>%
    dplyr::slice_max(n = 1, order_by = SingleIsolateAccession) %>%
    dplyr::ungroup() %>%
    dplyr::select(serotype,GPSC,SingleIsolateLane,SingleIsolateAccession) %>%
    dplyr::distinct() %>%
    dplyr::rename(UnambiguousSerotype = serotype) %>%
    dplyr::rename(UnambiguousGPSC = GPSC) %>%
    dplyr::rename(RepresentativeLane = SingleIsolateLane) %>%
    dplyr::rename(RepresentativeAccession = SingleIsolateAccession)

# Infer GPSC from wider cohort and identify representatives
combined.info %<>%
    dplyr::left_join(serotype_gpsc_links.df %>%
                       dplyr::rename(SerotypeRepresentativeLane = RepresentativeLane) %>%
                       dplyr::rename(SerotypeRepresentativeAccession = RepresentativeAccession),
                     by = c("serotype"="UnambiguousSerotype")) %>%
    dplyr::left_join(serotype_gpsc_links.df %>%
                       dplyr::rename(GPSCRepresentativeLane = RepresentativeLane) %>%
                       dplyr::rename(GPSCRepresentativeAccession = RepresentativeAccession),
                     by = c("GPSC"="UnambiguousGPSC"))

###################
# REPRESENTATIVES #
###################

# Fill in GPSC/serotype matches from the same mother/child pair 
combined.info %<>%
  dplyr::group_by(codenum,InferredGPSC,InferredSerotype) %>%
  dplyr::arrange(RepresentativeEvidence) %>%
  tidyr::fill(RepresentativeGenome,.direction="down") %>%
  tidyr::fill(RepresentativeAccession,.direction="down") %>%
  dplyr::mutate(RepresentativeGenome = dplyr::if_else((is.na(InferredGPSC)|is.na(InferredSerotype)),
                                                      NA_character_,
                                                      RepresentativeGenome)) %>%
  dplyr::mutate(RepresentativeAccession = dplyr::if_else((is.na(InferredGPSC)|is.na(InferredSerotype)),
                                                      NA_character_,
                                                      RepresentativeAccession)) %>%
  dplyr::mutate(RepresentativeEvidence = 
                  dplyr::if_else((is.na(RepresentativeEvidence) & !is.na(RepresentativeGenome)),
                  "Matched serotype & GPSC within mother-child pair",
                  as.character(RepresentativeEvidence))
                ) %>%
  dplyr::ungroup()

# Fill in GPSC/serotype matches from the any mother/child pair in the cohort
combined.info %<>%
    dplyr::group_by(InferredGPSC,InferredSerotype) %>%
    dplyr::arrange(RepresentativeEvidence) %>%
    tidyr::fill(RepresentativeGenome,.direction="down") %>%
    tidyr::fill(RepresentativeAccession,.direction="down") %>%
    dplyr::mutate(RepresentativeGenome = dplyr::if_else((is.na(InferredGPSC)|is.na(InferredSerotype)),
                                                      NA_character_,
                                                      RepresentativeGenome)) %>%
    dplyr::mutate(RepresentativeAccession = dplyr::if_else((is.na(InferredGPSC)|is.na(InferredSerotype)),
                                                      NA_character_,
                                                      RepresentativeAccession)) %>%
    dplyr::mutate(RepresentativeEvidence = 
                  dplyr::if_else((is.na(RepresentativeEvidence) & !is.na(RepresentativeGenome)),
                 "Matched serotype & GPSC within immunology cohort",
                 RepresentativeEvidence)
                ) %>%
    dplyr::ungroup()

# Identify representatives from across Maela cohort
combined.info %<>%
    # Infer unambiguous pairings where representatives can be identified
    dplyr::mutate(InferUnambiguousGPSC = (is.na(InferredGPSC) & !is.na(UnambiguousGPSC) & is.na(DeepSeqLane) & is.na(SingleIsolateLane))) %>%
    dplyr::mutate(InferUnambiguousSerotype = (is.na(InferredSerotype) & !is.na(UnambiguousSerotype) & is.na(DeepSeqLane) & is.na(SingleIsolateLane))) %>%
    # Infer representatives where serotype was unknown based on GPSC representative
    dplyr::mutate(RepresentativeGenome = dplyr::if_else(InferUnambiguousSerotype,
                                                      GPSCRepresentativeLane,
                                                      RepresentativeGenome)) %>%
    dplyr::mutate(RepresentativeAccession = dplyr::if_else(InferUnambiguousSerotype,
                                                           GPSCRepresentativeAccession,
                                                           RepresentativeAccession)) %>%
    dplyr::mutate(InferredGPSC = dplyr::if_else(InferUnambiguousGPSC,
                                                           as.character(UnambiguousGPSC),
                                                           InferredGPSC)) %>%
    dplyr::mutate(RepresentativeEvidence = 
                  dplyr::if_else(InferUnambiguousGPSC,
                                "Matched serotype & GPSC within Maela cohort",
                                 RepresentativeEvidence)
    ) %>%
    # Infer representatives where GPSC was unknown based on serotype representative
    dplyr::mutate(RepresentativeGenome = dplyr::if_else(InferUnambiguousGPSC,
                                                        SerotypeRepresentativeLane,
                                                        RepresentativeGenome)) %>%
    dplyr::mutate(RepresentativeAccession = dplyr::if_else(InferUnambiguousGPSC,
                                                           SerotypeRepresentativeAccession,
                                                           RepresentativeAccession)) %>%
    dplyr::mutate(InferredGPSC = dplyr::if_else(InferUnambiguousSerotype,
                                                           as.character(UnambiguousSerotype),
                                                           InferredGPSC)) 

# Arrange into sensible order
combined.info %<>%
    dplyr::arrange(codenum,age.v) %>%
    # remove unneeded columns
    dplyr::select(-c(UnambiguousGPSC,
                     SerotypeRepresentativeLane,
                     SerotypeRepresentativeAccession,
                     UnambiguousSerotype,
                     GPSCRepresentativeLane,
                     GPSCRepresentativeAccession,
                     InferUnambiguousGPSC,
                     InferUnambiguousSerotype)
                  ) %>%
    # filter inferred representatives
    dplyr::mutate(RepresentativeEvidence = dplyr::if_else((is.na(RepresentativeGenome) & is.na(DeepSeqLane) & is.na(SingleIsolateLane)),
                                                          "No genome data",
                                                          RepresentativeEvidence)) %>%
    dplyr::mutate(RepresentativeGenome = dplyr::if_else(RepresentativeEvidence %in% c("No carriage","No genome data","Single isolate data","Deep sequencing data"),
                                                        NA_character_,
                                                        RepresentativeGenome)
                ) %>%
    dplyr::mutate(RepresentativeEvidence = dplyr::if_else((!is.na(SingleIsolateLane) & !is.na(DeepSeqLane)),
                                                          "Single isolate and deep sequencing data",
                                                          RepresentativeEvidence)) %>%
    dplyr::mutate(RepresentativeAccession = dplyr::if_else((is.na(RepresentativeGenome)|RepresentativeAccession=="None"),
                                                           NA_character_,
                                                           RepresentativeAccession)) %>%
    dplyr::rename(GenomicData = RepresentativeEvidence) %>%
    dplyr::select(c("specnum", "codenum", "specdate", "dob", "age.d", "age.v", 
"category", "diagnosis_clinical", "Xray_diagnosis", "pnc", "serotype", "InferredSerotype", "sequence_type", 
"GPSC", "InferredGPSC", "sangernum", "SingleIsolateLane", "SingleIsolateAccession", "DeepSeqLane", 
"DeepSeqAccession", "RepresentativeGenome", "RepresentativeAccession","GenomicData"))

# Write output file
write.csv(combined.info,
          file="Maela_antigens_epidemiological_data.csv",
          quote=FALSE,
          row.names = FALSE)

save(combined.info,
     file="Maela_antigens_epidemiological_data.Rdata")

# Write out list of files to analyse
lanes_list.out <-
  combined.info %>%
    dplyr::select(SingleIsolateLane,DeepSeqLane,RepresentativeGenome) %>%
    tidyr::pivot_longer(c(SingleIsolateLane,DeepSeqLane,RepresentativeGenome),names_to = "Labels",values_to = "Lanes") %>%
    dplyr::select(Lanes) %>%
    dplyr::distinct() %>%
    dplyr::filter(!is.na(Lanes)) %>%
    dplyr::pull()

write.table(lanes_list.out,
          file="Maela_antigens_lanes_list.tab",
          quote=FALSE,
          row.names = FALSE,
          col.names = FALSE)

```

This identified `{r} length(lanes_list.out)` assemblies as providing information on these carriage episodes.

```{bash, eval = FALSE}

# Load environment
module load common-apps/anaconda3/2011.11
source activate /lustre/scratch118/infgen/team284/nc3/conda/envs/maela_anti
export PERL5LIB=$CONDA_PREFIX/lib/perl5/site_perl/5.22.0/:$PERL5LIB
module load cogsoft/04.19.2012-c2
module load blast/2.7.1=h96bfa4b_5-c1
# Get assemblies
pf assembly -t file -i Maela_antigens_lanes_list.tab -l .
# Annotate and process assemblies
./run_clustering.pl -r Spn23F.dna *.contigs_velvet.fa
# Convert files from independent clustering format to querying format
bsub -o conv.o -e conv.e -M 1000 -R 'select[mem>1000] rusage[mem=1000]' ./convert_to_maela.pl 
# Get filtered non-redundant protein set in new directory
./get_nr_proteins.sh
mv blastDB/ maela.blastDB
cd maela.blastDB
segmasker -in maela_nr_proteins.aa -infmt fasta -outfmt fasta -out filtered.maela_nr_proteins.aa
# Run self-self BLAST
makeblastdb -in maela_nr_proteins.aa -dbtype prot
makeblastdb -in filtered.maela_nr_proteins.aa -dbtype prot
cd ..
module load blat/36--0
mkdir self_blaf
mkdir self_blan
bsub -o self_blast.o -e self_blast.e -M 1000 -R 'select[mem>1000] rusage[mem=1000]' ./run_blast.sh
bsub -o filtered_self_blast.o -e filtered_self_blast.e -M 1000 -R 'select[mem>1000] rusage[mem=1000]' ./run_filtered_blast.sh
# Run BLAST against original clustering
mkdir ma_blan
mkdir ma_blaf
mkdir ma.blastDB
sqlite3 /lustre/scratch118/infgen/team284/nc3/Mass/pub_run/protein_sequences.db 'select * from Proteins group by Sequence' | perl -lane 'my @d = split(/\|/,$_); print ">$d[0]\n$d[1]";' > ma.blastDB/ma_nr_proteins.aa
cd ma.blastDB
segmasker -in ma_nr_proteins.aa -infmt fasta -outfmt fasta -out filtered.ma_nr_proteins.aa
makeblastdb -in ma_nr_proteins.aa -dbtype prot
makeblastdb -in filtered.ma_nr_proteins.aa -dbtype pro
cd ..
bsub -o ma_blast.o -e ma_blast.e -M 1000 -R 'select[mem>1000] rusage[mem=1000]' ./run_ma_blast.sh
bsub -o filtered_ma_blast.o -e filtered_ma_blast.e -M 1000 -R 'select[mem>1000] rusage[mem=1000]' ./run_filtered_ma_blast.sh
# Prepare files for COGcognitor
./make_cognitor_input.pl
mkdir BLASTconv
$CONDA_PREFIX/bin/COGmakehash -i=combined.input.cds.csv -o=./BLASTconv -s=, -n=1
bsub -o rb.o -e rb.e -M 4000 -R 'select[mem>4000] rusage[mem=4000]'  $CONDA_PREFIX/bin/COGreadblast -d=./BLASTconv/ -e=0.000001 -q=1 -t=2 -v -f=ma_blaf -s=self_blaf -u=ma_blan
bsub -o cognitor.o -e cognitor.e -M 4000 -R 'select[mem>4000] rusage[mem=4000]' $CONDA_PREFIX/bin/COGcognitor -i=./BLASTconv/ -t=mass.input.cls.csv -q=maela.input.cls.csv -o=maela.cognitor.out
bsub -o pp.o -e pp.e -M 12000 -R 'select[mem>12000] rusage[mem=12000]'  ./process_cognitor_output.sh
bsub -o anno_all.o -e anno_all.e -M 1000 -R 'select[mem>1000] rusage[mem=1000]' ./annotate_all_proteins.pl

```

-   Plot separation of GPSCs for single isolates by accessory content

-   Plot proportion of cohort exposed to each antigen over time

Some other evaluations

```{r Evalate assemblies, fig.width = 8, fig.height = 10}

load("Maela_antigens_epidemiological_data.Rdata")

strain.df <-
  read.table(file = "maela_isolate_index.tab",
             comment.char = "")
colnames(strain.df) <- c("index","lane")

representative_single_isolates <- 
  unique(combined.info$RepresentativeGenome[combined.info$GenomicData=="Matched serotype & GPSC within Maela cohort"])

lane.df <-
    combined.info %>%
      dplyr::select(InferredGPSC,InferredSerotype,DeepSeqLane,SingleIsolateLane,RepresentativeGenome) %>%
      tidyr::pivot_longer(cols = c(DeepSeqLane,SingleIsolateLane,RepresentativeGenome),
                          names_to = "lane_type",
                          values_to = "lane") %>%
      dplyr::mutate(lane_type = dplyr::case_when(
            lane_type %in% c("SingleIsolateLane") ~ "single_isolate",
            lane_type %in% c("DeepSeqLane") ~ "deep_seq",
            lane %in% representative_single_isolates ~ "single_isolate",
            TRUE ~ NA_character_
          )
      ) %>%
      dplyr::filter(!is.na(lane)) %>%
      dplyr::group_by(lane) %>%
      tidyr::fill(lane_type,.direction = "updown") %>%
      dplyr::ungroup() %>%
      dplyr::distinct() %>%
      dplyr::group_by(lane) %>%
      dplyr::mutate(genotype_count = n()) %>%
      dplyr::ungroup()

assemblies.df <-
  read.table(file = "assembly_stats.tab", header = TRUE, comment.char = "") %>%
    dplyr::mutate(lane = gsub(".contigs_velvet.fa","",File)) %>%
    dplyr::left_join(lane.df, by=c("lane"))

length_plot <-
  ggplot(assemblies.df,
         aes(x = Length,
             colour = lane_type,
             fill = lane_type)) +
    geom_density(alpha = 0.2) +
    xlab("Total assembly length (bp)") +
    theme_bw() +
    theme(legend.position = "bottom")

n50_plot <-
  ggplot(assemblies.df,
         aes(x = N50,
             colour = lane_type,
             fill = lane_type)) +
    geom_density(alpha = 0.2) +
    xlab(expression("N50 (bp)")) +
    theme_bw() +
    theme(legend.position = "bottom") +
    scale_colour_manual(labels = c("Deep sequencing datasets","Single isolate datasets"),
                        values = c("deep_seq" = "red","single_isolate" = "blue"),
                        name = "Dataset type",
                        aesthetics = c("colour","fill"))

ncontig_plot <-
  ggplot(assemblies.df,
         aes(x = n_contigs,
             colour = lane_type,
             fill = lane_type)) +
    geom_density(alpha = 0.2) +
    xlab(expression("Number of contigs")) +
    scale_x_continuous(trans = "log10") +
    theme_bw() +
    theme(legend.position = "bottom") +
    scale_colour_manual(labels = c("Deep sequencing datasets","Single isolate datasets"),
                        values = c("deep_seq" = "red","single_isolate" = "blue"),
                        name = "Dataset type",
                        aesthetics = c("colour","fill"))

assembly_qc_plot <-
  ggplot(assemblies.df,
         aes(x = Length,
             y = n_contigs,
             colour = lane_type,
             fill = lane_type)) +
    geom_point(alpha = 0.2) +
    ylab("Number of contigs") +
    xlab("Total assembly length (bp)") +
    scale_x_continuous(trans = "log10") +
    scale_y_continuous(trans = "log10") +
    theme_bw() +
    theme(legend.position = "bottom") +
    scale_colour_manual(labels = c("Deep sequencing datasets","Single isolate datasets"),
                        values = c("deep_seq" = "red","single_isolate" = "blue"),
                        name = "Dataset type",
                        aesthetics = c("colour","fill"))

cowplot::plot_grid(
  plotlist = list(
    cowplot::plot_grid(
      plotlist = list(
        n50_plot + theme(legend.position = "none"),
        ncontig_plot + theme(legend.position = "none"),
        assembly_qc_plot + theme(legend.position = "none")
      ),
      nrow = 3,
      ncol = 1,
      labels = "auto"
    ),
    cowplot::get_legend(n50_plot)
  ),
  nrow = 2,
  rel_heights = c(0.95,0.05)
)

```

Note that all PBP alleles are treated as equivalent.

```{r Link COGs to sequences}

cog.df <-
  read.csv(archive::archive_read("maela_cog_assignments.csv.tar.gz"), header = FALSE)
colnames(cog.df) <- c("cds","cog")

cog_antigens <-
  rownames(igg.data)
cog_antigens <- c(cog_antigens[(!grepl("_",cog_antigens) & !(cog_antigens %in% c("ZmpE","PsrP","LytA","PblB","CBD","PhageAmi")) & !is.na(cog_antigens))],
                  "CLS00355",
                  "CLS00381",
                  "CLS01435"
)

mass.cog.df <-
  read.csv(archive::archive_read("massachusetts_cog_assignments.csv.tar.gz"))
colnames(mass.cog.df) <- c("seq","cog")
mass.cog.frequency.df <-
  mass.cog.df %>%
    tidyr::separate(seq,"orf",into=c("index","cds_index"),remove=FALSE) %>%
    dplyr::select(index,cog) %>%
    dplyr::mutate(present = 1) %>%
    dplyr::distinct() %>%
    tidyr::complete(cog,index,fill = list(present = 0)) %>%
    dplyr::group_by(cog) %>%
    dplyr::summarise(mass_frequency = mean(present)) %>%
    dplyr::ungroup()

maela.cog.df <-
  cog.df %>%
      tidyr::separate(cds,"orf",into=c("index","cds_index"),remove=FALSE) %>%
      dplyr::mutate(index = as.integer(gsub("seq","",index))) %>%
      dplyr::left_join(
        strain.df %>%
            dplyr::mutate(lane = gsub(".contigs_velvet.fa","",lane)),
        by=c("index")
      ) %>%
      dplyr::select(cog,lane) %>%
      dplyr::filter(lane != "Spn23F.dna") %>%
      dplyr::mutate(cog = factor(cog, levels = cog_antigens)) %>%
      dplyr::filter(cog %in% cog_antigens) %>%
      dplyr::mutate(present = 1) %>%
      dplyr::distinct() %>%
      tidyr::complete(cog,lane,fill = list(present = 0))

cog.frequency.df <-
  maela.cog.df %>%
      dplyr::left_join(lane.df,by=c("lane")) %>%
      dplyr::filter(lane_type == "single_isolate") %>%
      dplyr::group_by(cog) %>%
      dplyr::summarise(maela_frequency = mean(present)) %>%
      dplyr::ungroup() %>%
      dplyr::left_join(mass.cog.frequency.df,by=c("cog"))

outliers.df <-
  dplyr::bind_rows(
    cog.frequency.df %>%
      dplyr::filter(mass_frequency < 0.6 & maela_frequency > 0.8) %>%
      dplyr::mutate(outlier = "Too high"),
    cog.frequency.df %>%
      dplyr::filter(mass_frequency > 0.35 & maela_frequency < 0.15) %>%
      dplyr::mutate(outlier = "Too low")
  )

ggplot(cog.frequency.df,
       aes(x = mass_frequency,
           y = maela_frequency)) +
  geom_point() +
  ylab("Frequency in Maela dataset") +
  xlab("Frequency in Massachusetts dataset") +
  geom_smooth(method = "lm", formula = y ~ x) +
  ggrepel::geom_label_repel(data = outliers.df,
                            aes(label = cog, colour = outlier),
                            ) +
  scale_color_manual(values = c("Too high" = "red","Too low" = "orange")) +
  stat_cor(method = "pearson",
             p.accuracy = 0.000001,
             label.sep='\n'
  ) +
  theme_bw() +
  theme(legend.position = "none")

cog.frequency.df[cog.frequency.df$mass_frequency < 0.6 & cog.frequency.df$maela_frequency > 0.8,]
cog.frequency.df[cog.frequency.df$mass_frequency > 0.35 & cog.frequency.df$maela_frequency < 0.15,]

# Annotate frequency outliers

# for f in $(grep CLS00934 all.strains.filtered.cls.csv | awk -F ',' '{print $1}'); do grep -A 1 ${f} maela.blastDB/maela_nr_proteins.aa; done
# grep -A 1 MVMVFDANRIISEDSEGFVIPHGDHNHYIKVQTKGY ma.blastDB/ma_nr_proteins.aa
# CLS00934 - too high - different start codons means this is actually a merger of CLS00934, CLS04012 and CLS02423, all ICE-type proteins found on PPI-1 of different lengths - total frequency of these COGs is 0.94987 - fragmentary - mass.cog.frequency.df[mass.cog.frequency.df$cog %in% c("CLS00934","CLS04012","CLS02423"),]
# CLS00904 - too high - histidine triad protein fragment - highly variable, often fragmentary assemblies
# CLS01596 - too high - matches to CLS00397 (Glyoxalase) - fragments of a glyoxalase that is often a pseudogene - mass.cog.frequency.df[mass.cog.frequency.df$cog %in% c("CLS00397","CLS01596"),]

# CLS01009 - too low - putative hydrolase (pseudogene)
# CLS01954 - too low - hsdS fragment
# CLS02419 - too low - short fragment
# CLS02572 - too low - IS3-Spn1 orf A (pseudogene)
# CLS02595 - too low - putative periplasmic solute-binding transport protein (pseudogene)
# CLS00856 - too low - putative alanine dehydrogenase 1 (pseudogene)
# CLST2327713 - too low - transposase

```

```{r COG frequencies in deep sequencing data}

deep_seq.cog.frequency.df <-
  maela.cog.df %>%
      dplyr::left_join(lane.df,by=c("lane")) %>%
      dplyr::group_by(lane_type) %>%
      dplyr::mutate(type_count = n()) %>%
      dplyr::ungroup() %>%
      dplyr::group_by(cog,lane_type) %>%
      dplyr::summarise(frequency = mean(present)) %>%
      dplyr::ungroup() %>%
      dplyr::distinct() %>%
      tidyr::pivot_wider(id_cols = cog, values_from = frequency, names_from = lane_type)

outliers.df <-
  dplyr::bind_rows(
    deep_seq.cog.frequency.df %>%
      dplyr::filter(single_isolate < 0.45 & deep_seq > 0.6) %>%
      dplyr::mutate(outlier = "Too high"),
    deep_seq.cog.frequency.df %>%
      dplyr::filter(single_isolate > 0.35 & deep_seq < 0.15) %>%
      dplyr::mutate(outlier = "Too low")
  )

ggplot(deep_seq.cog.frequency.df,
       aes(x = single_isolate,
           y = deep_seq)) +
  geom_point() +
  xlab("Frequency in single isolate genomes") +
  ylab("Frequency in deep sequencing datasets") +
  geom_smooth(method = "lm", formula = y ~ x) +
  ggrepel::geom_label_repel(data = outliers.df,
                            aes(label = cog, colour = outlier),
                            ) +
  scale_color_manual(values = c("Too high" = "red","Too low" = "orange")) +
  stat_cor(method = "pearson",
             p.accuracy = 0.000001,
             label.sep='\n'
  ) +
  theme_bw() +
  theme(legend.position = "none")

# CLS01741 - IS630-Spn1 transposase fragment
# CLS02091 - transposase fragment
# CLS02572 - hypotheetical protein fragment

```

```{r Generate matrix and analyse distribution by GPSC, fig.height = 11, fig.width = 8}

# Convert to matrix-type dataset
maela_presence_absence.df <-
  maela.cog.df %>%
      tidyr::pivot_wider(names_from = cog, values_from = present)

maela_presence_absence.mat <-
  as.matrix(
      tibble::column_to_rownames(maela_presence_absence.df, var = "lane")
  )

# Get subset of single isolate genomes
maela_single_isolate_presence_absence.mat <-
  as.matrix(
    maela_presence_absence.df %>%
      dplyr::left_join(lane.df %>% dplyr::select(lane,lane_type), by = c("lane")) %>%
      dplyr::filter(lane_type == "single_isolate") %>%
      dplyr::select(-lane_type) %>%
      tibble::column_to_rownames(var = "lane")
  )


(single_isolate_tsne <-
  plot_tsne(maela_single_isolate_presence_absence.mat,
            lane.df,
            perplexity = 40,
            match_by = "lane",
            colour_by = "InferredGPSC",
            group_by = "InferredGPSC",
            group_min = 5
            )
)

```

```{bash Psp mapping analysis, eval = FALSE}

# Run mapping to identify psp sequences
module load common-apps/anaconda3/2011.11
source activate /lustre/scratch118/infgen/team284/nc3/conda/envs/maela_anti
export PERL5LIB=$CONDA_PREFIX/lib/perl5/site_perl/5.22.0/:$PERL5LIB
module load samtools/1.9--h91753b0_8
cd /lustre/scratch118/infgen/team284/nc3/Maela_acc_genome/deep_sequencing_genomes/antigen_mapping
pf data -t file -i ../Maela_antigens_lanes_list.tab -l .
declare -i N; N=1; for f in $(ls -d [123456789]*_*); do echo -en "gunzip -c ${f}/${f}_1.fastq.gz > ${f}_1.fastq\ngunzip -c ${f}/${f}_2.fastq.gz > ${f}_2.fastq\n~nc3/Scripts/psp_mapping_analysis.pl -f ${f}_1.fastq -r ${f}_2.fastq -s filtered.pspA_upstream_region.fa -d trimmed.recurated.pspA.fa -m -o ${f}_pspA\n~nc3/Scripts/psp_mapping_analysis.pl -f ${f}_1.fastq -r ${f}_2.fastq -s filtered.pspC_upstream_region.fa -d trimmed.recurated.pspC.fa -m -o ${f}_pspC\nrm ${f}_1.fastq ${f}_2.fastq\n" > psp_job.$N; chmod +x ./psp_job.$N; N=($N+1); done
bsub -M 4000 -R 'select[mem>4000] rusage[mem=4000]' -J psp_mapping[1-1117]%100 -o psp_mapping.%I.o -e psp_mapping.%I.e ./psp_job.\${LSB_JOBINDEX}

```

Make variable antigen translation tables, and calculate how many distinct antigens may be present within each sequence dataset:

```{r Generate variable antigen translation tables}

assign_by_matches <- function(df,threshold = 0) {
   df %<>%
    # get top rank for each representative on the array
    dplyr::group_by(lane,protein,lane_type,InferredGPSC) %>%
    dplyr::mutate(min_rank = min(rank)) %>%
    dplyr::ungroup() %>%
    # get number of representatives equivalent to number of genotypes
    dplyr::filter(min_rank <= genotype_count & mapping > threshold) %>%
    dplyr::select(lane,protein,lane_type,InferredGPSC,genotype_count) %>%
    #dplyr::select(-allele) %>%
    dplyr::rename(allele = protein)
}

# difference here is grouping/counting - assume one cds in
# assembly per genotype
assign_zmp_by_matches <- function(df,threshold = 0) {
   df %<>%
    # get top rank for each representative on the array
    dplyr::group_by(lane,cds,lane_type,InferredGPSC) %>%
    dplyr::mutate(min_rank = min(rank)) %>%
    dplyr::ungroup() %>%
    # get number of representatives equivalent to number of genotypes
    dplyr::filter(min_rank <= genotype_count & mapping > threshold) %>%
    dplyr::select(lane,protein,lane_type,InferredGPSC,genotype_count) %>%
    #dplyr::select(-allele) %>%
    dplyr::rename(allele = protein)
}

add_lane_data <- function(df,lane.df) {
   df %>%
    dplyr::left_join(lane.df %>%
                       dplyr::select(lane,lane_type,genotype_count,InferredGPSC) %>%
                       dplyr::group_by(lane,lane_type,genotype_count) %>%
                       tidyr::nest(InferredGPSC = InferredGPSC) %>%
                       dplyr::mutate(InferredGPSC = paste(unlist(InferredGPSC),collapse = ",")) %>%
                       dplyr::distinct(),
                     by = c("lane")) %>%
    dplyr::mutate(lane_type = dplyr::if_else(lane_type == "single_isolate",
                                                         "Single isolate",
                                                         "Deep sequencing")
                  ) %>%
    dplyr::mutate(Rank = dplyr::if_else(rank <= genotype_count,"Top rank","Lower ranks"))
}

translate_psp_alleles <- function(prefix,type_file,mapping_file) {

  # Translate PspA sequences
  pspA_types.df <-
    read.table(type_file, comment.char = "") %>%
      dplyr::rename(allele = V1, protein = V2) %>%
      dplyr::mutate(protein = paste0(prefix,"_",protein))

  sequences_to_mapping.df <-
    read.table(file = mapping_file,
               header = TRUE,
               comment.char = "") %>%
      group_by(Taxon) %>%
      dplyr::slice_max(order_by = desc(Reads), n = 1) %>%
      dplyr::ungroup()

  pspA_translation.df <-
    sequences_to_mapping.df %>%
      dplyr::select(-Reads) %>%
      dplyr::left_join(pspA_types.df, by = setNames("allele",prefix)) %>%
      dplyr::rename(allele = Taxon)
}

plot_antigen_distribution <- function(df) {
  
    df %<>%
      dplyr::filter(lane_type == "Single isolate") %>%
      dplyr::select(lane,InferredGPSC,allele) %>%
      dplyr::distinct() %>%
      dplyr::group_by(InferredGPSC) %>%
      dplyr::mutate(GPSC_count = n()) %>%
      dplyr::ungroup() %>%
      dplyr::group_by(InferredGPSC,allele) %>%
      dplyr::mutate(Allele_count = n()) %>%
      dplyr::mutate(Frequency = n()/GPSC_count) %>%
      dplyr::ungroup()
    
      df %<>%
        dplyr::mutate(InferredGPSC = factor(InferredGPSC,
                                        levels = df %>%
                                                    dplyr::select(InferredGPSC) %>%
                                                    dplyr::distinct() %>%
                                                    dplyr::arrange(as.numeric(as.character(InferredGPSC))) %>%
                                                    dplyr::pull()
                                        )
                  )
      
      df %<>%
        dplyr::mutate(allele = factor(allele,
                                  levels = df %>%
                                              dplyr::arrange(InferredGPSC,desc(Allele_count)) %>%
                                              dplyr::select(allele) %>%
                                              dplyr::distinct() %>%
                                              dplyr::pull()
                                  )
                  )
  
  ggplot(df %>%
          dplyr::select(-Frequency) %>%
          dplyr::rename(Frequency = Allele_count),
       aes(x=InferredGPSC,
           y=allele,
           fill=Frequency)) +
  geom_tile() +
  scale_fill_distiller(palette = "Spectral") +
  ylab("Antigen allele") +
  xlab("GPSC") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "bottom")
}

pspA_alleles.df <- translate_psp_alleles("PspA",
                                         "pspA_representatives_to_probes.tab",
                                         "pspA_sequences_to_representatives.tab"
)

pspC_alleles.df <- translate_psp_alleles("PspC",
                                         "pspC_representatives_to_probes.tab",
                                         "pspC_sequences_to_representatives.tab"
)

```

```{r Identify PspA sequences}

pspA.raw.df <- 
  read.table(file = "maela_pspA.tab",
             comment.char = "") %>%
    dplyr::rename(Taxon = V1, allele = V2, mapping = V3) %>%
    dplyr::left_join(pspA_alleles.df, by = c("allele")) %>%
    dplyr::group_by(Taxon) %>%
    dplyr::mutate(rank = rank(desc(mapping), ties.method = "first")) %>%
    dplyr::ungroup() %>%
    dplyr::rename(lane = Taxon) %>%
    add_lane_data(lane.df)

(pspA_mapping_plot <-
  ggplot(pspA.raw.df,
         aes(x = mapping,
             fill = Rank)) +
    geom_density(alpha = 0.25) +
    scale_x_continuous(trans = "log10") +
    geom_vline(xintercept = 1500, linetype = 2) +
    xlab("Alignments to reference allele") +
    theme_bw() +
    facet_wrap(.~lane_type) +
    theme(strip.background = element_blank())
)

```

```{r Plot inferred PspA distribution, fig.height = 7, fig.width = 8}

# test by taking the top ranking allele if >600 reads
# need to check whether completed for all mappings
pspA.df <- assign_by_matches(pspA.raw.df, threshold = 1500)

plot_antigen_distribution(pspA.df)

```

```{r Identify PspC sequences}

pspC.raw.df <- 
  read.table(file = "maela_pspC.tab",
             comment.char = "") %>%
    dplyr::rename(Taxon = V1, allele = V2, mapping = V3) %>%
    dplyr::left_join(pspC_alleles.df, by = c("allele")) %>%
    dplyr::group_by(Taxon) %>%
    dplyr::mutate(rank = rank(desc(mapping), ties.method = "first")) %>%
    dplyr::ungroup() %>%
    dplyr::rename(lane = Taxon) %>%
    add_lane_data(lane.df)

(pspC_mapping_plot <-
  ggplot(pspC.raw.df,
         aes(x = mapping,
             fill = Rank)) +
    geom_density(alpha = 0.25) +
    scale_x_continuous(trans = "log10") +
    geom_vline(xintercept = 1500, linetype = 2) +
    xlab("Alignments to reference allele") +
    theme_bw() +
    facet_wrap(.~lane_type) +
    theme(strip.background = element_blank())
)

```

```{r Plot inferred PspC distribution, fig.height = 7, fig.width = 8}

# Was 500
pspC.df <- assign_by_matches(pspC.raw.df, threshold = 1500)

plot_antigen_distribution(pspC.df)

```

```{r Extract zmp sequences, eval = FALSE}

# Get ZMP sequences
bsub -M 4000 -R 'select[mem>4000] rusage[mem=4000]' -o get_zmp.o -e get_zmp.e ./get_zmp_sequences.pl
# Trim to remove LPXTG close to the N terminus
./trim_zmp.pl maela_zmpA_sequences.fa > trimmed.maela_zmpA_sequences.fa
./trim_zmp.pl maela_zmpB_sequences.fa > trimmed.maela_zmpB_sequences.fa
# Calculate distances
bsub -o kmer.a.o -e kmer.a.e -M 12000 -R 'select[mem>12000] rusage[mem=12000]' "./kmer_comparison.pl trimmed.zmpA.protein.seq trimmed.maela_zmpA_sequences.fa > maela_zmpA_distances.tab"
bsub -o kmer.b.o -e kmer.b.e -M 12000 -R 'select[mem>12000] rusage[mem=12000]' "./kmer_comparison.pl trimmed.zmpB.protein.seq trimmed.maela_zmpB_sequences.fa > maela_zmpB_distances.tab"

```

We then extract the zmp distances:

```{r Identify ZmpA sequences}

return_zmp_distances <- function(zmp_fn) {

  zmp.dist.df <-
    read.table(file = zmp_fn,
               header = TRUE,
               comment.char = "") %>%
      tidyr::separate(cds, into = c("lane","cds"), sep = "-") %>%
      dplyr::group_by(lane) %>%
      dplyr::mutate(rank = rank(desc(dist), ties.method = "first")) %>%
      dplyr::ungroup() %>%
    dplyr::mutate(Rank = dplyr::if_else(rank == 1,"Top rank","Lower ranks"))
  
  return(zmp.dist.df)

}  

zmpA.dist.df <-
  return_zmp_distances("maela_zmpA.tab") %>%
  add_lane_data(lane.df)

(zmpA_mapping_plot <-
  ggplot(zmpA.dist.df,
         aes(x = dist,
             fill = Rank)) +
    geom_density(alpha = 0.25) +
    geom_vline(xintercept = 0.575, linetype = 2) +
    xlab(expression(italic(k)*"-mer distance")) +
    theme_bw() +
    facet_wrap(.~lane_type) +
    theme(strip.background = element_blank())
)

```

```{r Identify ZmpB sequences}

zmpB.dist.df <-
  return_zmp_distances("maela_zmpB.tab") %>%
  add_lane_data(lane.df)

(zmpB_mapping_plot <-
  ggplot(zmpB.dist.df,
         aes(x = dist,
             fill = Rank)) +
    geom_density(alpha = 0.25) +
    geom_vline(xintercept = 0.475, linetype = 2) +
    xlab(expression(italic(k)*"-mer distance")) +
    theme_bw() +
    theme(legend.position = "bottom") +
    facet_wrap(.~lane_type) +
    theme(strip.background = element_blank())
)

```

```{r Make multipanel mapping plot, fig.width = 8, fig.height = 8}


cowplot::plot_grid(
  plotlist = list(
    cowplot::plot_grid(
      plotlist = list(
        pspA_mapping_plot + theme(legend.position = "none"),
        pspC_mapping_plot + theme(legend.position = "none"),
        zmpA_mapping_plot + theme(legend.position = "none"),
        zmpB_mapping_plot + theme(legend.position = "none")
      ),
      nrow = 2,
      ncol = 2,
      labels = "auto"
    ),
    cowplot::get_legend(zmpB_mapping_plot)
  ),
  nrow = 2,
  ncol = 1,
  rel_heights = c(0.95,0.05)
)

```

```{r Plot inferred ZmpA distribution, fig.height = 7, fig.width = 8}

zmpA.alleles.df <-
  read.table("massachusetts_zmpA_assignments.tab",
             header = TRUE,
             comment.char = "") %>%
  dplyr::mutate(protein = paste0("ZmpA_",ZmpA)) %>%
  dplyr::rename(allele = Taxon)

zmpA.df <-
  assign_zmp_by_matches(zmpA.dist.df %>%
                      dplyr::rename(mapping = dist, allele = zmp) %>%
                      dplyr::left_join(zmpA.alleles.df, by = c("allele")),
                    threshold = 0.575) %>%
  dplyr::distinct()

plot_antigen_distribution(zmpA.df)

```

```{r Plot inferred ZmpB distribution, fig.height = 7, fig.width = 8}

zmpB.alleles.df <-
  read.table("massachusetts_zmpB_assignments.tab",
             header = TRUE,
             comment.char = "") %>%
  dplyr::mutate(protein = paste0("ZmpB_",ZmpB)) %>%
  dplyr::rename(allele = Taxon)

zmpB.df <-
  assign_zmp_by_matches(zmpB.dist.df %>%
                      dplyr::rename(mapping = dist, allele = zmp) %>%
                      dplyr::left_join(zmpB.alleles.df, by = c("allele")),
                    threshold = 0.475) %>%
  dplyr::distinct()

plot_antigen_distribution(zmpB.df)

```

```{r Plot pspA frequency comparison}

compare_antigen_frequencies <- function(df) {
  comp_freq.df <-
    df %>%
      dplyr::select(lane_type,allele,genotype_count) %>%
      dplyr::group_by(lane_type) %>%
      dplyr::mutate(type_count = sum(genotype_count)) %>%
      dplyr::ungroup() %>%
      dplyr::group_by(lane_type,allele) %>%
      dplyr::summarise(frequency = n()/type_count) %>%
      dplyr::ungroup() %>%
      dplyr::distinct() %>%
      tidyr::pivot_wider(id_cols = allele,names_from = lane_type,values_from = frequency)
  return(comp_freq.df)
}

plot_comparison_of_antigen_frequencies <- function(comp_freq.df) {

  outlier_difference<-0.05
  outliers.df <-
    comp_freq.df %>%
      dplyr::filter((`Deep sequencing` < `Single isolate`-outlier_difference) | (`Deep sequencing` > `Single isolate`+outlier_difference))
  
  ggplot(comp_freq.df,
         aes(x = `Single isolate`,
             y = `Deep sequencing`)) +
    geom_point() +
    scale_x_continuous(trans= "log10") +
    scale_y_continuous(trans= "log10") +
    geom_smooth(method = "lm", formula = y ~ x) +
    xlab("Frequency in single isolate genomes") +
    ylab("Frequency in deep sequencing datasets") +
    ggrepel::geom_label_repel(data = outliers.df, aes(x = `Single isolate`,
                                                       y = `Deep sequencing`,
                                                       label = allele)) +
    geom_abline(intercept = 0, slope = 1, linetype = 2) +
    stat_cor(method = "pearson",
             p.accuracy = 0.000001,
             label.sep='\n'
    ) +
    theme_bw()
}

pspA.freq.df <- compare_antigen_frequencies(pspA.df)
pspA_freq_plot <-
  plot_comparison_of_antigen_frequencies(pspA.freq.df)

```

```{r Plot pspC frequency comparison}

pspC.freq.df <- compare_antigen_frequencies(pspC.df)
pspC_freq_plot <-
  plot_comparison_of_antigen_frequencies(pspC.freq.df)

```

```{r Plot zmpA frequency comparison}

zmpA.freq.df <- compare_antigen_frequencies(zmpA.df)
zmpA_freq_plot <-
  plot_comparison_of_antigen_frequencies(zmpA.freq.df)

```

```{r Plot zmpB frequency comparison}

zmpB.freq.df <- compare_antigen_frequencies(zmpB.df)
zmpB_freq_plot <-
  plot_comparison_of_antigen_frequencies(zmpB.freq.df)

```

```{r Plot multipanel frequency plots, fig.width = 8, fig.height = 8}

cowplot::plot_grid(
  plotlist = list(
    pspA_freq_plot,
    pspC_freq_plot,
    zmpA_freq_plot,
    zmpB_freq_plot
  ),
  nrow = 2,
  ncol = 2,
  labels = "auto"
)

```

```{r Create overall presence-absence matrix across lanes, fig.height = 11, fig.width = 8}

variable_antigen_list <- rownames(igg.data)
variable_antigen_list <- variable_antigen_list[grepl("PspA_",variable_antigen_list) | grepl("PspC_",variable_antigen_list) | grepl("ZmpA_",variable_antigen_list) | grepl("ZmpB_",variable_antigen_list)]

variable_antigen.df <-
  dplyr::bind_rows(
    pspA.df,
    pspC.df,
    zmpA.df,
    zmpB.df
  ) %>%
  dplyr::select(lane,allele) %>%
  dplyr::distinct() %>%
  dplyr::mutate(presence = 1) %>%
  dplyr::filter(allele %in% variable_antigen_list) %>%
  dplyr::mutate(allele = factor(allele, levels = variable_antigen_list)) %>%
  dplyr::mutate(lane = factor(lane, levels = lane.df %>% dplyr::select(lane) %>% dplyr::distinct() %>% dplyr::pull())) %>%
  tidyr::complete(allele,lane,fill = list(presence = 0)) %>%
  tidyr::pivot_wider(lane,names_from = allele,values_from = presence)

maela_variable_antigen.mat <-
  as.matrix(
    variable_antigen.df %>%
      dplyr::left_join(lane.df %>% dplyr::select(lane,lane_type), by = c("lane")) %>%
      dplyr::filter(lane_type == "single_isolate") %>%
      dplyr::select(-lane_type) %>%
      tibble::column_to_rownames(var = "lane")
  )

(single_isolate_variable_antigen_tsne <-
  plot_tsne(maela_variable_antigen.mat,
            lane.df,
            perplexity = 40,
            match_by = "lane",
            colour_by = "InferredGPSC",
            group_by = "InferredGPSC",
            group_min = 10,
            max_iter = 1000,
            ) +
  guides(colour=guide_legend(ncol=10,title = "Inferred GPSC"))
)

```

```{bash Analyse zmpE distribution, eval = FALSE}

# Search for zmpE sequences
cd  /lustre/scratch118/infgen/team284/nc3/Maela_acc_genome/deep_sequencing_genomes/zmpE
bsub -o cat.o -e cat.e -M 1000 -R 'select[mem>1000] rusage[mem=1000]' "cat ../[123456789]*.dna > maela_immunology_cohort_genomes.mfa"
makeblastdb -in maela_immunology_cohort_genomes.mfa -dbtype nucl
bsub -o zmpE_blast.o -e zmpE_blast.e -M 1000 -R 'select[mem>1000] rusage[mem=1000]' tblastn -db maela_immunology_cohort_genomes.mfa -query zmpE.aa -outfmt 6 -out maela_immunology_cohort_zmpE.blast.out

```

To analyse these data:

```{r Analyse ZmpE BLAST hits}

read_blast <- function(fn) {
  
  blast.df <- read.table(file=fn, comment.char = "", header = FALSE)
  colnames(blast.df) <- c("name1", "name2", "per_id", "aln_len", "mism", 
                                       "gaps", "start1", "end1", "start2", "end2", "e_value", 
                                       "bit_score")
  
  blast.df %<>%
    dplyr::group_by(name2) %>%
    dplyr::slice_max(order_by = bit_score, n = 1) %>%
    dplyr::ungroup()
  
  return(blast.df)
}

zmpE.blast.df <- read_blast("maela_immunology_cohort_zmpE.blast.out")

zmpE_threshold <- 2000
(zmpE_blast_plot <-
  ggplot(zmpE.blast.df,
         aes(x = aln_len)) +
    geom_density(alpha = 0.2, fill = "blue", line = "blue") +
    geom_vline(xintercept = zmpE_threshold,
               linetype = 2,
               colour = "red") +
    xlab("Length of highest-scoring BLAST alignment") +
    theme_bw()
)

make_pa_df <- function(df,gene,threshold) {
  
  df %>%
    dplyr::filter(aln_len > threshold) %>%
    dplyr::mutate(lane = gsub(".contigs_velvet.fa","",name2)) %>%
    dplyr::select(lane) %>%
    dplyr::distinct() %>%
    dplyr::mutate(!!gene := 1)
  
}

zmpE.pa.df <- make_pa_df(zmpE.blast.df,"ZmpE",zmpE_threshold)

```

Checking the top hit in `28042_1#51.contigs_velvet.fa`, the protein is adjacent to *ileS*, the recorded insertion site in Massachusetts SC12 non-typeables. The hit in `7054_6#3.contigs_velvet.fa` sits in the *zmpAD* insertion site.

```{bash Analyse phage distribution, eval = FALSE}

cd /lustre/scratch118/infgen/team284/nc3/Maela_acc_genome/deep_sequencing_genomes/phage_search
cat phi*dna prophage*dna > all_prophage.mfa
bsub -o phage_blast.o -e phage_blast.e -M 1000 -R 'select[mem>1000] rusage[mem=1000]' blastn -db maela_immunology_cohort_genomes.mfa -query all_prophage.mfa -outfmt 6 -out maela_immunology_cohort_phage.blast.out

```

```{r Analyse prophage BLAST hits}

phage.blast.df <- 
  read_blast("maela_immunology_cohort_phage.blast.out")

phage_threshold <- 6000
(phage_blast_plot <-
  ggplot(phage.blast.df,
         aes(x = aln_len)) +
    geom_density(alpha = 0.2, fill = "blue", line = "blue") +
    geom_vline(xintercept = phage_threshold,
               linetype = 2,
               colour = "red") +
    xlab("Length of highest-scoring BLAST alignment") +
    theme_bw()
)

PhageAmi.pa.df <- make_pa_df(phage.blast.df,"PhageAmi",phage_threshold)

```

```{bash Analyse PblB distribution, eval = FALSE}

cd /lustre/scratch118/infgen/team284/nc3/Maela_acc_genome/deep_sequencing_genomes/pblB_search
bsub -o pblB_blast.o -e pblB_blast.e -M 1000 -R 'select[mem>1000] rusage[mem=1000]' tblastn -db maela_immunology_cohort_genomes.mfa -query pblB.aa -outfmt 6 -out maela_immunology_cohort_pblB.blast.out

```

```{r Analyse PblB BLAST hits}

pblB.blast.df <- read_blast("maela_immunology_cohort_pblB.blast.out")

pblB_threshold <- 500
(pblB_blast_plot <-
  ggplot(pblB.blast.df,
         aes(x = aln_len)) +
    geom_density(alpha = 0.2, fill = "blue", line = "blue") +
    geom_vline(xintercept = pblB_threshold,
               linetype = 2,
               colour = "red") +
    xlab("Length of highest-scoring BLAST alignment") +
    theme_bw()
)

pblB.pa.df <- make_pa_df(pblB.blast.df,"PblB",pblB_threshold)

```

```{r Multi-panel plot of BLAST hits, fig.height = 8, fig.width = 8}

cowplot::plot_grid(
  plotlist = list(
    zmpE_blast_plot,
    phage_blast_plot,
    pblB_blast_plot
  ),
  nrow = 3,
  ncol = 1,
  labels = "auto"
)

```

```{r Summarise additional antigens and create presence-absence matrix}

# Combine all additional antigens
additional_antigens.df <-
  maela_presence_absence.df %>%
    dplyr::select(lane,CLS01513) %>%
    dplyr::rename(PsrP = CLS01513) %>%
    dplyr::left_join(
        zmpE.pa.df,
      by = c("lane")
    ) %>%
    dplyr::left_join(
        PhageAmi.pa.df,
      by = c("lane")
    ) %>%
    dplyr::left_join(
        pblB.pa.df,
      by = c("lane")
    ) %>%
    dplyr::mutate(across(where(is.numeric), ~ replace_na(.x, 0))) %>%
    dplyr::mutate(LytA = 1,
                  CBD = 1)

# Combined COGs, variable antigens and additional antigens
complete_antigen_presence_absence.df <-
  maela_presence_absence.df %>%
    dplyr::left_join(
      variable_antigen.df,
      by = c("lane")
    ) %>%
    dplyr::left_join(
      additional_antigens.df,
      by = c("lane")
    ) %>%
    dplyr::mutate(CLS00355_a1 = CLS00355,
                  CLS00355_a2 = CLS00355,
                  CLS00355_a3 = CLS00355,
                  CLS00381_a1 = CLS00381,
                  CLS00381_a2 = CLS00381,
                  CLS00381_a3 = CLS00381,
                  CLS01435_a1 = CLS01435,
                  CLS01435_a2 = CLS01435,
                  CLS01435_a3 = CLS01435,
                  CLS01435_a4 = CLS01435) %>%
    dplyr::select(-c(CLS00355,CLS00381,CLS01435))

# Convert to matrix
complete_antigen_presence_absence.mat <-
  as.matrix(
      tibble::column_to_rownames(complete_antigen_presence_absence.df, var = "lane")
  )
    
in_array_but_not_genomes<-rownames(igg.data)[!(rownames(igg.data) %in% colnames(complete_antigen_presence_absence.mat))]
in_genomes_but_not_array<-colnames(complete_antigen_presence_absence.mat)[!(colnames(complete_antigen_presence_absence.mat) %in% rownames(igg.data))]
if (length(in_array_but_not_genomes) == 0) {
  message("All array proteins in genomic analysis")
} else {
  message(paste("Array proteins missing from genomic analysis:",in_array_but_not_genomes))
}

if (length(in_genomes_but_not_array) == 0) {
  message("All genomic analysis proteins in array")
} else {
  message(paste("Genome proteins missing from array:",in_genomes_but_not_array))
}
```

# Analyse the immunological data

```{r Processing of raw data}

antigen_classification.df <-
  read.csv(archive::archive_read("antigen_classification.csv.tar.gz")) %>%
  dplyr::mutate(type = dplyr::case_when(
                    classification == "non-ABT" ~ "non-ABT",
                    classification == "ABT" & type == "COG" ~ "Other ABT",
                    TRUE ~ type
                  )
                )

antigens.df <-
  igg.data %>%
    # Reformat
    dplyr::as_tibble(rownames = "antigen") %>%
    tidyr::pivot_longer(cols = -antigen,
                        names_to = "sero_specnum",
                        values_to = "igg") %>%
    # Add functional information
    dplyr::mutate(antigen = dplyr::case_when(
                    grepl("zmp",antigen) ~ gsub("zmp","Zmp",antigen),
                    antigen == "psrP" ~ "PsrP",
                    antigen == "lytA" ~ "LytA",
                    antigen == "pblB" ~ "PblB",
                    TRUE ~ antigen
                  )
                ) %>%
    dplyr::filter(antigen != "CLS02728") %>% # left over pspC
    dplyr::left_join(antigen_classification.df, by = c("antigen")) %>%
    dplyr::mutate(type = dplyr::case_when(
                        antigen == "CBD" ~ "Adhesin",
                        antigen == "PhageAmi" ~ "CellWall",
                        antigen == "ZmpE" ~ "Degradative",
                        antigen == "LytA" ~ "CellWall",
                        antigen == "PblB" ~ "Adhesin",
                        TRUE ~ type
                      )
                  ) %>%
    # Add sample information
    dplyr::left_join(translation.df, by = c("sero_specnum")) %>%
    # Calculate change from last timepoint
    dplyr::group_by(codenum,category,antigen) %>%
    dplyr::mutate(igg_delta = igg - dplyr::lag(igg,1,order_by = sero_age)) %>%
    dplyr::ungroup() %>%
    # Change type ordering
    dplyr::mutate(type = factor(type,
                                levels = c("Adhesin",
                                           "CellWall",
                                           "Degradative",
                                           "SBP",
                                           "PspA",
                                           "PspC",
                                           "ZmpA",
                                           "ZmpB",
                                           "Other ABT",
                                           "non-ABT"
                                           )
                                )
                  ) %>%
    dplyr::mutate(prot_label = dplyr::if_else(
                      (is.na(gene) | gene == "-" | type %in% c("PspA","PspC","ZmpA","ZmpB")),
                      antigen,
                      firstup(gene)
                    )
                  )

```

```{r Plot complete frequencies across dataset, fig.width = 8, fig.height = 8}

antigen_frequencies_in_population.df <-
  as.data.frame(apply(complete_antigen_presence_absence.df[,-1],
                      2,
                      mean)
                ) %>%
  dplyr::rename(Frequency = starts_with("apply")) %>%
  tibble::rownames_to_column(var = "antigen") %>%
  dplyr::left_join(antigens.df %>%
                     dplyr::select(antigen,type) %>%
                     dplyr::distinct(),
                   by = c("antigen"))

core.list <-
  antigen_frequencies_in_population.df %>%
    dplyr::filter(Frequency > 0.99) %>%
    dplyr::select(antigen) %>%
    dplyr::pull()

low_frequency.df <-
  antigen_frequencies_in_population.df %>%
    dplyr::filter((Frequency < 0.9 & !(type %in% c("PspA","PspC","ZmpA","ZmpB","non-ABT"))) |
                  (Frequency > 0.1 & (type %in% c("PspA","PspC","ZmpA","ZmpB")))
                  ) %>%
    dplyr::left_join(antigens.df %>% dplyr::select(antigen,prot_label) %>% dplyr::distinct(),
                     by = c("antigen"))

ggplot(antigen_frequencies_in_population.df,
       aes(x = type,
           y = Frequency,
           colour = type,
           fill = type)) +
  # geom_violin(draw_quantiles = c(0.5)) +
  geom_point(position = position_jitter(width = 0.1, seed = 42), alpha = 0.5) +
  xlab("Protein Type") +
  ylab("Frequency in genomic datasets") +
  theme_bw() +
            scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  ggrepel::geom_label_repel(data = low_frequency.df,
                            aes(x = type,
                                y = Frequency,
                                label = prot_label,
                                colour = type),
                            min.segment.length = 10,
                            inherit.aes = FALSE,
                            position = position_jitter(width = 0.1, seed = 42)) +
  guides(colour = guide_legend(override.aes = list(size=1, alpha = 1)),
         fill = guide_legend(override.aes = list(size=1, alpha = 1))) +
  theme(legend.position = "bottom")

```

```{r Plot overall trends over time}

# Plot overall trends over time
ggplot(antigens.df,
       aes(x = sero_age,
           y = igg,
           group = type,
           colour = type)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.01),
             size = 0.25,
             alpha = 0.2) +
  theme_bw() +
            scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  guides(colour = guide_legend(override.aes = list(size=1, alpha = 1))) +
  theme(legend.position = "bottom")

```

-   The variation within each class is greater than the variation between classes, or over time.

-   ZmpA/B ratio changes over time, as does adhesin/SBP ratio

```{r Summary stats on individuals}

table(antigens.df %>% dplyr::select(codenum,sero_age) %>% dplyr::distinct() %>% dplyr::select(sero_age)) %>%
  kableExtra::kable() %>%
  kableExtra::kable_styling(latex_options = "scale_down")

```

```{r Summary stats on antigens}

table(antigens.df %>% dplyr::select(antigen,type) %>% dplyr::distinct() %>% dplyr::select(type)) %>%
  kableExtra::kable() %>%
  kableExtra::kable_styling(latex_options = "scale_down")

```

```{r Test whether antigens cluster together, fig.height = 10, fig.width = 8}

pbp_alleles <- c("CLS00355_a1","CLS00355_a2","CLS00355_a3","CLS00381_a1","CLS00381_a2","CLS00381_a3","CLS01435_a1","CLS01435_a2","CLS01435_a3","CLS01435_a4")
pclA_alleles <- c("CLS01333", "CLS99466", "CLS03178", "CLS03616", "CLS03265")

(all_antigens <-
  plot_tsne(igg.data,
            antigens.df %>%
              dplyr::select(antigen,type,gene) %>%
              dplyr::distinct() %>%
              dplyr::mutate(type = dplyr::case_when(
                                antigen %in% pbp_alleles ~ "PBP",
                                type == "CellWall" ~ "CellWall (not PBP)",
                                antigen %in% pclA_alleles ~ "PclA",
                                type == "Adhesin" ~ "Adhesin (not PclA)",
                                TRUE ~ type
                              )
                            ),
            perplexity = 40,
            match_by="antigen",
            colour_by="type",
            group_by="type",
            group_min=5,
            max_iter=1000) +
            scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type")
)

```

```{r Plot t-SNE across individuals, fig.height = 11, fig.width = 8, messages = FALSE}

(all_individuals <-
  plot_tsne(t(igg.data),
            antigens.df,
            perplexity = 40,
            match_by="sero_specnum",
            colour_by="codenum",
            group_by="codenum",
            shape_by="sero_age",
            group_min=3,
            max_iter=50000) +
  scale_shape_manual(values = sero_age_shapes, name = "Age") +
  guides(colour = guide_legend(title.position = "top",
                               title = "Household"),
         shape = guide_legend(title.position = "top"))
)

```

```{r Plot t-SNE across individuals before six months, fig.height = 11, fig.width = 8, messages = FALSE}

early_igg.data <-
  igg.data[,colnames(igg.data) %in% antigens.df$sero_specnum[antigens.df$sero_age %in% c("Cord","Birth","6 months")]]

(all_individuals_early <-
  plot_tsne(t(early_igg.data),
            antigens.df[antigens.df$sero_age %in% c("Cord","Birth","6 months"),],
            perplexity = 40,
            match_by="sero_specnum",
            colour_by="codenum",
            group_by="codenum",
            shape_by="sero_age",
            group_min=2,
            max_iter=5000) +
  scale_shape_manual(values = sero_age_shapes, name = "Age") +
  guides(colour = guide_legend(title.position = "top",
                               title = "Household"),
         shape = guide_legend(title.position = "top"))
)

```

```{r Plot t-SNE across individuals after six months, fig.height = 11, fig.width = 8, messages = FALSE}

late_igg.data <-
  igg.data[,colnames(igg.data) %in% antigens.df$sero_specnum[(antigens.df$sero_age %in% c("12 months","18 months","24 months"))]]

(all_individuals_late <-
  plot_tsne(t(late_igg.data),
            antigens.df[(antigens.df$sero_age %in% c("12 months","18 months","24 months")),],
            perplexity = 40,
            match_by="sero_specnum",
            colour_by="codenum",
            group_by="codenum",
            shape_by="sero_age",
            group_min=2,
            max_iter=5000) +
  scale_shape_manual(values = sero_age_shapes, name = "Age") +
  guides(colour = guide_legend(title.position = "top",
                               title = "Individual"),
         shape = guide_legend(title.position = "top",
                              title="Sample type"))
)

```

```{r Plot t-SNE to test for batch effects, fig.height = 11, fig.width = 8, messages = FALSE}

(all_individuals_late_by_slide <-
  plot_tsne(t(late_igg.data),
            antigens.df[(antigens.df$sero_age %in% c("12 months","18 months","24 months")),] %>%
              dplyr::left_join(pheno.df %>% dplyr::select(Sample.ID,Slide.ID),
                               by = c("sero_specnum"="Sample.ID")),
            perplexity = 40,
            match_by="sero_specnum",
            colour_by="codenum",
            group_by="Slide.ID",
            shape_by="sero_age",
            group_min=2,
            max_iter=50000) +
  scale_shape_manual(values = sero_age_shapes, name = "Age") +
  guides(colour = guide_legend(title.position = "top",
                               title = "Individual"),
         shape = guide_legend(title.position = "top",
                              title="Sample type"))
)

```

## Run SIMLR clustering

```{r SIMLR clustering of samples, eval = FALSE}

cluster_range <- 2:10

cluster_number_scoring <-
  SIMLR::SIMLR_Estimate_Number_of_Clusters(igg.data,NUMC = cluster_range)

save(cluster_number_scoring,
     file="sample_clustering_scoring.Rdata")

number_of_clusters <-
  cluster_range[which.min(cluster_number_scoring$K1)]

sample_clustering_output <-
  SIMLR::SIMLR(igg.data, # igg.data[!(rownames(igg.data) %in% c("CLS01991","CLS02425","CLS01450")),], - unstable clustering caused by CLS02425
               k = 12, # stable to the inclusion of CLS02425 (same as k=10); similar to k=15 with CLS02425; hence more robust than k=10
               number_of_clusters)

save(sample_clustering_output,
     file="sample_clustering_output.Rdata")

```

```{r Plot number of SIMLR clusters, fig.height = 6, fig.width = 8}

load(file="sample_clustering_scoring.Rdata")

ggplot(data = tibble(
                    "Number of clusters" = 2:10,
                    "Score" = cluster_number_scoring$K1,
                  ),
        aes(x = `Number of clusters`,
            y = Score)) +
  geom_line() +
  geom_point() +
  theme_bw()

```

```{r Plot SIMLR output, fig.height = 9, fig.width = 8}

load(file="sample_clustering_output.Rdata")

clustering_plot.df <-
  tibble::as_tibble(
      cbind(colnames(igg.data),sample_clustering_output$ydata,sample_clustering_output$y$cluster)
  ) %>%
  dplyr::rename(sero_specnum = V1,
                x = V2,
                y = V3,
                cluster = V4) %>%
  dplyr::mutate(
    x = as.numeric(x),
    y = as.numeric(y),
    cluster = factor(cluster)
    ) %>%
  dplyr::left_join(translation.df,
                   by = c("sero_specnum")) %>%
  dplyr::left_join(cluster_names,
                   by = "cluster") %>%
  dplyr::select(-cluster) %>%
  dplyr::rename(cluster = cluster_name)


ggplot(clustering_plot.df,
       aes(x = x,
           y = y,
           colour = cluster,
           groups = cluster)) +
  geom_point(size = 0.25, alpha = 0.25) +
  stat_ellipse(type = "t") +
  scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster") +
  theme_bw() +
  xlab("SIMLR dimension 1") +
  ylab("SIMLR dimension 2") +
  theme(legend.position = "bottom")

```

```{r Plot clusters by age, fig.height = 8, fig.width = 8}

(clustering_by_age_plot <- 
  ggplot(clustering_plot.df,
         aes(x = sero_age,
             y = codenum,
             fill = cluster)) +
    geom_tile() +
    scale_colour_manual(values = cluster_palette,
                        aesthetics = c("fill","colour"),
                        drop = TRUE, limits = force,
                        name = "Immune cluster") +
    theme_bw() +
    ylab("Individual") +
    xlab("Age") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
)

```

```{r Fisher exact test on age}

clustering_contingency_tables.df <-
  clustering_plot.df %>%
    dplyr::select(codenum,sero_age,cluster) %>%
    dplyr::filter(sero_age %in% c("6 months","12 months","18 months","24 months")) %>%
    dplyr::mutate(sero_age = factor(sero_age)) %>%
    dplyr::mutate(infant_low = dplyr::if_else(cluster == "Infant low","Infant low","Not infant low")) %>%
    dplyr::select(-cluster)

clustering_by_age.tab <-
  table(clustering_contingency_tables.df$sero_age,clustering_contingency_tables.df$infant_low)

clustering_by_age.tab.test <-
  fisher.test(clustering_by_age.tab)

clustering_by_age.tab.test

```

```{r Fisher exact test on individual}

clustering_by_early_late.df <-
  clustering_contingency_tables.df %>%
    dplyr::filter(sero_age %in% c("6 months","24 months")) %>%
    tidyr::pivot_wider(id_cols = codenum,names_from = sero_age,values_from = infant_low) %>%
    dplyr::select(-codenum)

early_late_clustering.tab <-
  table(clustering_by_early_late.df$`6 months`,clustering_by_early_late.df$`24 months`)

early_late_clustering.test <-
  fisher.test(early_late_clustering.tab)

early_late_clustering.test

```

```{r Identify discriminant antigens, messages = FALSE}

antigen_importance.ranking <-
  SIMLR::SIMLR_Feature_Ranking(sample_clustering_output$S, igg.data)

antigen_importance.df <-
  tibble::as_tibble(
    cbind(
      rownames(igg.data),
      antigen_importance.ranking$pval,
      antigen_importance.ranking$aggR
    )
  ) %>%
  dplyr::rename(
    antigen = V1,
    p_val = V2,
    rank = V3
  ) %>%
  dplyr::mutate(
    p_val = as.double(p_val),
    rank = as.integer(rank)
  ) %>%
  dplyr::arrange(rank) %>%
  dplyr::left_join(antigens.df %>% dplyr::select(antigen,type,gene) %>% dplyr::distinct(),
                   by = c("antigen"))

kruskal.test(rank~type,antigen_importance.df)
```

```{r Plot distribution over ranks}

(importance_rank_plot <-
  ggplot(antigen_importance.df,
         aes(x = type,
             y = rank,
             colour = type,
             fill = type)) +
    geom_point(position = position_jitter(), size = 0.5) +
    geom_violin(alpha = 0.25,
                draw_quantiles = c(0.5),,
               scale = "width") +
    theme_bw() +
    xlab("Protein type") +
    ylab("Importance rank") +
    scale_colour_manual(values = type_palette,
                        aesthetics = c("fill","colour"),
                        drop = TRUE, limits = force,
                        name = "Protein type") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
)

```

Explanation of p values: <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3278763/>.

```{r Plot SIMLR p values, fig.width = 8, fig.height = 8}

rank_p_val_plot <-
  ggplot(antigen_importance.df,
         aes(x = type,
             y = p_val,
             colour = type,
             fill = type)) +
      geom_point(position = position_jitter(), size = 0.5) +
      geom_violin(alpha = 0.25,
                  draw_quantiles = c(0.5),
               scale = "width") +
    xlab("Protein type") +
    ylab(expression(rho~"value")) +
    scale_colour_manual(values = type_palette,
                        aesthetics = c("fill","colour"),
                        drop = TRUE, limits = force,
                        name = "Protein type") +
    scale_y_continuous(trans = "log10") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom")

simlr_plot_legend <-
  cowplot::get_legend(rank_p_val_plot)

cowplot::plot_grid(
  plotlist = list(
    cowplot::plot_grid(plotlist = list(
                          importance_rank_plot + theme(legend.position = "none"),
                          rank_p_val_plot + theme(legend.position = "none")
                        ),
                        nrow = 1,
                        ncol = 2,
                        labels = "auto"
                       ),
    simlr_plot_legend
  ),
  nrow = 2,
  ncol = 1,
  rel_heights = c(0.9,0.1)
)

```

-   Cell wall and SBPs cause relatively little distinctiveness

-   Degradative (e.g. bgaA) and PspC and adhesion cause more distinctiveness

```{r Test for skewness}

library(moments)

get_stats <- function(protein, antigen_df = antigens.df) {
  values <- antigen_df %>% dplyr::filter(antigen == protein) %>% dplyr::select(igg) %>% dplyr::pull()
  return(
    as.data.frame(
      list(
      "antigen" = protein,
      "mean" = mean(values),
      "variance" = var(values),
      "median" = median(values),
      "skewness" = skewness(values),
      "kurtosis" = kurtosis(values)
      )
    )
  )
}

protein_list <- antigens.df %>% dplyr::select(antigen) %>% dplyr::distinct() %>% dplyr::pull()
igg_dist_stats.df <-
  dplyr::bind_rows(
    lapply(protein_list,get_stats,antigen_df = antigens.df)
  )

ranked_antigen_distribution.df <-
  antigen_importance.df %>%
    dplyr::left_join(igg_dist_stats.df, by = c("antigen")) %>%
    tidyr::pivot_longer(cols = c(mean,variance,median,skewness,kurtosis),
                        names_to = "statistic",
                        values_to = "value") %>%
  dplyr::mutate(statistic = factor(statistic,levels = c("mean","median","variance","skewness","kurtosis")))

ggplot(ranked_antigen_distribution.df,
       aes(x = rank,
           y = value,
           colour = type)) +
  geom_point(alpha = 0.5, size = 0.25) +
  theme_bw() +
    scale_colour_manual(values = type_palette,aesthetics = c("fill","colour")) +
  scale_y_continuous(trans = "log10") +
  scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  facet_wrap(.~statistic, scales = 'free_y') +
    theme(strip.background = element_blank())

```

```{r Plot t-SNE across individuals before six months by cluster, fig.height = 11, fig.width = 8, messages = FALSE}


six_month_classification <-
  clustering_plot.df %>%
    dplyr::group_by(codenum) %>%
    dplyr::mutate(distance = sqrt((x - dplyr::lag(x, n = 1,order_by = sero_age))**2 + (y - dplyr::lag(y, n = 1,order_by = sero_age))**2)) %>%
    dplyr::ungroup()

antigens.df %<>%
  dplyr::left_join(clustering_plot.df %>% dplyr::select(sero_specnum,cluster))

early_igg_by_cluster.data <-
  igg.data[,colnames(igg.data) %in% antigens.df$sero_specnum[antigens.df$sero_age %in% c("Cord","Birth")]]

(all_individuals_early_by_cluster <-
  plot_tsne(t(early_igg_by_cluster.data),
            antigens.df[antigens.df$sero_age %in% c("Cord","Birth"),],
            perplexity = 40,
            match_by="sero_specnum",
            colour_by="cluster",
            group_by="codenum",
            shape_by="sero_age",
            group_min=2,
            max_iter=50000) +
  scale_shape_manual(values = sero_age_shapes, name = "Age") +
  guides(colour = guide_legend(title.position = "top",
                               title = "Cluster"),
         shape = guide_legend(title.position = "top")) +
  scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster")

)

# all_individuals_early +
#   ggrepel::geom_label_repel(data=. %>% dplyr::filter(sero_age == "6 months"), aes(label = codenum))

```

```{r Plot t-SNE across individuals at six months by cluster, fig.height = 11, fig.width = 8, messages = FALSE}


antigens.df %<>%
  dplyr::left_join(clustering_plot.df %>% dplyr::select(sero_specnum,cluster))

mid_igg.data <-
  igg.data[,colnames(igg.data) %in% antigens.df$sero_specnum[antigens.df$sero_age %in% c("6 months","12 months","18 months","24 months")]]

(all_individuals_mid_by_cluster <-
  plot_tsne(t(mid_igg.data),
            antigens.df[antigens.df$sero_age %in% c("6 months","12 months","18 months","24 months"),],
            perplexity = 40,
            match_by="sero_specnum",
            colour_by="cluster",
            group_by="codenum",
            shape_by="sero_age",
            group_min=2,
            max_iter=5000) +
  scale_shape_manual(values = sero_age_shapes, name = "Age") +
  guides(colour = guide_legend(title.position = "top",
                               title = "Cluster"),
         shape = guide_legend(title.position = "top")) +
  scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster")
)

# all_individuals_mid +
#   ggrepel::geom_label_repel(data=. %>% dplyr::filter(sero_age == "6 months"), aes(label = codenum))

```

```{r Plot t-SNE across individuals after six months coloured by cluster, fig.height = 11, fig.width = 8, messages = FALSE}

late_specnums <-
  antigens.df %>%
    dplyr::filter(sero_age %in% c("12 months","18 months","24 months") | (sero_age == "6 months" & cluster == "Infant high")) %>%
    dplyr::select(sero_specnum) %>%
    dplyr::pull()
  
late_igg_by_cluster.data <-
  igg.data[,colnames(igg.data) %in% late_specnums]

(all_individuals_late_by_cluster <-
  plot_tsne(t(late_igg_by_cluster.data),
            antigens.df[antigens.df$sero_specnum %in% late_specnums,],
            perplexity = 40, # was 40
            match_by="sero_specnum",
            colour_by="cluster",
            group_by="codenum",
            shape_by="sero_age",
            group_min=2,
            max_iter=50000) +
  scale_shape_manual(values = sero_age_shapes, name = "Age") +
  guides(colour = guide_legend(title.position = "top",
                               title = "Cluster"),
         shape = guide_legend(title.position = "top")) +
  scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster")
)

# all_individuals_late +
#   ggrepel::geom_label_repel(data=. %>% dplyr::filter(sero_age == "6 months"), aes(label = codenum))

```

```{r Plot t-SNE across individuals after later coloured by cluster, fig.height = 11, fig.width = 8, messages = FALSE,eval=FALSE}

just_late_specnums <-
  antigens.df %>%
    dplyr::filter(sero_age %in% c("12 months","18 months","24 months")) %>%
    dplyr::select(sero_specnum) %>%
    dplyr::pull()
  
just_late_igg_by_cluster.data <-
  igg.data[,colnames(igg.data) %in% just_late_specnums]

all_individuals_just_late_by_cluster <-
  plot_tsne(t(just_late_igg_by_cluster.data),
            antigens.df[antigens.df$sero_specnum %in% just_late_specnums,],
            perplexity = 40,
            match_by="sero_specnum",
            colour_by="cluster",
            group_by="codenum",
            shape_by="sero_age",
            group_min=2,
            max_iter=50000) +
  scale_shape_manual(values = sero_age_shapes, name = "Age") +
  guides(colour = guide_legend(title.position = "top",
                               title = "Cluster"),
         shape = guide_legend(title.position = "top")) +
  scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster")

save(all_individuals_just_late_by_cluster,
     file="Maela_late_timepoint_tSNE.Rdata")

```

```{r Display late timepoint tSNE plot, fig.height = 11, fig.width = 8}

load("Maela_late_timepoint_tSNE.Rdata")

all_individuals_just_late_by_cluster

# all_individuals_late +
#   ggrepel::geom_label_repel(data=. %>% dplyr::filter(sero_age == "6 months"), aes(label = codenum))

```

```{r Plot t-SNE across individuals without six months coloured by cluster, fig.height = 8, fig.width = 8, messages = FALSE}

no_6mo_specnums <-
  antigens.df %>%
    dplyr::filter(sero_age != "6 months") %>%
    dplyr::select(sero_specnum) %>%
    dplyr::pull()
  
no_6mo_by_cluster.data <-
  igg.data[,colnames(igg.data) %in% no_6mo_specnums]

(all_individuals_no_6mo_by_cluster <-
  plot_tsne(t(no_6mo_by_cluster.data),
            antigens.df[antigens.df$sero_specnum %in% no_6mo_specnums,] %>%
              dplyr::mutate(indiv_stage = dplyr::if_else(
                                sero_age %in% c("Birth","Cord"),
                                paste0(codenum,"_early"),
                                paste0(codenum,"_late")
                              )
                            ),
            perplexity = 40,
            match_by="sero_specnum",
            colour_by="cluster",
            group_by="indiv_stage",
            shape_by="sero_age",
            group_min=2,
            max_iter=5000) +
  scale_shape_manual(values = sero_age_shapes, name = "Age") +
  guides(colour = guide_legend(title.position = "top",
                               title = "Cluster"),
         shape = guide_legend(title.position = "top")) +
  scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster")
)

# all_individuals_late +
#   ggrepel::geom_label_repel(data=. %>% dplyr::filter(sero_age == "6 months"), aes(label = codenum))

```

# Individual antigen analysis

```{r Create individual analysis data frame, fig.width = 8, fig.height = 10}

sampling.df <-
  translation.df %>%
    # Add technical information
    dplyr::left_join(pheno.df %>%
                        dplyr::rename(sero_specnum = Sample.ID) %>%
                        dplyr::select(sero_specnum,Slide.ID),
                     by = c("sero_specnum")
                     ) %>%
    # Add epidemiological information
    dplyr::full_join(combined.info,
                     by = c("codenum","specdate","category","dob"),
                     multiple = "all"
    ) %>%
    # Order
    dplyr::arrange(codenum,specdate) %>%
    dplyr::mutate(serotype = dplyr::if_else(
                      serotype == "NA",
                      NA_character_,
                      serotype
                    )
                  )

# Plot a summary of epidemiological data here
ggplot(sampling.df %>%
         dplyr::filter(category == "Infant" & !is.na(specnum)),
       aes(x = specdate,
           y = codenum,
           colour = serotype,
           shape = diagnosis_clinical)) +
  geom_point() +
  ylab("Individual") +
  xlab("Sample collection date") +
  theme_bw() +
  theme(legend.position = "bottom",
        legend.direction='vertical') +
  guides(colour=guide_legend(ncol=6,title = "Serotype"),
         shape=guide_legend(title="Sample type"))

```

```{r Tabulate diagnosis at time of swabbing}

table(sampling.df %>%
        dplyr::filter(!is.na(specnum)) %>%
        dplyr::select(diagnosis_clinical,pnc)) %>%
  kableExtra::kable() %>%
  kableExtra::kable_styling(latex_options = "scale_down")

```

```{r Tabulate genome tallies}

table(sampling.df %>%
        dplyr::filter(!is.na(specnum)) %>%
        dplyr::select(GenomicData)) %>%
  kableExtra::kable() %>%
  kableExtra::kable_styling(latex_options = "scale_down")

```

```{r Combine individual data with genomic data}


no_genome_data <-
  sampling.df %>%
    dplyr::filter(GenomicData == "No genome data") %>%
    dplyr::select(specnum)  %>%
    dplyr::pull()

add_core_presence <- function(x,serotypes,lanes) {
    dplyr::if_else(serotypes!="character(0)" & (is.na(lanes) | lanes=="character(0)"),
                          as.integer(1),
                          as.integer(x))
}

antigen_names <- colnames(complete_antigen_presence_absence.df)[2:ncol(complete_antigen_presence_absence.df)]

genomic_immunology.df <-
  sampling.df %>%
    # Associate epidemiological samples with subsequent IgG sample
    dplyr::select(-c(age.d,age.v,pnc,serotype,sequence_type,GPSC,sangernum,GenomicData)) %>%
    dplyr::mutate(InferredSerotype = dplyr::if_else(InferredSerotype=="None",NA_character_,InferredSerotype)) %>%
    dplyr::mutate(InferredGPSC = dplyr::if_else(InferredGPSC=="None",NA_character_,InferredGPSC)) %>%
    dplyr::group_by(codenum,category) %>%
    tidyr::fill(c(sero_specnum,sero_age,previous_specdate,first_specdate,Slide.ID),.direction = "updown") %>%
    dplyr::ungroup() %>%
    # Record cases of disease
    dplyr::group_by(sero_specnum,category) %>%
    dplyr::mutate(LRTI_since_last_sample = sum(diagnosis_clinical %in% c("Pneumonia","Severe pneumonia","Very severe pneumonia"))) %>%
    dplyr::mutate(severe_LRTI_since_last_sample = sum(diagnosis_clinical %in% c("Severe pneumonia","Very severe pneumonia"))) %>%
    dplyr::mutate(very_severe_LRTI_since_last_sample = sum(diagnosis_clinical %in% c("Very severe pneumonia"))) %>%
    dplyr::ungroup() %>%
    dplyr::select(-c(diagnosis_clinical,Xray_diagnosis,specnum)) %>%
    # Summarise data relevant to each IgG analysis
    dplyr::group_by(sero_specnum,
                    codenum,
                    category,
                    sero_age,
                    LRTI_since_last_sample,
                    severe_LRTI_since_last_sample,
                    very_severe_LRTI_since_last_sample,
                    dob,
                    specdate,
                    previous_specdate,
                    first_specdate,
                    Slide.ID) %>%
    dplyr::reframe(serotypes = list(InferredSerotype),
                     GPSCs = list(InferredGPSC),
                     lanes = list(SingleIsolateLane,DeepSeqLane,RepresentativeGenome),
                     accessions = list(SingleIsolateAccession,DeepSeqAccession,RepresentativeAccession)) %>%
    dplyr::ungroup() %>%
    dplyr::group_by(sero_specnum,codenum,category,sero_age,LRTI_since_last_sample,severe_LRTI_since_last_sample,very_severe_LRTI_since_last_sample,dob,specdate,previous_specdate,first_specdate,Slide.ID,serotypes,GPSCs) %>%
    dplyr::summarise(lanes = list(lanes),
                     accessions = list(accessions),
                     .groups = "drop_last") %>%
    dplyr::ungroup() %>%
    # Process lists individually
    dplyr::rowwise() %>%
    dplyr::mutate(serotypes =
                    list(
                      unique(
                        unlist(
                          lapply(serotypes,function(x) x[!is.na(x)]))
                      )
                    )
                  ) %>%
    dplyr::mutate(GPSCs = 
                    list(
                      unique(
                        unlist(
                          lapply(GPSCs,function(x) x[!is.na(x)]))
                        )
                      )
                  ) %>%
    dplyr::mutate(lanes = 
                    list(
                      unique(
                        unlist(
                          lapply(lanes,function(x) x[!is.na(x)]))
                        )
                      )
                    ) %>%
    dplyr::mutate(accessions = 
                    list(
                      unique(
                        unlist(
                          lapply(accessions,function(x) x[!is.na(x)]))
                        )
                      )
                  ) %>%
    dplyr::ungroup() %>%
    # Record the first serotypes and GPSCs for each child
    dplyr::rowwise() %>%
    dplyr::mutate(empty_sero_list = dplyr::if_else(length(serotypes)==0,"Empty","Not_empty")) %>%
    dplyr::mutate(empty_gpsc_list = dplyr::if_else(length(GPSCs)==0,"Empty","Not_empty")) %>%
    dplyr::ungroup() %>%
    dplyr::group_by(codenum,category,empty_sero_list,
                     .groups = "drop_last") %>%
    dplyr::mutate(first_serotypes = list(dplyr::first(serotypes[!is_empty(unlist(serotypes))]))) %>%
    dplyr::ungroup() %>%
    dplyr::group_by(codenum,category,empty_gpsc_list) %>%
    dplyr::mutate(first_GPSCs = list(dplyr::first(GPSCs[!is_empty(unlist(GPSCs))]))) %>%
    dplyr::ungroup() %>%
    dplyr::rowwise() %>%
    dplyr::mutate(empty_sero_list = paste(first_serotypes,collapse = ", ")) %>%
    dplyr::mutate(empty_sero_list = dplyr::if_else(
      (empty_sero_list==""|category=="Mother"),
      NA_character_,
      empty_sero_list)) %>%
    dplyr::mutate(empty_gpsc_list = paste(first_GPSCs,collapse = ", ")) %>%
    dplyr::mutate(empty_gpsc_list = dplyr::if_else(
      (empty_gpsc_list==""|category=="Mother"),
      NA_character_,
      empty_gpsc_list)) %>%
    dplyr::ungroup() %>%
    dplyr::group_by(codenum) %>%
    tidyr::fill(empty_sero_list,.direction="updown") %>%
    tidyr::fill(empty_gpsc_list,.direction="updown") %>%
    dplyr::mutate(first_serotypes = str_split(empty_sero_list,", ",simplify = FALSE)) %>%
    dplyr::mutate(first_GPSCs = str_split(empty_gpsc_list,", ",simplify = FALSE)) %>%
    dplyr::ungroup() %>%
    dplyr::select(-c(empty_gpsc_list,empty_sero_list)) %>%
    # Link to genomic data
    tidyr::unnest(cols = lanes,
                  keep_empty = TRUE) %>%
    dplyr::left_join(complete_antigen_presence_absence.df,
                     by = c("lanes"="lane")) %>%
    # Convert to integers
    dplyr::mutate(
      across(
        all_of(antigen_names),
        ~ dplyr::if_else(is.na(.x),as.integer(0),as.integer(.x))
      )
    ) %>%
    # Add in core genes where only serotype available
    dplyr::mutate(
      across(
        all_of(core.list[core.list %in% colnames(complete_antigen_presence_absence.df)]),
        # ~ dplyr::if_else(serotypes!="character(0)" & (is.na(lanes) | lanes=="character(0)")),
        #                  as.integer(1),
        #                  as.integer(.x))
        ~ add_core_presence(.x,serotypes,lanes)
        # ~ dplyr::if_else(serotypes!="character(0)"),
        #                  as.integer(1),
        #                  as.integer(as.name(cur_column()))
      )
    ) %>%
    # Exposed if any coloniser carried protein
    dplyr::group_by(codenum,category,specdate) %>%
    dplyr::mutate(
      across(
        all_of(antigen_names),
        max
      )
    ) %>%
    tidyr::nest(lanes = lanes) %>%
    dplyr::mutate(lanes = purrr::flatten(lanes)) %>%
    dplyr::ungroup()

# Combined data with time of exposure and response
antigen_exposure_response.df <-
  genomic_immunology.df %>%
    # Pivot to one row per antigen per sample
    tidyr::pivot_longer(all_of(antigen_names),
                        names_to = "antigen",
                        values_to = "distribution") %>%
    # Record time of first exposure to antigen
    dplyr::group_by(codenum,category,antigen,distribution) %>%
    dplyr::mutate(first_exposure = dplyr::first(specdate)) %>%
    dplyr::ungroup() %>%
    dplyr::group_by(codenum,category,antigen) %>%
    dplyr::mutate(first_exposure = dplyr::if_else(distribution == 0,lubridate::as_date(NA),first_exposure)) %>%
    tidyr::fill(first_exposure,.direction = "updown") %>%
    dplyr::ungroup() %>%
    # Record exposure in first month
    dplyr::group_by(codenum,category,antigen) %>%
    dplyr::mutate(FirstMonth = dplyr::if_else((as.numeric(specdate-dob) < 40 & distribution == 1),
                                             "ExposedInFirstMonth",
                                             NA_character_)
                  ) %>%
    tidyr::fill(FirstMonth, .direction = "updown") %>%
    dplyr::mutate(FirstMonth = dplyr::if_else(is.na(FirstMonth),"NotExposedInFirstMonth",FirstMonth)) %>%
    dplyr::ungroup() %>%
    # Record exposure in first six months
    dplyr::group_by(codenum,category,antigen) %>%
    dplyr::mutate(FirstSixMonths = dplyr::if_else((as.numeric(specdate-dob) < 180 & distribution == 1),
                                             "ExposedInFirstSixMonths",
                                             NA_character_)
                  ) %>%
    tidyr::fill(FirstSixMonths, .direction = "updown") %>%
    dplyr::mutate(FirstSixMonths = dplyr::if_else(is.na(FirstSixMonths),"NotExposedInFirstSixMonths",FirstSixMonths)) %>%
    dplyr::ungroup() %>%
    # Record exposure since birth
    dplyr::group_by(sero_specnum,antigen) %>%
    dplyr::mutate(SinceBirth = dplyr::if_else(max(distribution) == 1,
                                              "ExposedSinceBirth",
                                              NA_character_)
    ) %>%
    tidyr::fill(SinceBirth, .direction = "updown") %>%
    dplyr::ungroup() %>%
    dplyr::group_by(codenum,category,antigen) %>%
    dplyr::arrange(specdate) %>%
    tidyr::fill(SinceBirth, .direction = "down") %>%
    dplyr::mutate(SinceBirth = dplyr::if_else(is.na(SinceBirth),"NotExposedSinceBirth",SinceBirth)) %>%
    dplyr::ungroup() %>%
    # Record exposure since last sample
    dplyr::group_by(sero_specnum,antigen) %>%
    dplyr::mutate(LastSample = dplyr::if_else(max(distribution) == 1,
                                              "ExposedSinceLastSample",
                                              "NotExposedSinceLastSample")) %>%
    dplyr::ungroup() %>%
    # Record whether exposure was within 40 days of blood sample being taken
    dplyr::group_by(sero_specnum,antigen) %>%
    dplyr::mutate(LastMonth = dplyr::if_else((as.numeric(max(specdate)-specdate) < 40 & distribution == 1),
                                             "ExposedInLastMonth",
                                             NA_character_)
                  ) %>%
    tidyr::fill(LastMonth, .direction = "updown") %>%
    dplyr::mutate(LastMonth = dplyr::if_else(is.na(LastMonth),"NotExposedInLastMonth",LastMonth)) %>%
    dplyr::ungroup() %>%
    # Record whether first exposure was in last month
    dplyr::group_by(sero_specnum) %>%
    dplyr::mutate(FirstInLastMonth = dplyr::if_else((as.numeric(max(specdate)-first_exposure) < 40 & as.numeric(max(specdate)-first_exposure) >= 0),
                                             "FirstExposedInLastMonth",
                                             "FirstExposedBeforeLastMonth")
                  ) %>%
    dplyr::mutate(FirstInLastMonth =  dplyr::if_else(is.na(FirstInLastMonth),"NotExposedSinceBirth",FirstInLastMonth)) %>%
    dplyr::ungroup() %>%
    # Number of exposures in each sampling period
    dplyr::group_by(codenum,category,specdate,antigen) %>%
    dplyr::mutate(ExposedInMonth = max(distribution)) %>%
    dplyr::ungroup() %>%
    dplyr::group_by(codenum,category,sero_specnum,antigen) %>%
    dplyr::mutate(ExposuresInMonth = sum(ExposedInMonth)) %>%
    dplyr::ungroup() %>%
    # Finally summarise all information
    dplyr::group_by(sero_specnum,
                      codenum,
                      category,
                      sero_age,
                      dob,
                      previous_specdate,
                      first_specdate,
                      Slide.ID,
                      first_serotypes,
                      first_GPSCs,
                      antigen,
                      first_exposure,
                      FirstMonth,
                      FirstSixMonths,
                      SinceBirth,
                      LastSample,
                      LastMonth,
                      FirstInLastMonth,
                      ExposuresInMonth,
                      LRTI_since_last_sample,
                      severe_LRTI_since_last_sample,
                      very_severe_LRTI_since_last_sample
                  ) %>%
    dplyr::summarise(specdate = max(specdate),
                     serotypesSinceLastSample = list(unique(unlist(serotypes))),
                     GPSCsSinceLastSample = list(unique(unlist(GPSCs))),
                     accessionsSinceLastSample = list(unique(unlist(accessions))),
                     lanesSinceLastSample = list(unique(unlist(lanes)))
                    ) %>%
    dplyr::ungroup() %>%
    # Calculate cumulative exposures since birth
    dplyr::group_by(codenum,category,antigen) %>%
    dplyr::arrange(specdate) %>%
    dplyr::mutate(ExposuresSinceBirth = cumsum(ExposuresInMonth)) %>%
    dplyr::ungroup() %>%
    # Calculate LRTIs between samples
    dplyr::mutate(AnyLRTISinceLastSample = dplyr::if_else(LRTI_since_last_sample==0,
                                                          "NoLRTISinceLastSample",
                                                          "LRTISinceLastSample")) %>%
    dplyr::mutate(AnySevereLRTISinceLastSample = dplyr::if_else(severe_LRTI_since_last_sample==0,
                                                          "NoSevereLRTISinceLastSample",
                                                          "SevereLRTISinceLastSample")) %>%
    dplyr::mutate(AnyVerySevereLRTISinceLastSample = dplyr::if_else(very_severe_LRTI_since_last_sample==0,
                                                          "NoVerySevereLRTISinceLastSample",
                                                          "VerySevereLRTISinceLastSample"))

n_antigens <- length(antigen_names)
n_igg_samples <- length(unique(translation.df$sero_specnum))

if (nrow(antigen_exposure_response.df) == n_antigens*n_igg_samples) {
  message("Correct number of datapoints")
} else {
  message("Incorrect number of data points")
}

# Run checks on outputs
early_stage_errors <-
  antigen_exposure_response.df %>%
    dplyr::filter(FirstSixMonths == "NotExposedInFirstSixMonths") %>%
    dplyr::filter(FirstMonth == "ExposedInFirstMonth")
if (nrow(early_stage_errors) > 0) {
  message("ERROR: Some antigens exposed in first month but not first six months")
} else {
  message("Assignments in first six months make sense logically")
}
first_month_count <- sum(antigen_exposure_response.df$FirstMonth=="ExposedInFirstMonth")
first_six_months_count <- sum(antigen_exposure_response.df$FirstSixMonths=="ExposedInFirstSixMonths")
if (first_six_months_count < first_month_count) {
  message("ERROR: More antigens exposed in first month than first six months")
} else {
  message("Assignments in first six months make sense numerically")
}

late_stage_errors <-
  antigen_exposure_response.df %>%
    dplyr::filter(LastSample == "NotExposedSinceLastSample") %>%
    dplyr::filter(LastMonth == "ExposedInLastMonth")
if (nrow(late_stage_errors) > 0) {
  message("ERROR: More antigens exposed in last month but not in last six months")
} else {
  message("Assignments in last six months make sense")
}
last_month_count <-
  sum(antigen_exposure_response.df$LastMonth=="ExposedInLastMonth")
since_last_sample_count <-
  sum(antigen_exposure_response.df$LastSample=="ExposedSinceLastSample")
first_in_last_count <-
  sum(antigen_exposure_response.df$FirstInLastMonth=="FirstExposedInLastMonth")
if (first_in_last_count < since_last_sample_count & last_month_count < since_last_sample_count) {
  message("Assignments in first pre-sample months make sense numerically")
} else {
  message("ERROR: Assignments in pre-sample months do not make numerical sense")
}

```

```{r Combine with IgG data}

antigen_exposure_response.df %<>%
  dplyr::left_join(antigens.df %>% dplyr::select(antigen,
                                                 sero_specnum,
                                                 igg,
                                                 type,
                                                 gene,
                                                 classification,
                                                 igg_delta,
                                                 cluster
                                                 ),
                   by=c("antigen","sero_specnum")
                   ) %>%
  # Remove slide for which we have no data
  dplyr::filter(!is.na(igg))

```

## Plot IgG levels

```{r Plot immune responses based on first six months, fig.height = 11, fig.width = 8}

plot_igg_levels <- function(df,var,rlevel = NULL,x_lab = NULL,
                            include_ages = c("6 months", "12 months", "18 months", "24 months"),
                            wrap_len = 25) {
  
  ref_level <- rlevel
  if (is.null(ref_level)) {
    ref_levels <- unique(df[[var]])[grep("Not",unique(df[[var]]))]
    if (length(ref_levels) > 0) {
      ref_level <- ref_levels[length(ref_levels)]
    } else {
      ref_level <- unique(df[[var]])[1]
    }
  }
  ref_level <- gsub("([[:lower:]])([[:upper:]]){1}", "\\1 \\2", ref_level)
  
  if (is.null(x_lab)) {
    x_lab = var
  }
  
  df %<>%
    dplyr::filter(sero_age %in% include_ages)
  
  if (is.factor(df[[var]])) {
    original_levels = gsub("([[:lower:]])([[:upper:]]){1}", "\\1 \\2", unique(df[[var]]))
    df[[var]] <- factor(gsub("([[:lower:]])([[:upper:]]){1}", "\\1 \\2", df[[var]]),
                        levels = original_levels)
  } else {
    df[[var]] <- gsub("([[:lower:]])([[:upper:]]){1}", "\\1 \\2", df[[var]])
  }
  
  col_var <- sym(var)
  test.df <-
    df %>%
      dplyr::select(sero_age,type,all_of(col_var),igg) %>%
      dplyr::group_by(sero_age,type,!!col_var) %>%
      dplyr::mutate(group_count = n()) %>%
      dplyr::ungroup() %>%
      dplyr::group_by(sero_age,type) %>%
      dplyr::mutate(level_count = n_distinct(!!col_var)) %>%
      dplyr::mutate(group_count = min(group_count)) %>%
      dplyr::mutate(ref_present = ref_level %in% !!col_var) %>%
      dplyr::filter(min(group_count) >= 3 & level_count > 1 & ref_present) %>%
      dplyr::ungroup() %>%
      dplyr::group_by(sero_age,type) %>%
      rstatix::wilcox_test(formula(paste("igg ~ ",var)),
                           ref.group = ref_level,
                           paired = FALSE,
                           p.adjust.method = "holm") %>%
      rstatix::adjust_pvalue() %>%
      rstatix::add_significance("p.adj") %>%
      dplyr::ungroup() %>%
      rstatix::add_xy_position(x = var, dodge = 1) %>%
      dplyr::filter(p.adj.signif != "ns")

  
  ggplot(df,
         aes(y = igg,
             x = !!col_var,
             colour = !!col_var,
             fill = !!col_var)) +
    geom_point(position = position_jitter(), alpha = 0.5, size = 0.25) +
    geom_violin(alpha = 0, colour = "black", draw_quantiles = c(0.5),
               scale = "width") +
    facet_grid(type~sero_age) +
    ggpubr::stat_pvalue_manual(test.df,
                         tip.length = 0,
                         label = "p.adj.signif",
                         inherit.aes = FALSE) +
    theme_bw() +
    xlab(x_lab) +
    scale_x_discrete(labels = function(x) str_wrap(x, width = wrap_len)) +
    ylab("IgG binding level") + 
    scale_y_continuous(expand = expansion(mult = c(0.05, 0.2))) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom",
          legend.title = element_blank(),
          strip.background = element_blank()) +
    guides(colour = guide_legend(override.aes = list(size=1, alpha = 1),
                                 title = x_lab),
           fill = guide_legend(override.aes = list(size=1, alpha = 1),
                                 title = x_lab))
  
}

plot_igg_levels(antigen_exposure_response.df,
                "FirstSixMonths",
                x_lab = "Protein exposure in first six months",
                wrap_len = 15)

```

```{r Plot based on first month, fig.height = 11, fig.width = 8}

plot_igg_levels(antigen_exposure_response.df,
                "FirstMonth",
                x_lab = "Protein exposure in first month",
                wrap_len = 15)

```

```{r Plot based on lifetime exposure, fig.height = 11, fig.width = 8}

plot_igg_levels(antigen_exposure_response.df,
                "SinceBirth",
                x_lab = "Protein exposure since birth",
                wrap_len = 15)

# PLOTTING FOR PRESENTATIONS
# (since_birth_plot <- plot_igg_levels(antigen_exposure_response.df,
#                                     "SinceBirth",
#                                     x_lab = "Protein exposure since birth",
#                                     wrap_len = 15) + theme(legend.position = "right",legend.key.size = unit(1.5, 'cm'),legend.text = element_text(size=15),axis.text.x = element_text(size=10))) + scale_y_continuous(expand = expansion(mult = c(0.05, 0.15)))

```

```{r Plot based on exposure in last month, fig.height = 11, fig.width = 8}

plot_igg_levels(antigen_exposure_response.df,
                "LastMonth",
                x_lab = "Protein exposure in the month prior to serum sampling",
                wrap_len = 15)

```

```{r Plot based on exposure since last test, fig.height = 11, fig.width = 8}

plot_igg_levels(antigen_exposure_response.df,
                "LastSample",
                x_lab = "Protein exposure since last serum sample",
                wrap_len = 15)

```

```{r Plot based on first exposure in last month, fig.height = 11, fig.width = 8}

plot_igg_levels(antigen_exposure_response.df,
                "FirstInLastMonth",
                x_lab = "First protein exposure occurred in the month prior to serum sampling",
                wrap_len = 15)

```

## Plot IgG changes

```{r Plot changing immune responses based on first six months, fig.height = 11, fig.width = 8}

plot_igg_change <- function(df,var,rlevel = NULL) {
  
  ref_level <- rlevel
  if (is.null(ref_level)) {
    ref_levels <- unique(df[[var]])[grep("Not",unique(df[[var]]))]
    if (length(ref_levels) > 0) {
      ref_level <- ref_levels[length(ref_levels)]
    } else {
      ref_level <- unique(df[[var]])[1]
    }
  }
  
  df %<>%
    dplyr::filter(!(sero_age %in% c("Cord","Birth")))
  
  col_var <- sym(var)
  test.df <-
    df %>%
      dplyr::group_by(sero_specnum) %>%
      dplyr::mutate(na_count = sum(is.na(igg_delta))) %>%
      dplyr::ungroup() %>%
      dplyr::filter(na_count < 1000) %>%
      dplyr::select(sero_age,type,all_of(col_var),igg_delta) %>%
      dplyr::group_by(sero_age,type,!!col_var) %>%
      dplyr::mutate(group_count = n()) %>%
      dplyr::ungroup() %>%
      dplyr::group_by(sero_age,type) %>%
      dplyr::mutate(level_count = n_distinct(!!col_var)) %>%
      dplyr::mutate(group_count = min(group_count)) %>%
      dplyr::mutate(ref_present = ref_level %in% !!col_var) %>%
      dplyr::filter(min(group_count) >= 3 & level_count > 1 & ref_present) %>%
      dplyr::ungroup() %>%
      dplyr::group_by(sero_age,type) %>%
      rstatix::wilcox_test(formula(paste("igg_delta ~ ",var)),
                           ref.group = ref_level,
                           paired = FALSE,
                           p.adjust.method = "holm") %>%
      rstatix::adjust_pvalue() %>%
      rstatix::add_significance("p.adj") %>%
      dplyr::ungroup() %>%
      rstatix::add_xy_position(x = var, dodge = 1) %>%
      dplyr::filter(p.adj.signif != "ns")
  
  ggplot(df,
         aes(y = igg_delta,
             x = !!col_var,
             colour = !!col_var,
             fill = !!col_var)) +
    geom_point(position = position_jitter(), alpha = 0.5, size = 0.25) +
    geom_violin(alpha = 0, colour = "black", draw_quantiles = c(0.5),
               scale = "width") +
    facet_grid(type~sero_age) +
    ggpubr::stat_pvalue_manual(test.df,
                         tip.length = 0,
                         label = "p.adj.signif",
                         inherit.aes = FALSE) +
    ylab(expression(Delta~"IgG level")) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom",
          strip.background = element_blank()) +
    guides(colour = guide_legend(override.aes = list(size=1, alpha = 1)))
  
}

plot_igg_change(antigen_exposure_response.df,"FirstSixMonths")

```

```{r Plot change based on first month, fig.height = 11, fig.width = 8}

plot_igg_change(antigen_exposure_response.df,"FirstMonth")

```

```{r Plot change based on lifetime exposure, fig.height = 11, fig.width = 8}

plot_igg_change(antigen_exposure_response.df,"SinceBirth")

```

```{r Plot change based on exposure in last month, fig.height = 11, fig.width = 8}

plot_igg_change(antigen_exposure_response.df,"LastMonth")

```

```{r Plot change based on exposure since last test, fig.height = 11, fig.width = 8}

plot_igg_change(antigen_exposure_response.df,"LastSample")

```

```{r Plot change based on first exposure in last month, fig.height = 11, fig.width = 8}

plot_igg_change(antigen_exposure_response.df,"FirstInLastMonth")

```

# Plot changes in IgG between timepoints

```{r Plot changes between timepoints by type, fig.width = 8, fig.height = 8}

ggplot(antigen_exposure_response.df %>%
         dplyr::filter(!(sero_age %in% c("Birth","Cord"))),
       aes(x = sero_age,
           y = igg_delta,
           colour = type,
           fill = type)) +
  geom_point(position = position_jitter(width = 0.1),
             size = 0.5,
             alpha = 0.25) +
  geom_violin(draw_quantiles = c(0.5), alpha = 0.2,
               scale = "width") +
  facet_wrap(.~type) +
  geom_hline(yintercept = 0, linetype = 2) +
  scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  ylab(expression(Delta*"IgG level")) +
  xlab("Age") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        strip.background = element_blank())

```

## Plot changes for intermediate-frequency antigens

```{r Plot changing immune responses to int freq proteins based on first six months, fig.height = 11, fig.width = 8}

variable_antigens.df <-
  tibble(
      "CLS00380" = "Eng",
      "CLS00346" = "AgaD",
      "CLS00336" = "HylA",
      "CLS00596" = "BgaA",
      "CLS00379" = "AliA",
      "CLS00926" = "PiaA",
      "CLS00389" = "CbpC/J",
      "CLS00362" = "Wzg",
      "CLS01822" = "SP_2105",
      "CLS00902" = "PhtB/D",
      "CLS01445" = "NanB",
      "CLS01852" = "PcpA",
      "CLS01541" = "Peptidase",
      "CLS00011" = "Membrane protein",
      "CLS02608" = "ZmpD",
      "CLS01333" = "PclA",
      "CLS03178" = "PclA",
      "CLS03616" = "PclA",
      "CLS99466" = "PclA",
      "CLS03265" = "PclA",
      "CLS02425" = "PhtA",
      "CLS02157" = "Glf",
      "CLS01887" = "Phage protein",
      "CLS01160" = "NanC",
      "CLS01991" = "ZmpC",
      "CLS02871" = "PitB",
      "ZmpE" = "ZmpE"
  ) %>%
  tidyr::pivot_longer(cols = everything(),
                      names_to = "antigen",
                      values_to = "gene")

plot_int_freq_change <- function(df,var,freq.df) {
  
  ref_levels <- unique(df[[var]])[grep("Not",unique(df[[var]]))]
  ref_level <- ref_levels[length(ref_levels)]
  
  int_freq_anti <-
    freq.df$antigen
  
  df %<>%
    dplyr::filter(!(sero_age %in% c("Cord","Birth"))) %>%
    dplyr::filter(antigen %in% int_freq_anti) %>%
    dplyr::filter(type != "non-ABT") %>%
    dplyr::mutate(antigen = dplyr::case_when(
                        is.na(gene) | gene == '-' ~ antigen,
                        TRUE ~ paste0(antigen,"\n(",gene,")")
                      )
                    )
  
  col_var <- sym(var)
  test.df <-
    df %>%
      dplyr::group_by(antigen) %>%
      dplyr::mutate(na_count = sum(is.na(igg_delta))) %>%
      dplyr::ungroup() %>%
      dplyr::filter(na_count < 1000) %>%
      dplyr::select(sero_age,antigen,all_of(col_var),igg_delta) %>%
      dplyr::group_by(sero_age,antigen,!!col_var) %>%
      dplyr::mutate(group_count = n()) %>%
      dplyr::ungroup() %>%
      dplyr::group_by(sero_age,antigen) %>%
      dplyr::mutate(level_count = n_distinct(!!col_var)) %>%
      dplyr::mutate(group_count = min(group_count)) %>%
      dplyr::mutate(ref_present = ref_level %in% !!col_var) %>%
      dplyr::filter(min(group_count) >= 3 & level_count > 1 & ref_present) %>%
      dplyr::ungroup() %>%
      dplyr::group_by(sero_age,antigen) %>%
      rstatix::wilcox_test(formula(paste("igg_delta ~ ",var)),
                           ref.group = ref_level,
                           paired = FALSE,
                           p.adjust.method = "holm") %>%
      rstatix::adjust_pvalue() %>%
      rstatix::add_significance("p.adj") %>%
      dplyr::ungroup() %>%
      rstatix::add_xy_position(x = var, dodge = 1) %>%
      dplyr::filter(p.adj.signif != "ns")

  
  ggplot(df,
         aes(y = igg_delta,
             x = !!col_var,
             colour = !!col_var,
             fill = !!col_var)) +
    geom_point(position = position_jitter(), alpha = 0.5, size = 0.25) +
    geom_violin(alpha = 0, colour = "black", draw_quantiles = c(0.5)) +
    facet_grid(antigen~sero_age) +
    ggpubr::stat_pvalue_manual(test.df,
                         tip.length = 0,
                          label = "p.adj.signif",
                         inherit.aes = FALSE) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom",
          strip.background = element_blank())
  
}

plot_int_freq_change(antigen_exposure_response.df,"FirstSixMonths",variable_antigens.df)

```

```{r Plot int freq change based on first month, fig.height = 11, fig.width = 8}

plot_int_freq_change(antigen_exposure_response.df,"FirstMonth",variable_antigens.df)

```

```{r Plot int freq change based on lifetime exposure, fig.height = 11, fig.width = 8}

plot_int_freq_change(antigen_exposure_response.df,"SinceBirth",variable_antigens.df)

```

```{r Plot int freq change based on exposure in last month, fig.height = 11, fig.width = 8}

plot_int_freq_change(antigen_exposure_response.df,"LastMonth",variable_antigens.df)

```

```{r Plot int freq change based on exposure since last test, fig.height = 11, fig.width = 8}

plot_int_freq_change(antigen_exposure_response.df,"LastSample",variable_antigens.df)

```

```{r Plot int freq change based on first exposure in last month, fig.height = 11, fig.width = 8}

plot_int_freq_change(antigen_exposure_response.df,"FirstInLastMonth",variable_antigens.df)

```

### Plot changes by number of exposures

```{r Plot responses by number of exposures}

int_freq_anti <-
    cog.frequency.df %>%
      dplyr::filter(maela_frequency > 0.05 & maela_frequency < 0.95) %>%
      dplyr::select(cog) %>%
      dplyr::pull()

response_by_num_exposures.df <-
  antigen_exposure_response.df %>%
    dplyr::filter(!(sero_age %in% c("Cord","Birth"))) %>%
    dplyr::filter(antigen %in% int_freq_anti) %>%
    dplyr::filter(type != "non-ABT") %>%
    dplyr::mutate(antigen = dplyr::case_when(
                        is.na(gene) | gene == '-' ~ antigen,
                        TRUE ~ paste0(antigen,"\n(",gene,")")
                      )
                    )

ggplot(response_by_num_exposures.df,
       aes(x = ExposuresSinceBirth,
           y = igg,
           colour = sero_age)
      ) +
  geom_point(position = position_jitter(width = 0.1), alpha= 0.5) +
  facet_grid(antigen~sero_age) +
  theme_bw() +
  theme(strip.background = element_blank())



ggplot(antigen_exposure_response.df %>% dplyr::filter(antigen == "PspA_10"),
       aes(x = ExposuresSinceBirth,
           y = igg,
           colour = sero_age)
      ) +
  geom_point(position = position_jitter(width = 0.1), alpha= 0.5) +
  facet_wrap(~sero_age) +
  theme_bw() +
  theme(strip.background = element_blank())

```

# Compare immune responses with first coloniser

```{r Tabulate relative to NT as first coloniser, fig.height = 11, fig.width = 8}

antigen_exposure_response.df %<>%
  dplyr::rowwise() %>%
  dplyr::mutate(NT_first = dplyr::if_else("NT" %in% first_serotypes,"NT","OnlyEncapsulated")) %>%
  dplyr::ungroup()

ref_level <- "OnlyEncapsulated"

test.df <-
    antigen_exposure_response.df %>%
      dplyr::select(sero_age,type,NT_first,igg) %>%
      dplyr::group_by(sero_age,type,NT_first) %>%
      dplyr::mutate(group_count = n()) %>%
      dplyr::ungroup() %>%
      dplyr::group_by(sero_age,type) %>%
      dplyr::mutate(level_count = n_distinct(NT_first)) %>%
      dplyr::mutate(group_count = min(group_count)) %>%
      dplyr::mutate(ref_present = ref_level %in% NT_first) %>%
      dplyr::filter(min(group_count) >= 3 & level_count > 1 & ref_present) %>%
      dplyr::ungroup() %>%
      dplyr::group_by(sero_age,type) %>%
      rstatix::wilcox_test(formula("igg ~ NT_first"),
                           ref.group = ref_level,
                           paired = FALSE,
                           p.adjust.method = "holm") %>%
      rstatix::adjust_pvalue() %>%
      rstatix::add_significance("p.adj") %>%
      dplyr::ungroup() %>%
      rstatix::add_xy_position(x = "NT_first", dodge = 1) %>%
      dplyr::filter(p.adj.signif != "ns")

ggplot(antigen_exposure_response.df,
         aes(y = igg,
             x = NT_first,
             colour = NT_first,
             fill = NT_first)) +
    geom_point(position = position_jitter(), alpha = 0.5, size = 0.25) +
    geom_violin(alpha = 0, colour = "black", draw_quantiles = c(0.5)) +
    facet_grid(type~sero_age) +
    ggpubr::stat_pvalue_manual(test.df,
                         tip.length = 0,
                         label = "p.adj.signif",
                         inherit.aes = FALSE) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom",
          strip.background = element_blank())

```

# Compare immune profiles with disease

```{r Plot differences between clusters, fig.height = 11, fig.width = 8}

# antigen_exposure_response.df %<>%
#   dplyr::left_join(
#       clustering_plot.df %>%
#         dplyr::select(sero_specnum,cluster),
#       by = c("sero_specnum")
#     ) %>%
#     dplyr::mutate(cluster = as.character(cluster))

plot_igg_levels(antigen_exposure_response.df,
                "cluster",
                rlevel = "Infant low",
                x_lab = "Immune cluster",
                include_ages = c("Cord", "Birth", "6 months", "12 months", 
"18 months", "24 months"),
                wrap_len = 25) +
  scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster")

```

```{r Tabulate relative to NTs}

clustering_info.df <-
  antigen_exposure_response.df %>%
    dplyr::select(sero_specnum,first_serotypes,first_GPSCs,sero_age,NT_first,cluster) %>%
    dplyr::distinct()

ggplot(clustering_info.df,
       aes(x = cluster,
           fill = NT_first,
           colour = NT_first)) +
  geom_bar() +
  facet_grid(NT_first~sero_age) +
  theme_bw() +
  theme(strip.background = element_blank())

```

```{r Tabulate relative to LRTI}

clustering_info.df %<>%
  dplyr::left_join(genomic_immunology.df %>%
                    dplyr::select(sero_specnum,LRTI_since_last_sample) %>%
                    dplyr::distinct()
                  )

ggplot(clustering_info.df,
       aes(x = cluster,
           y = LRTI_since_last_sample)) +
  geom_boxplot() +
  facet_grid(.~sero_age) +
  theme_bw() +
  theme(strip.background = element_blank())

```

```{r Plot IgG levels by LRTI status, fig.height = 11, fig.width = 8}

# antigen_exposure_response.df %<>%
#   dplyr::left_join(genomic_immunology.df %>%
#                     dplyr::select(sero_specnum,LRTI_since_last_sample) %>%
#                     dplyr::distinct(),
#                    by = c("sero_specnum")
#                   ) %>%
#   dplyr::mutate(AnyLRTISinceLastSample = dplyr::if_else(LRTI_since_last_sample==0,"NoLRTISinceLastSample","LRTISinceLastSample"))

plot_igg_levels(antigen_exposure_response.df %>%
                  dplyr::mutate(AnyLRTISinceLastSample = gsub("LRTI","Pneumonia",AnyLRTISinceLastSample)) %>%
                  dplyr::mutate(AnyLRTISinceLastSample = factor(AnyLRTISinceLastSample,levels = c("PneumoniaSinceLastSample","NoPneumoniaSinceLastSample"))),
                "AnyLRTISinceLastSample",
                x_lab = "Pneumonia diagnosis since last serum sample",
                wrap_len = 15)

# Plot for presentation
# (pneumonia_plot <- plot_igg_levels(antigen_exposure_response.df %>%
#                                         dplyr::filter(sero_age %in% c("6 months","12 months","18 months","24 months")) %>%
#                                         dplyr::mutate(AnyLRTISinceLastSample = gsub("LRTI","Pneumonia",AnyLRTISinceLastSample)) %>%
#                                         dplyr::mutate(AnyLRTISinceLastSample = factor(AnyLRTISinceLastSample,levels = c("PneumoniaSinceLastSample","NoPneumoniaSinceLastSample"))),
#                                     "AnyLRTISinceLastSample",
#                                     x_lab = "Pneumonia diagnosis since last serum sample",rlevel = "PneumoniaSinceLastSample",
#                                     wrap_len = 15) + theme(legend.position = "right",legend.key.size = unit(1.5, 'cm'),legend.text = element_text(size=15),axis.text.x = element_text(size=10))) + scale_y_continuous(expand = expansion(mult = c(0.05, 0.15)))


```

```{r Plot IgG change by LRTI status based on exposure, fig.height = 11, fig.width = 8}

plot_igg_change(antigen_exposure_response.df %>%
                  dplyr::mutate(AnyLRTISinceLastSample = gsub("LRTI","Pneumonia",AnyLRTISinceLastSample)),
                "AnyLRTISinceLastSample")

```

```{r Plot IgG levels by severe LRTI status, fig.height = 11, fig.width = 8}

plot_igg_levels(antigen_exposure_response.df %>%
                  dplyr::mutate(AnySevereLRTISinceLastSample = gsub("LRTI","Pneumonia",AnySevereLRTISinceLastSample)) %>%
                  dplyr::mutate(AnyLRTISinceLastSample = factor(AnyLRTISinceLastSample,levels = c("SeverePneumoniaSinceLastSample","NoSeverePneumoniaSinceLastSample"))),
                "AnySevereLRTISinceLastSample",
                x_lab = "Severe pneumonia diagnosis since last serum sample",
                wrap_len = 15)

```

```{r Plot IgG levels by very severe LRTI status, fig.height = 11, fig.width = 8}

plot_igg_levels(antigen_exposure_response.df %>%
  dplyr::mutate(AnyVerySevereLRTISinceLastSample = gsub("LRTI","Pneumonia",AnyVerySevereLRTISinceLastSample)) %>%
                  dplyr::mutate(AnyLRTISinceLastSample = factor(AnyLRTISinceLastSample,levels = c("VerySeverePneumoniaSinceLastSample","NoVerySeverePneumoniaSinceLastSample"))),
                "AnyVerySevereLRTISinceLastSample",
                x_lab = "Severe pneumonia diagnosis since last serum sample",
                wrap_len = 15)

```

### Correlation in variable antigen responses

```{r Plot PspA correlations, fig.height = 10, fig.width = 8}

get_age_specific_cor_df <- function(age,antigen_df = NA,antigen_type = NA) {
  
  var.df <-
    antigen_df %>%
      dplyr::filter(type == antigen_type) %>%
      dplyr::filter(sero_age == age) %>%
      dplyr::select(codenum,antigen,igg) %>%
      tidyr::pivot_wider(names_from = antigen, values_from = igg)
  
  var_cor <-
    var.df %>%
      corrr::correlate(
          use = "pairwise.complete.obs",
          method = "pearson"
      ) %>%
      corrr::rearrange() %>%
      tidyr::pivot_longer(cols = -term,
                          names_to = "antigen",
                          values_to = "correlation") %>%
      dplyr::mutate(sero_age = age)
  
  return(var_cor)
}
  
get_cor_df <- function(antigen_df,antigen_type) {
  
  ages <- c("Cord","Birth","6 months","12 months","18 months","24 months")
  
  var_antigen_cor <-
    dplyr::bind_rows(
      lapply(ages,
             get_age_specific_cor_df,
             antigen_df = antigen_df,
             antigen_type = antigen_type)  
    ) %>%
    dplyr::mutate(sero_age = factor(sero_age, levels = ages))
  
  var_antigen_dist <-
    dist(
      var_antigen_cor %>%
        dplyr::filter(sero_age == "24 months") %>%
        dplyr::select(term,antigen,correlation) %>%
        tidyr::pivot_wider(id_cols = term,
                           names_from = antigen,
                           values_from = correlation) %>%
        tibble::column_to_rownames(var = "term")
    )
  permute_method <- "OLO"
  var_antigen_ser_permutation_x <- seriation::seriate(var_antigen_dist,
                                                    method = permute_method)
  var_antigen_dist_reordered <- seriation::permute(var_antigen_dist,
                                            var_antigen_ser_permutation_x)
  
  var_antigen_cor %<>%
    dplyr::mutate(term = factor(term,
                                levels = labels(var_antigen_dist_reordered))) %>%
    dplyr::mutate(antigen = factor(antigen,
                                levels = labels(var_antigen_dist_reordered)))
  
  
  return(var_antigen_cor)

}

plot_variable_antigen_cor <- function(df) {
  ggplot(df,
       aes(
          x = term,
          y = antigen,
          size = correlation,
          colour = correlation
        )
       ) +
  geom_point(shape = 16, alpha = 0.75) +
  scale_size_area(breaks = NULL,limits = c(-1,1)) +
  scale_colour_gradient2(low = "indianred2",
                        mid = "white",
                        high = "skyblue1",
                        limits = c(-1,1)) +
  scale_size(range = c(-4,4),limits=c(-1,1)) +
  guides(size = "none") +
  ylab("Protein") +
  xlab("Protein") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "bottom",
        strip.background = element_blank()) +
  facet_wrap(.~sero_age,
             nrow = 3,
             ncol = 2)
}

pspA_cor <- get_cor_df(antigen_exposure_response.df,"PspA")
plot_variable_antigen_cor(pspA_cor)

# For presentation
# (pspA_cor_plot <- plot_variable_antigen_cor(pspA_cor)  + theme(legend.position = "right",legend.key.size = unit(1.5, 'cm'),legend.title = element_text(size=15),legend.text = element_text(size=15),axis.text = element_text(size=10)))

```

```{r Plot PspA levels, fig.height = 8, fig.width = 8}

plot_var_immunity_level <- function(df,antigen_type,cor_df) {
  
  ggplot(df %>%
         dplyr::filter(type == antigen_type) %>%
         dplyr::mutate(antigen = factor(antigen,
                                        levels = levels(cor_df$antigen))),
       aes(x = antigen,
           y = codenum,
           fill = igg)
       ) +
  geom_tile() +
  theme_bw() +
  facet_wrap(.~sero_age,
             nrow = 3,
             ncol = 2) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        strip.background = element_blank())
  
}

plot_var_immunity_level(antigen_exposure_response.df,"PspA",pspA_cor)

```

```{r Plot PspA frequencies against responses}

pspA_comparison.df <-
  pspA.freq.df %>%
    dplyr::left_join(
      antigen_exposure_response.df %>% dplyr::select(antigen,
                                                     codenum,
                                                     sero_age,
                                                     igg),
      by = c("allele"="antigen")
    ) %>%
    dplyr::mutate(allele = factor(allele,levels = levels(pspA_cor$term)))

ggplot(pspA_comparison.df,
       aes(x = allele,
           y = `Single isolate`,
           group = allele)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(alpha = 0.1, size = 0.25) +
  theme_bw()

```

```{r Plot PspC correlations, fig.height = 22, fig.width = 16}

pspC_cor <- get_cor_df(antigen_exposure_response.df,"PspC")
plot_variable_antigen_cor(pspC_cor) +
  theme(axis.text.x = element_text(size = 6))

```

```{r Plot PspC levels, fig.height = 8, fig.width = 8}

plot_var_immunity_level(antigen_exposure_response.df,"PspC",pspC_cor)

```

```{r Plot PspC frequencies against responses}

pspC_comparison.df <-
  pspC.freq.df %>%
    dplyr::left_join(
      antigen_exposure_response.df %>% dplyr::select(antigen,
                                                     codenum,
                                                     sero_age,
                                                     igg),
      by = c("allele"="antigen")
    ) %>%
    dplyr::mutate(allele = factor(allele,levels = levels(pspC_cor$term)))

ggplot(pspC_comparison.df,
       aes(x = allele,
           y = `Single isolate`,
           group = allele)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(alpha = 0.1, size = 0.25) +
  theme_bw()

```

```{r Plot ZmpA correlations, fig.height = 10, fig.width = 8}

zmpA_cor <- get_cor_df(antigen_exposure_response.df,"ZmpA")
plot_variable_antigen_cor(zmpA_cor)

```

```{r Plot ZmpA levels, fig.height = 8, fig.width = 8}

plot_var_immunity_level(antigen_exposure_response.df,"ZmpA",zmpA_cor)

```

```{r Plot ZmpB correlations, fig.height = 10, fig.width = 8}

zmpB_cor <- get_cor_df(antigen_exposure_response.df,"ZmpB")
plot_variable_antigen_cor(zmpB_cor)

```

```{r Plot ZmpB levels, fig.height = 8, fig.width = 8}

plot_var_immunity_level(antigen_exposure_response.df,"ZmpB",zmpB_cor)

```

### Mixed effects modelling of the level of IgG binding

```{r IgG binding mixed effects model}

library(lme4)
library(sjPlot)

mef_input.df <-
  antigen_exposure_response.df %>%
    dplyr::select(codenum,sero_age,antigen,type,igg) %>%
    dplyr::mutate(type = factor(type,
                                levels = c("non-ABT",
                                           "Adhesin",
                                           "Other ABT",
                                           "CellWall",
                                           "Degradative",
                                           "SBP",
                                           "PspA",
                                           "PspC",
                                           "ZmpA",
                                           "ZmpB")
                                )
                  )

igg.model_1 <-
  lmer(igg ~ sero_age*type + (1|codenum) + 0,
       data = mef_input.df,
       REML = FALSE)

igg.model_2 <-
  lmer(igg ~ sero_age + type + (1|codenum) + 0,
       data = mef_input.df,
       REML = FALSE)

igg.model_3 <-
  lmer(igg ~ sero_age + (1|type) + (1|codenum) + 0,
       data = mef_input.df,
       REML = FALSE)

# save(igg.model_1,igg.model_2,igg.model_3,
#      file="Maela_antigens_igg_model_fits.Rdata")

```

```{r Model comparison}

#load("Maela_antigens_igg_model_fits.Rdata")

# Test model fits
anova(igg.model_1,igg.model_2)
anova(igg.model_1,igg.model_3)

```

```{r Plot coefficients, fig.height = 11, fig.width = 8}

sjPlot::plot_model(igg.model_1)

```

```{r Tabulate model output}

sjPlot::tab_model(igg.model_1)

```

# move to 1572

```{r Plot trends over time, fig.height = 8, fig.width = 8}

igg_prediction.df <-
  expand.grid("sero_age" = base::structure(c(1L,2L,3L,4L,5L,6L), levels = c("Cord", "Birth", "6 months", "12 months", "18 months", "24 months"), class = "factor"),
              "type" = base::structure(c(2L, 1L, 3L, 4L, 5L, 6L, 7L, 8L, 9L, 10L), levels = c("non-ABT", 
"Adhesin", "Other ABT", "CellWall", "Degradative", "SBP", "PspA", 
"PspC", "ZmpA", "ZmpB"), class = "factor")
              )

new_pred <- predict(igg.model_1, newdata = igg_prediction.df, re.form = NA)

igg_prediction.df %<>%
  dplyr::mutate(prediction = new_pred)

ggplot(igg_prediction.df,
       aes(x = sero_age,
           y = prediction,
           group = type,
           colour = type)) +
  geom_line() +
  geom_point() +
  scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  theme_bw() +
  xlab("Age") +
  ylab("Predicted IgG binding") +
  theme(legend.position = "bottom")

```

```{r Compare with raw data, fig.height = 8, fig.width = 8}

(model_comp_raw_data_plot <-
  ggplot(antigen_exposure_response.df,
         aes(x = sero_age,
             y = igg,
             group = interaction(antigen,codenum),
             colour = type)) +
    geom_line(alpha = 0.05, linetype = 1) +
    geom_line(data = igg_prediction.df,
              aes(x = sero_age,
                  y = prediction,
                  group = type),
              size = 1,
              inherit.aes = FALSE,
              colour = "white") +
    facet_wrap(.~type,nrow = 2) +
    theme_bw() +
    scale_colour_manual(values = type_palette,
                        aesthetics = c("fill","colour"),
                        drop = TRUE, limits = force,
                        name = "Protein type") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom",
          strip.background = element_blank()) +
    xlab("Age") +
    ylab("IgG binding") +
    guides(colour = guide_legend(override.aes = list(alpha = 1)))
)

```

```{r Fit logit model to exposed or not exposed}

fit_saturation_model <- function(l.df) {
  a_start <- 0.5
  lmod <- nls(igg ~ B - C*exp(-A*ExposuresSinceBirth), 
                        l.df,
                  start = list("A" = a_start, "B" = median(l.df$igg), "C" = median(l.df$igg)/exp(-1*a_start)),
                  lower = list("A" = 0, "B" = min(l.df$igg), "C" = 0),
                  upper = list("A" = 25, "B" = max(l.df$igg), "C" = 25), #max(l.df$igg)+1),
                  algorithm = "port",
                  control = nls.control(maxiter = 10000))
  return(lmod)
}

fit_response_models <- function(protein,antigen_df = NA) {
 #protein <- "CLS02871"
  #protein<-"PsrP"

  #message(paste("Analyse",protein))
    
  logit.df <-
    antigen_df %>%
      dplyr::select(antigen,type,SinceBirth,ExposuresSinceBirth,codenum,sero_age,igg) %>%
      dplyr::filter(sero_age %in% c("12 months","18 months","24 months")) %>%
      dplyr::filter(antigen == protein)

  lmod<-NULL
  try_result <- tryCatch({lmod<-fit_saturation_model(logit.df)},
                          error = function(e){}
                        )
  
  results_list <- list("antigen" = protein)
  if (class(try_result)[1] != "simpleError" & !is.null(lmod)) {
    results_list <- list("antigen" = protein, "A" = coef(lmod)[[1]], "B" = coef(lmod)[[2]], "C" = coef(lmod)[[3]])
  }
  
  return(as.data.frame(results_list))
}

res_out <- 
  dplyr::bind_rows(
    lapply(antigen_exposure_response.df %>%
             dplyr::filter(type != "non-ABT") %>%
             dplyr::select(antigen) %>%
             dplyr::distinct() %>%
             dplyr::pull(),
                  fit_response_models,
                  antigen_df = antigen_exposure_response.df)
  )

candidate_vaccine_antigens <- c("rrgB1","rrgB2","rrgB3","ply","piaA","piuA")

exposure_summary.df <-
  antigen_exposure_response.df %>%
    dplyr::select(antigen,type,gene) %>%
    dplyr::filter(type != "non-ABT" | gene %in% candidate_vaccine_antigens) %>%
    dplyr::distinct() %>%
    dplyr::left_join(res_out, by = c("antigen"))

ggplot(exposure_summary.df,
       aes(x = type,
           y = A)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(position = position_jitter(), alpha = 0.25) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

```

```{r Plot factor C}

ggplot(exposure_summary.df,
       aes(x = type,
           y = C)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(position = position_jitter(), alpha = 0.25) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

```

```{r Plot examples of IgG responses increasing with exposures, fig.height = 11, fig.width = 8}

n_exp = 0:24

exposure_summary_plot.df <-
  exposure_summary.df %>%
    dplyr::filter((A >= 1 | C >= 2.5) & (A < 6 & C < 6)) %>%
    dplyr::left_join(antigen_exposure_response.df %>%
                       dplyr::filter(sero_age %in% c("12 months","18 months","24 months")) %>%
                       dplyr::select(antigen,codenum,ExposuresSinceBirth,igg),
                     by = c("antigen")) %>%
    dplyr::mutate(name = paste0(gene,"\n(",antigen,")"))

exposure_summary_models_plot.df <-
  exposure_summary_plot.df %>%
    dplyr::select(antigen,name,A,B,C) %>%
    dplyr::rowwise() %>%
    dplyr::mutate(n_exp = list(n_exp),
                  pred_val = list(B - C*exp(-1*A*n_exp))) %>%
    dplyr::ungroup() %>%
    tidyr::unnest(cols = c(n_exp,pred_val))

ggplot(exposure_summary_plot.df,
         aes(x = ExposuresSinceBirth,
             y = igg)) +
    geom_point(position = position_jitter(width = 0.05), alpha = 0.25, size = 0.25) +
    geom_line(data = exposure_summary_models_plot.df,
              aes(x = n_exp,
                  y = pred_val),
              inherit.aes = FALSE,
              colour = "red") +
    theme_bw() +
    facet_wrap(.~name) +
    theme(strip.background = element_blank())

```

```{r Seroconversion for those with large exposure values, fig.height = 11, fig.width = 8}

n_exp = 0:24

example_exposure_summary_plot.df <-
  exposure_summary.df %>%
    dplyr::filter(antigen %in% c("CLS00682","CLS00379","CLS01803","CLS01160","CLS01023","CLS02871","CLS01289","CLS99466","ZmpE","PspA_14","PspA_17","PspA_33")) %>%
    dplyr::left_join(antigen_exposure_response.df %>%
                       dplyr::filter(sero_age %in% c("12 months","18 months","24 months")) %>%
                       dplyr::select(antigen,codenum,ExposuresSinceBirth,igg),
                     by = c("antigen")) %>%
    dplyr::mutate(name = paste0(gene,"\n(",antigen,")"))

example_exposure_summary_models_plot.df <-
  example_exposure_summary_plot.df %>%
    dplyr::select(antigen,name,A,B,C) %>%
    dplyr::rowwise() %>%
    dplyr::mutate(n_exp = list(n_exp),
                  pred_val = list(B - C*exp(-1*A*n_exp))) %>%
    dplyr::ungroup() %>%
    tidyr::unnest(cols = c(n_exp,pred_val))

ggplot(example_exposure_summary_plot.df,
         aes(x = ExposuresSinceBirth,
             y = igg)) +
    geom_point(position = position_jitter(width = 0.05), alpha = 0.25, size = 0.25) +
    geom_line(data = example_exposure_summary_models_plot.df,
              aes(x = n_exp,
                  y = pred_val),
              inherit.aes = FALSE,
              colour = "red") +
    ylim(c(-1,8)) +
    theme_bw() +
    facet_wrap(.~name) +
    theme(strip.background = element_blank())

```

### Fit mixed models to understand clustering

```{r Mixed effects models to understand clustering - data prep}

library(lme4)

mef_input_with_clustering.df <-
  antigen_exposure_response.df %>%
    dplyr::select(codenum,cluster,sero_age,antigen,type,igg) %>%
    dplyr::mutate(type = factor(type,
                                levels = c("non-ABT",
                                           "Adhesin",
                                           "Other ABT",
                                           "CellWall",
                                           "Degradative",
                                           "SBP",
                                           "PspA",
                                           "PspC",
                                           "ZmpA",
                                           "ZmpB")
                                )
                  )

```

```{r Mixed effects models to understand clustering - fit models}

igg.no_clustering_model_1 <-
  lmer(igg ~ sero_age + type + (1|codenum) + 0,
       data = mef_input_with_clustering.df,
       REML = FALSE)

igg.fe_clustering_model_1 <-
  lmer(igg ~ sero_age + type + cluster + (1|codenum) + 0,
       data = mef_input_with_clustering.df,
       REML = FALSE)

igg.re_clustering_model_1 <-
  lmer(igg ~ sero_age + type + (1|cluster) + (1|codenum) + 0,
       data = mef_input_with_clustering.df,
       REML = FALSE)

igg.int_clustering_model_1 <-
  lmer(igg ~ sero_age + type + cluster*type + (1|codenum) + 0,
       data = mef_input_with_clustering.df,
       REML = FALSE)


igg.no_clustering_model_2 <-
  lmer(igg ~ sero_age + (1|type) + (1|codenum) + 0,
       data = mef_input_with_clustering.df,
       REML = FALSE)

igg.fe_clustering_model_2 <-
  lmer(igg ~ sero_age + (1|type) + cluster + (1|codenum) + 0,
       data = mef_input_with_clustering.df,
       REML = FALSE)

igg.re_clustering_model_2 <-
  lmer(igg ~ sero_age + (1|type) + (1|cluster) + (1|codenum) + 0,
       data = mef_input_with_clustering.df,
       REML = FALSE)

igg.int_clustering_model_2 <-
  lmer(igg ~ sero_age + (1|type) + cluster*type + (1|codenum) + 0,
       data = mef_input_with_clustering.df,
       REML = FALSE)


igg.no_clustering_model_3 <-
  lmer(igg ~ sero_age*type + (1|codenum) + 0,
       data = mef_input_with_clustering.df,
       REML = FALSE)

igg.fe_clustering_model_3 <-
  lmer(igg ~ sero_age*type + cluster + (1|codenum) + 0,
       data = mef_input_with_clustering.df,
       REML = FALSE)

igg.re_clustering_model_3 <-
  lmer(igg ~ sero_age*type + (1|cluster) + (1|codenum) + 0,
       data = mef_input_with_clustering.df,
       REML = FALSE)

igg.int_clustering_model_3 <-
  lmer(igg ~ sero_age*type + cluster*type + (1|codenum) + 0,
       data = mef_input_with_clustering.df,
       REML = FALSE)

igg.int_clustering_model_4 <-
  lmer(igg ~ sero_age*type*cluster + (1|codenum) + 0,
       data = mef_input_with_clustering.df,
       REML = FALSE)

#save(igg.no_clustering_model_1,igg.fe_clustering_model_1,igg.re_clustering_model_1,igg.int_clustering_model_1,igg.no_clustering_model_2,igg.fe_clustering_model_2,igg.re_clustering_model_2,igg.int_clustering_model_2,igg.no_clustering_model_3,igg.fe_clustering_model_3,igg.re_clustering_model_3,igg.int_clustering_model_3,igg.int_clustering_model_4,
#     file="Maela_antigens_igg_model_fits_with_clustering.Rdata")

```

```{r Model comparison with clustering}

#load("Maela_antigens_igg_model_fits_with_clustering.Rdata")

clustering_model_list <- c(igg.no_clustering_model_1,igg.fe_clustering_model_1,igg.re_clustering_model_1,igg.int_clustering_model_1,igg.no_clustering_model_2,igg.fe_clustering_model_2,igg.re_clustering_model_2,igg.int_clustering_model_2,igg.no_clustering_model_3,igg.fe_clustering_model_3,igg.re_clustering_model_3,igg.int_clustering_model_3,igg.int_clustering_model_4)

clustering_model_comparison.df <-
  pairwise_model_comparisons(clustering_model_list)

ggplot(clustering_model_comparison.df,
       aes(x = i,
           y = j,
           colour = delta_aic,
           fill = delta_aic)) +
  geom_tile() +
  scale_colour_distiller(type="div",aesthetics = c("colour","fill")) +
  xlab("Reference model index") +
  ylab("Comparison model index") +
  theme_bw() +
  guides(colour = guide_colourbar(title = expression("AIC"[ref]-"AIC"[comp])),
         fill = guide_colourbar(title = expression("AIC"[ref]-"AIC"[comp])))

```

```{r Tabulate model comparisons}

clustering_model_comparison.df %>%
  kableExtra::kable() %>%
  kableExtra::kable_styling(latex_options = "scale_down")

```

```{r Plot coefficients with clustering}

sjPlot::tab_model(igg.int_clustering_model_4)

```

```{r Plot prediction againt observation for responses by cluster, fig.height = 8, fig.width = 8}

igg.int_clustering_model_4_predictions <- predict(igg.int_clustering_model_4)

inheritance_and_exposure_with_predictions.df <-
  mef_input_with_clustering.df %>%
    dplyr::mutate(Prediction = igg.int_clustering_model_4_predictions)

ggplot(inheritance_and_exposure_with_predictions.df,
       aes(x = igg,
           y = Prediction,
           colour = type)) +
  geom_point(alpha = 0.25,
             size = 0.25) +
  geom_smooth(method = "lm") +
  xlab("Observed IgG level") +
  ylab("Predicted IgG level") +
  scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  stat_cor(method = "pearson",
             p.accuracy = 0.000001,
             label.sep='\n'
  ) +
  facet_grid(sero_age~type) +
  theme_bw()

```

```{r Plot simple prediction againt observation for responses by cluster, fig.height = 8, fig.width = 8}

ggplot(inheritance_and_exposure_with_predictions.df,
       aes(x = igg,
           y = Prediction)) +
  geom_point(alpha = 0.25,
             size = 0.25) +
  geom_smooth(method = "lm") +
  xlab("Observed IgG level") +
  ylab("Predicted IgG level") +
  stat_cor(method = "pearson",
             p.accuracy = 0.000001,
             label.sep='\n'
  ) +
  theme_bw()

```

```{r Tabulate model output with clustering, fig.height = 8, fig.width = 8}

igg_prediction_with_clustering.df <-
  expand.grid("sero_age" = base::structure(1:6, levels = c("Cord", "Birth", "6 months", "12 months", "18 months", "24 months"), class = "factor"),
              "type" = base::structure(c(2L, 1L, 3L, 4L, 5L, 6L, 7L, 8L, 9L, 10L), levels = c("non-ABT", 
"Adhesin", "Other ABT", "CellWall", "Degradative", "SBP", "PspA", 
"PspC", "ZmpA", "ZmpB"), class = "factor"),
              "cluster" = c("Neonatal highest", "Neonatal infant-like", "Neonatal intermediate", "Neonatal lowest", 
"Infant high", "Infant low")
              ) %>%
  dplyr::filter((grepl("Neonatal",cluster) & sero_age %in% c("Cord","Birth")) | (grepl("Infant",cluster) & !(sero_age %in% c("Cord","Birth"))))

new_pred_with_clustering <- predict(igg.int_clustering_model_4,
                                    newdata = igg_prediction_with_clustering.df,
                                    re.form = NA)

igg_prediction_with_clustering.df %<>%
  dplyr::mutate(prediction = new_pred_with_clustering)

ggplot(igg_prediction_with_clustering.df,
       aes(x = sero_age,
           y = prediction,
           group = interaction(type,cluster),
           colour = type)) +
  geom_line() +
  geom_point() +
  theme_bw() +
  facet_wrap(.~cluster) +
  scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "bottom",
        strip.background = element_blank()) +
  xlab("Age") +
  ylab("Predicted IgG level")

```

```{r Split by antigen type, fig.height = 9, fig.width = 8}

ggplot(igg_prediction_with_clustering.df,
       aes(x = sero_age,
           y = prediction,
           group = interaction(type,cluster),
           colour = cluster)) +
  geom_line() +
  geom_point() +
  theme_bw() +
  facet_wrap(.~type) +
  scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "bottom",
        strip.background = element_blank()) +
  xlab("Age")

```

```{r Compare with raw data and clustering, fig.height = 11, fig.width = 8}

ggplot(antigen_exposure_response.df,
       aes(x = sero_age,
           y = igg,
           group = interaction(antigen,codenum),
           colour = type)) +
  geom_line(alpha = 0.05) +
  geom_line(data = igg_prediction_with_clustering.df,
            aes(x = sero_age,
                y = prediction,
                group = type),
            size = 1,
            inherit.aes = FALSE) +
  facet_grid(type~cluster) +
  theme_bw() +
  scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        strip.background = element_blank()) +
  guides(colour = guide_legend(override.aes = list(alpha=1))) +
  xlab("Age") +
  ylab("IgG prediction")

```

```{r Simple plot by type from MEF, fig.width = 8}

mef_type.df <-
  as.data.frame(coef(igg.int_clustering_model_4)[1]) %>%
    dplyr::select(-contains("cluster")) %>%
    dplyr::select(-contains("Intercept")) %>%
    dplyr::slice_head(n = 1) %>%
    tidyr::pivot_longer(everything(),
                        names_to = "coef",
                        values_to = "value") %>%
    dplyr::mutate(coef = gsub("codenum.","",coef)) %>%
    dplyr::mutate(type = dplyr::case_when(
                      grepl("type",coef) ~ gsub(".*type",
                                                "",
                                                coef),
                      TRUE ~ NA_character_
                    )
                  ) %>%
    dplyr::mutate(age = dplyr::case_when(
                      grepl("type",coef) & grepl("sero_age",coef) ~ gsub("\\.type.*",
                                                                        "",
                                                                        gsub("sero_age",
                                                                             "",
                                                                             coef)
                                                                        ),
                      grepl("sero_age",coef) ~ gsub("sero_age",
                                                   "",
                                                   coef
                                                   ),
                      TRUE ~ NA_character_
                    )
                  ) %>%
    dplyr::mutate(type_coef =  dplyr::if_else(is.na(age),value,NA_real_)) %>%
    dplyr::mutate(age_coef =  dplyr::if_else(is.na(type),value,NA_real_)) %>%
    dplyr::group_by(type) %>%
    tidyr::fill(type_coef,.direction = "updown") %>%
    dplyr::ungroup() %>%
    dplyr::group_by(age) %>%
    tidyr::fill(age_coef,.direction = "updown") %>%
    dplyr::ungroup() %>%
    dplyr::mutate(type = dplyr::if_else(is.na(type),"non-ABT",type)) %>%
    dplyr::mutate(age = dplyr::if_else(is.na(age),"Cord",age)) %>%
    dplyr::mutate(interaction_coef = dplyr::if_else((!is.na(type_coef) & !is.na(age_coef)),
                                                    value,
                                                    0
                                                    )
                  ) %>%
    dplyr::mutate(age_coef = dplyr::if_else(is.na(age_coef),0,age_coef)) %>%
    dplyr::mutate(type_coef = dplyr::if_else(is.na(type_coef),0,type_coef)) %>%
    dplyr::select(-value) %>%
    dplyr::mutate(igg = type_coef+age_coef+interaction_coef) %>%
    dplyr::mutate(type = gsub("\\."," ",type)) %>%
    dplyr::mutate(age = gsub("\\."," ",age)) %>%
    dplyr::mutate(age = factor(age,
                               levels = c("Cord",
                                          "Birth",
                                          "6 months",
                                          "12 months",
                                          "18 months",
                                          "24 months")))

ggplot(mef_type.df,
       aes(x = age,
           y = igg,
           colour = type,
           group = type)) +
  geom_point() +
  geom_line() +
  theme_bw()

```

```{r Simpler comparison of model output by type, fig.width = 8}

simple_mef_type.df <-
  igg_prediction_with_clustering.df %>%
    dplyr::mutate(cluster_type = dplyr::if_else(
                      grepl("Neonatal",cluster),
                      "Neonatal",
                      "Infant"
                    )
                  ) %>%
    dplyr::filter((sero_age %in% c("Cord","Birth") & cluster_type == "Neonatal") |
                    (!(sero_age %in% c("Cord","Birth")) & cluster_type == "Infant")) %>%
    dplyr::group_by(sero_age,type) %>%
    dplyr::summarise(igg = mean(prediction)) %>%
    dplyr::ungroup()

(simple_mef_type_plot <-
  ggplot(simple_mef_type.df,
         aes(x = sero_age,
             y = igg,
             colour = type,
             group = type)) +
    geom_line(size = 1) +
    geom_point() +
    ylab("IgG binding prediction") +
    xlab("Age") +
    scale_colour_manual(values = type_palette,
                        aesthetics = c("fill","colour"),
                        drop = TRUE, limits = force,
                        name = "Protein type") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
)

```

```{r What causes a transition between clusters}

transitions.df <-
  antigen_exposure_response.df %>%
    dplyr::select(codenum,cluster,specdate,LRTI_since_last_sample) %>%
    dplyr::distinct() %>%
    dplyr::group_by(codenum) %>%
    dplyr::mutate(LastCluster = dplyr::lag(cluster,order_by = specdate)) %>%
    dplyr::ungroup() %>%
    dplyr::mutate(ClusterTransition = paste0(LastCluster," to ",cluster)) %>%
    dplyr::arrange(codenum,specdate)

ggplot(transitions.df,
       aes(x = ClusterTransition,
           fill = factor(LRTI_since_last_sample),
           colour = factor(LRTI_since_last_sample))) +
  geom_bar() +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

```

```{r Test for consistent rises to final timepoint across clusters, fig.height = 11, fig.width = 8}

# Obsolete and very slow?
igg_levels_by_cluster.df <-
  antigen_exposure_response.df %>%
    dplyr::select(codenum,cluster,antigen,type,sero_age,specdate,igg) %>%
    dplyr::group_by(codenum,antigen) %>%
    dplyr::mutate(ValueAt12Months = dplyr::if_else(
                          sero_age == "12 months",
                          igg,
                          NA_real_
                      )
                  ) %>%
    tidyr::fill(ValueAt12Months,.direction = "updown") %>%
    dplyr::mutate(ValueAt6Months = dplyr::if_else(
                          sero_age == "6 months",
                          igg,
                          NA_real_
                      )
                  ) %>%
    tidyr::fill(ValueAt6Months,.direction = "updown") %>%
    dplyr::mutate(ValueAt24Months = dplyr::if_else(
                          sero_age == "24 months",
                          igg,
                          NA_real_
                      )
                  ) %>%
    tidyr::fill(ValueAt24Months,.direction = "updown") %>%
    dplyr::mutate(Change6to24 = ValueAt24Months - ValueAt6Months) %>%
    dplyr::mutate(Change12to24 = ValueAt24Months - ValueAt12Months) %>%
    dplyr::mutate(FinalCluster = dplyr::if_else(
                          sero_age == "24 months",
                          cluster,
                          NA_character_
                      )
                  ) %>%
    tidyr::fill(FinalCluster,.direction = "updown") %>%
    dplyr::ungroup() %>%
    dplyr::select(codenum,FinalCluster,antigen,type,ValueAt6Months,ValueAt12Months,ValueAt24Months,Change6to24,Change12to24) %>%
    dplyr::distinct()

igg_levels_by_cluster.df %<>%
  tidyr::pivot_longer(cols = -c(codenum,FinalCluster,antigen,type),names_to = "Timepoint",values_to = "igg") %>%
  dplyr::mutate(Timepoint = factor(Timepoint,
                                   levels = c("ValueAt6Months",
                                              "ValueAt12Months",
                                              "ValueAt24Months",
                                              "Change6to24",
                                              "Change12to24")
                                   )
                )

ggplot(igg_levels_by_cluster.df,
       aes(x = type,
           y = igg,
           colour = FinalCluster,
           fill = FinalCluster)) +
  geom_boxplot(outlier.shape = NA,
               position = position_dodge(),
               fill = NA) +
  geom_point(position = position_jitterdodge(jitter.width = 0.1),
             size = 0.5,
             alpha = 0.25) +
  facet_wrap(.~Timepoint,ncol = 1) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        strip.background = element_blank())

```

```{r What determines cluster membership}

lrti_counts.df <-
  antigen_exposure_response.df %>%
    dplyr::select(codenum,sero_age,LRTI_since_last_sample) %>%
    dplyr::filter(sero_age %in% c("12 months","18 months","24 months")) %>%
    dplyr::distinct() %>%
    dplyr::group_by(codenum) %>%
    summarise(Total_LRTI = sum(LRTI_since_last_sample)) %>%
    dplyr::ungroup()

combined_clustering_info.df <-
  igg_levels_by_cluster.df %>%
    dplyr::select(codenum,FinalCluster) %>%
    dplyr::distinct() %>%
    dplyr::left_join(antigen_exposure_response.df %>%
                       dplyr::select(codenum,sero_age,NT_first,first_serotypes,LRTI_since_last_sample,serotypesSinceLastSample) %>%
                       dplyr::filter(sero_age == "6 months") %>%
                       dplyr::distinct(),
                     by = c("codenum")) %>%
    dplyr::left_join(antigen_exposure_response.df %>%
                       dplyr::select(codenum,sero_age,LRTI_since_last_sample,AnyLRTISinceLastSample) %>%
                       dplyr::rename(LRTI_in_next_6_months = LRTI_since_last_sample) %>%
                       dplyr::rename(Any_LRTI_in_next_6_months = AnyLRTISinceLastSample) %>%
                       dplyr::filter(sero_age == "12 months") %>%
                       dplyr::distinct(),
                     by = c("codenum")) %>%
  dplyr::left_join(lrti_counts.df, by = c("codenum"))

message("FinalClusters")
message(combined_clustering_info.df$FinalCluster)

ggplot(combined_clustering_info.df,
       aes(x = FinalCluster,
           colour = NT_first,
           fill = NT_first)) +
  geom_bar() +
  theme_bw()

fisher.test(table(combined_clustering_info.df$FinalCluster,combined_clustering_info.df$NT_first))

```

```{r LRTI by cluster}

ggplot(combined_clustering_info.df,
       aes(x = FinalCluster,
           y = LRTI_in_next_6_months)) +
  geom_beeswarm() +
  ggpubr::stat_compare_means(method = "wilcox.test") +
  theme_bw()

```

```{r Exposures to different Zmp}

early_zmp_exposures.df <-
  antigen_exposure_response.df %>%
    dplyr::filter(sero_age == "6 months" & type %in% c("ZmpA","ZmpB")) %>%
    dplyr::select(codenum,antigen,type,ExposuresSinceBirth) %>%
    dplyr::distinct() %>%
    dplyr::left_join(combined_clustering_info.df %>%
                       dplyr::select(codenum,FinalCluster),
                     by = c("codenum")) %>%
    dplyr::group_by(antigen,type,FinalCluster) %>%
    dplyr::summarise(Count = sum(ExposuresSinceBirth > 0)) %>%
    dplyr::ungroup() %>%
    tidyr::pivot_wider(id_cols = c(antigen,type),
                       names_from = FinalCluster,
                       values_from = Count,
                       values_fill = 0)

zmp_labels.df <-
  early_zmp_exposures.df %>%
    dplyr::filter((`Infant low` > 20 & `Infant high` < 5) | (`Infant low` < 10 & `Infant high` > 7))

ggplot(early_zmp_exposures.df,
       aes(x = `Infant high`,
           y = `Infant low`,
           colour = type)) +
  geom_point() +
  ggrepel::geom_label_repel(data = zmp_labels.df,
                            aes(x = `Infant high`,
                                y = `Infant low`,
                                label = antigen)) +
  scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  theme_bw()

```

## How do immune responses mature?

```{r What type of rise?, fig.height = 8, fig.width = 8}

summarise_igg_distributions.df <-
  antigen_exposure_response.df %>%
    dplyr::select(codenum,sero_age,antigen,type,igg) %>%
    dplyr::group_by(sero_age,type) %>%
    dplyr::summarise(mean = mean(igg),
                     max = max(igg),
                     min = min(igg),
                     variance = var(igg),
                     skewness = skewness(igg),
                     kurtosis = kurtosis(igg)) %>%
    dplyr::ungroup() %>%
    tidyr::pivot_longer(cols = -c(sero_age,type),names_to = "statistic",values_to = "value")

ggplot(summarise_igg_distributions.df,
       aes(x = sero_age,
           y = value,
           group = type,
           colour = type)
       ) +
  geom_line() +
  theme_bw() +
  scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  facet_wrap(.~statistic, scales = 'free_y') +
  theme(strip.background = element_blank()) +
  xlab("Age")

```

```{r Plot quantiles, fig.height = 10, fig.width = 8}

resummarise_igg_distributions.df <-
  antigen_exposure_response.df %>%
    dplyr::select(codenum,sero_age,antigen,type,igg)

ggplot(resummarise_igg_distributions.df,
       aes(x = sero_age,
           y = igg,
           group = type,
           colour = type,
           fill = type)
       ) +
  stat_summary(geom="ribbon",
               fun.min = function(x) quantile(x, 0.25),
               fun.max = function(x) quantile(x, 0.75),
               alpha = 0.25) + 
  stat_summary(geom="line", fun.y=median) +
  scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  theme_bw() +
  facet_wrap(.~type, scales = 'free_y') +
  theme(strip.background = element_blank()) +
  xlab("Age")

```

# Test correlation between early samples

```{r Plot early sample correlations, fig.height = 8, fig.width = 4}

early_sample_correlations.df <-
  antigen_exposure_response.df %>%
    dplyr::select(codenum,sero_age,antigen,type,igg) %>%
    tidyr::pivot_wider(id_cols = c(codenum,antigen,type),
                       names_from = sero_age,
                       values_from = igg) %>%
    dplyr::group_by(antigen,type) %>%
    dplyr::summarise(cor_birth = cor(Cord,Birth,
                               use = "complete.obs"),
                     cor_6mo = cor(Cord,`6 months`,
                               use = "complete.obs"),
                     cor_12mo = cor(Cord,`12 months`,
                               use = "complete.obs"),
                     cor_18mo = cor(Cord,`18 months`,
                               use = "complete.obs"),
                     cor_24mo = cor(Cord,`24 months`,
                               use = "complete.obs"),
                     cor6to_12mo = cor(`6 months`,`12 months`,
                               use = "complete.obs"),
                     cor6to_18mo = cor(`6 months`,`18 months`,
                               use = "complete.obs"),
                     cor6to_24mo = cor(`6 months`,`24 months`,
                               use = "complete.obs"),
                     ) %>%
    dplyr::ungroup() %>%
    tidyr::pivot_longer(cols = c(cor_birth,cor_6mo,cor_12mo,cor_18mo,cor_24mo,cor6to_12mo,cor6to_18mo,cor6to_24mo),
                        names_to = "comparison",
                        values_to = "correlation") %>%
    dplyr::mutate(comparison = factor(comparison,
                                      levels = c("cor_birth","cor_6mo","cor_12mo","cor_18mo","cor_24mo",
                                                 "cor6to_12mo","cor6to_18mo","cor6to_24mo")
                                      )
                  )

early_sample_correlation_tests.df <-
  early_sample_correlations.df %>%
    dplyr::group_by(type,comparison) %>%
    rstatix::wilcox_test(correlation ~ 1,
                             mu = 0,
                             paired = FALSE,
                             p.adjust.method = "holm") %>%
    rstatix::adjust_pvalue() %>%
    rstatix::add_significance("p.adj") %>%
    dplyr::ungroup() %>%
    rstatix::add_xy_position(x = "type", dodge = 1)



outlier_labels_1.df <-
  early_sample_correlations.df %>%
    dplyr::filter(!grepl("cor6to",comparison)) %>%
    dplyr::filter(correlation < -0.1 & type  != "non-ABT") %>%
    dplyr::mutate(type = factor(type,levels = unique(antigen_exposure_response.df$type))) %>%
    dplyr::group_by(antigen) %>%
    dplyr::slice_head(n = 1) %>%
    dplyr::ungroup() %>%
    dplyr::left_join(antigens.df %>% dplyr::select(antigen,prot_label) %>% dplyr::distinct(),
                     by = c("antigen"))

j_width = 1.5
d_width = 0.75

(early_sample_correlations_plot <-
  ggplot(early_sample_correlations.df %>% dplyr::filter(!grepl("cor6to",comparison)),
         aes(x = comparison,
             y = correlation,
             colour = type,
             fill = type)) +
    geom_point(alpha = 0.5,
               size = 0.5,
               position = position_jitterdodge(jitter.width = j_width,
                                               dodge.width = d_width,
                                               seed = 42)
               ) +
    geom_violin(position = position_dodge(width = d_width),
                alpha = 0.1,
                draw_quantiles = c(0.5),
               scale = "width") +
    scale_x_discrete(labels = c("Birth",
                            "6 months",
                            "12 months",
                            "18 months",
                            "24 months")
                     ) +
    scale_y_continuous(limits = c(-0.75,1.2)) +
    scale_colour_manual(values = type_palette,
                        aesthetics = c("fill","colour"),
                        drop = TRUE, limits = force,
                        name = "Protein type") +
    ggrepel::geom_label_repel(data = outlier_labels_1.df,
                              aes(label = prot_label),
                              # position = position_jitterdodge(jitter.width = j_width,
                              #                                 dodge.width = d_width,
                              #                                 seed = 42),
                              fill = NA,
                              force_pull = 1e-6,
                              size = 3,
                              box.padding = 0.0,
                              label.padding = 0.05,
                              nudge_y = -0.1
                              ) +
    geom_hline(yintercept = 0, linetype = 4) +
    theme_bw() +
    ylab("Pearson correlation coefficient with umbilical cord sample") +
    xlab("Comparator sample") +
    facet_wrap(.~type, nrow = 5) +
    theme(legend.position = "bottom",
          axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          strip.background = element_blank()) +
    guides(colour = guide_legend(nrow = 1, title.position = "top"),
           fill = guide_legend(nrow = 1, title.position = "top"),
           ) +
    stat_pvalue_manual(early_sample_correlation_tests.df %>% dplyr::filter(!grepl("cor6to",comparison)),
                       color = "type",
                       x = "comparison",
                       y.position = 1.1)
)

```

```{r Plot later correlations, fig.height = 8, fig.width = 4}

outlier_labels_1b.df <-
  early_sample_correlations.df %>%
    dplyr::filter(grepl("cor6to",comparison)) %>%
    dplyr::filter(correlation < 0.05 & type  != "non-ABT") %>%
    dplyr::mutate(type = factor(type,levels = unique(antigen_exposure_response.df$type))) %>%
    dplyr::group_by(antigen) %>%
    dplyr::slice_head(n = 1) %>%
    dplyr::ungroup() %>%
    dplyr::left_join(antigens.df %>% dplyr::select(antigen,prot_label) %>% dplyr::distinct(),
                     by = c("antigen"))

(late_sample_correlations_plot <-
  ggplot(early_sample_correlations.df %>% dplyr::filter(grepl("cor6to",comparison)),
         aes(x = comparison,
             y = correlation,
             colour = type,
             fill = type)) +
    geom_point(alpha = 0.5,
               size = 0.5,
               position = position_jitterdodge(jitter.width = j_width,
                                               dodge.width = d_width)
               ) +
    geom_violin(position = position_dodge(width = 0.75),
                alpha = 0.1,
                draw_quantiles = c(0.5),
               scale = "width") +
    scale_x_discrete(labels = c("12 months",
                                "18 months",
                                "24 months")
                     ) +
    scale_y_continuous(limits = c(-0.75,1.2)) +
    scale_colour_manual(values = type_palette,
                        aesthetics = c("fill","colour"),
                        drop = TRUE, limits = force,
                        name = "Protein type") +
    ggrepel::geom_label_repel(data = outlier_labels_1b.df,
                              aes(label = prot_label),
                              # position = position_jitterdodge(jitter.width = j_width,
                              #                                 dodge.width = d_width,
                              #                                 seed = 42),
                              fill = NA,
                              force_pull = 1e-6,
                              size = 3,
                              box.padding = 0.0,
                              label.padding = 0.05,
                              nudge_y = -0.1
                              ) +
    geom_hline(yintercept = 0, linetype = 2) +
    theme_bw() +
    xlab("Comparator sample") +
    ylab("Pearson correlation coefficient with umbilical cord sample") +
    facet_wrap(.~type, nrow = 5) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom",
          strip.background = element_blank()) +
    guides(colour = guide_legend(nrow = 1, title.position = "top"),
           fill = guide_legend(nrow = 1, title.position = "top")) +
    stat_pvalue_manual(early_sample_correlation_tests.df %>% dplyr::filter(grepl("cor6to",comparison)),
                       color = "type",
                       x = "comparison",
                       y.position = 1.1)
)

```

Plot an example correlation:

```{r Plot example early v late correlation}

plot_example_correlation <- function(antigen_name) {
  
  example_early_late_correlation.df <-
    antigen_exposure_response.df %>%
      dplyr::filter(antigen == antigen_name & sero_age %in% c("Cord","18 months")) %>%
      dplyr::select(codenum,sero_age,igg) %>%
      tidyr::pivot_wider(id_cols = codenum, values_from = igg, names_from = sero_age)

  ggplot(example_early_late_correlation.df,
         aes(x = Cord,
             y = `18 months`)) +
    geom_smooth(method = "lm") +
    stat_cor(method = "pearson",
            aes(label = paste(..r.label.., ..p.label.., sep = "~`,`~"))) +
    geom_point() +
    ggtitle(antigen_name) +
    theme_bw()
}

cowplot::plot_grid(
  plotlist = list(
    plot_example_correlation("CLS01620"), # Close to 0.25
    plot_example_correlation("CLS00682") # Close to 0.1
  ),
  nrow = 1,
  labels = "auto"
)

```

Plot heritability against frequency in population?

```{r Plot heriability against frequency}

heritability_frequency.df <-
  early_sample_correlations.df %>%
    dplyr::inner_join(cog.frequency.df, by = c("antigen" = "cog"))

ggplot(heritability_frequency.df,
       aes(x = maela_frequency,
           y = correlation,
           colour = type)) +
  geom_point() +
  theme_bw() +
  scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  facet_wrap(.~type) +
  theme(strip.background = element_blank())

```

```{r Plot maternal Ab levels by cluster, fig.height = 11, fig.width = 8}

maternal_levels_and_clusters.df <-
    antigen_exposure_response.df %>%
      dplyr::select(codenum,sero_age,antigen,type,igg) %>%
      tidyr::pivot_wider(id_cols = c(codenum,antigen,type),
                           names_from = sero_age,
                           values_from = igg) %>% 
      dplyr::left_join(
        antigen_exposure_response.df %>%
          dplyr::select(codenum,sero_age,cluster) %>%
          dplyr::filter(sero_age == "24 months") %>%
          dplyr::select(-sero_age) %>%
          dplyr::rename(final_cluster = cluster) %>%
          dplyr::distinct(),
        by = c("codenum")
      ) %>%
      tidyr::pivot_longer(cols = c(Cord,Birth,`6 months`,`12 months`,`18 months`,`24 months`),
                        names_to = "Age",
                        values_to = "igg"
      ) %>%
      dplyr::mutate(Age = factor(Age,
                                 levels = c("Cord","Birth","6 months","12 months","18 months","24 months"))
                    )

cluster_test.df <-
    maternal_levels_and_clusters.df %>%
      dplyr::select(Age,type,final_cluster,igg) %>%
      dplyr::group_by(Age,type) %>%
      rstatix::wilcox_test(formula("igg ~ final_cluster"),
                           ref.group = "Infant low",
                           paired = FALSE,
                           p.adjust.method = "holm") %>%
      rstatix::adjust_pvalue() %>%
      rstatix::add_significance("p.adj") %>%
      dplyr::ungroup() %>%
      rstatix::add_xy_position(x = "type", dodge = 1)

(final_cluster_separation <-
    ggplot(maternal_levels_and_clusters.df,
       aes(x = type,
           y = igg,
           colour = final_cluster,
           fill = final_cluster)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.1),
             alpha = 0.25,
             size = 0.25) +
  geom_violin(draw_quantiles = c(0.5),
              alpha = 0.25,
               scale = "width") +
  ggpubr::stat_pvalue_manual(cluster_test.df,
                              label = "p.adj.signif",
                              tip.length = 0,
                              inherit.aes = FALSE) +
  facet_wrap(.~Age, ncol = 2) +
  xlab("Protein type") +
  ylab("IgG binding") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "right",
        strip.background = element_blank()) +
  scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster")
)

```

```{r Alternative fit antigen models to inheritance and exposure, fig.height = 11, fig.width = 8}

inheritance_and_exposure.df <-
  antigen_exposure_response.df %>%
    dplyr::select(codenum,sero_specnum,cluster,antigen,type,sero_age,igg,SinceBirth,FirstSixMonths,LastSample,AnyLRTISinceLastSample,AnySevereLRTISinceLastSample,AnyVerySevereLRTISinceLastSample) %>%
    dplyr::group_by(codenum) %>%
    dplyr::mutate(Cord_IgG = dplyr::if_else(sero_age == 'Cord',igg,NA)) %>%
    tidyr::fill(Cord_IgG, .direction = "updown") %>%
    dplyr::ungroup() %>%
    dplyr::filter(sero_age %in% c("6 months","12 months","18 months","24 months")) %>%
    dplyr::mutate(SinceBirth = factor(SinceBirth,
                                      levels = c("NotExposedSinceBirth","ExposedSinceBirth")
                                      )
                  ) %>%
    dplyr::mutate(FirstSixMonths = factor(FirstSixMonths,
                                      levels = c("NotExposedInFirstSixMonths","ExposedInFirstSixMonths")
                                      )
                  ) %>%
    dplyr::mutate(LastSample = factor(LastSample,
                                      levels = c("NotExposedSinceLastSample","ExposedSinceLastSample")
                                      )
    ) %>%
    dplyr::filter(type != "non-ABT") %>%
    dplyr::filter(!is.na(igg) & !is.na(Cord_IgG)) %>%
    dplyr::group_by(antigen,sero_age) %>%
    dplyr::mutate(UnexposedCount = sum(SinceBirth == "NotExposedSinceBirth")) %>%
    dplyr::mutate(ExposedCount = sum(SinceBirth == "ExposedSinceBirth")) %>%
    dplyr::ungroup() %>%
    dplyr::mutate(AnyLRTISinceLastSample = factor(AnyLRTISinceLastSample, levels = c("NoLRTISinceLastSample","LRTISinceLastSample"))) %>%
    dplyr::mutate(AnySevereLRTISinceLastSample = factor(AnySevereLRTISinceLastSample, levels = c("NoSevereLRTISinceLastSample","SevereLRTISinceLastSample"))) %>%
    dplyr::mutate(AnyVerySevereLRTISinceLastSample = factor(AnyVerySevereLRTISinceLastSample, levels = c("NoVerySevereLRTISinceLastSample","VerySevereLRTISinceLastSample")))

inheritance_and_exposure_ordered_individuals <-
  sort(unique(inheritance_and_exposure.df$codenum))

inheritance_and_exposure.df %<>%
  dplyr::mutate(codenum = factor(codenum,
                                 levels = inheritance_and_exposure_ordered_individuals))

# ensure all antigens remain unexposed at final timepoint in at least one individual
both_inheritance_and_exposure.antigen_list <-
  inheritance_and_exposure.df %>%
    dplyr::filter(sero_age == "12 months" & UnexposedCount > 1 & ExposedCount > 1) %>%
    dplyr::select(antigen) %>%
    dplyr::pull()

both_inheritance_and_exposure.df <-
  inheritance_and_exposure.df %>%
    dplyr::filter(antigen %in% both_inheritance_and_exposure.antigen_list)

just_inheritance.df <-
  inheritance_and_exposure.df %>%
    dplyr::filter(!(antigen %in% both_inheritance_and_exposure.antigen_list))

simple.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + (1|codenum) + 0,
       data = just_inheritance.df,
       REML = FALSE)

inheritance.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + Cord_IgG*antigen + (1|codenum) + 0,
       data = just_inheritance.df,
       REML = FALSE)

```

```{r Fit models including exposure}

inheritance_and_exposure_since_birth.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + SinceBirth*antigen + Cord_IgG*antigen + (1|codenum) + 0,
       data = both_inheritance_and_exposure.df,
       REML = FALSE)

inheritance_and_exposure_first_six_months.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + FirstSixMonths*antigen + Cord_IgG*antigen + (1|codenum) + 0,
       data = both_inheritance_and_exposure.df,
       REML = FALSE)

inheritance_and_exposure_last_six_months.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + LastSample*antigen + Cord_IgG*antigen + (1|codenum) + 0,
       data = both_inheritance_and_exposure.df,
       REML = FALSE)

antigen_response_models <- c(simple.antigen_model,
                             inheritance.antigen_model,
                             inheritance_and_exposure_since_birth.antigen_model,
                             inheritance_and_exposure_first_six_months.antigen_model,
                             inheritance_and_exposure_last_six_months.antigen_model)

antigen_response_model_comparisons.df <-
  pairwise_model_comparisons(antigen_response_models)

ggplot(antigen_response_model_comparisons.df,
       aes(x = i,
           y = j,
           colour = delta_aic,
           fill = delta_aic)) +
  geom_tile() +
  scale_colour_distiller(type="div",aesthetics = c("colour","fill")) +
  xlab("Reference model index") +
  ylab("Comparison model index") +
  theme_bw()+
  guides(colour = guide_colourbar(title = expression("AIC"[ref]-"AIC"[comp])),
         fill = guide_colourbar(title = expression("AIC"[ref]-"AIC"[comp])))

```

```{r Plot best fitting model properties, fig.height = 8, fig.width = 8}

get_type_values <- function(df) {
  df  %>%
     dplyr::select(dplyr::starts_with('type')) %>%
     tidyr::pivot_longer(cols = everything(),
                                names_to = "variable",
                                values_to = "type_response") %>%
     dplyr::mutate(type = gsub("type",
                                    "",
                                    variable)
                          ) %>%
    dplyr::distinct() %>%
    dplyr::select(-variable) %>%
    tibble::add_row(type_response = 1, type = "Adhesin")
}

combine_exposure_model_outputs <- function(i.model, ie.model) {

    inheritance_and_exposure.coef <- as.data.frame(coef(ie.model)[[1]])
    base_inheritance_and_exposure_igg_level <- inheritance_and_exposure.coef$`(Intercept)`
    base_inheritance_and_exposure_cord_correlation <- unique(inheritance_and_exposure.coef$Cord_IgG)
    base_inheritance_and_exposure_since_birth <- unique(inheritance_and_exposure.coef$SinceBirthExposedSinceBirth)
    
    base_inheritance_and_exposure_6_months <- unique(inheritance_and_exposure.coef$`sero_age6 months`)
    base_inheritance_and_exposure_12_months <- unique(inheritance_and_exposure.coef$`sero_age12 months`)
    base_inheritance_and_exposure_18_months <- unique(inheritance_and_exposure.coef$`sero_age18 months`)
    base_inheritance_and_exposure_24_months <- unique(inheritance_and_exposure.coef$`sero_age24 months`)
    
    inheritance.coef <- as.data.frame(coef(i.model)[[1]])
    base_inheritance_igg_level <- inheritance.coef$`(Intercept)`
    base_inheritance_cord_correlation <- unique(inheritance.coef$Cord_IgG)
    base_inheritance_6_months <- unique(inheritance.coef$`sero_age6 months`)
    base_inheritance_12_months <- unique(inheritance.coef$`sero_age12 months`)
    base_inheritance_18_months <- unique(inheritance.coef$`sero_age18 months`)
    base_inheritance_24_months <- unique(inheritance.coef$`sero_age24 months`)
    
    individuals_model_comparisons.df <-
      tibble::tibble(
        "Individual" = inheritance_and_exposure_ordered_individuals,
        "Inheritance_exposure" = inheritance_and_exposure.coef$`(Intercept)`,
        "Inheritance" = inheritance.coef$`(Intercept)`
      )
    inheritance_and_exposure_types.df <- get_type_values(inheritance_and_exposure.coef)
    inheritance_types.df <- get_type_values(inheritance.coef)
    
    combined_output.df <-
      dplyr::bind_rows(
        ############################################
        # Models including differences in exposure #
        ############################################
        # Factors for exposure of antigens
        inheritance_and_exposure.coef %>%
          dplyr::select(dplyr::ends_with('SinceBirthExposedSinceBirth')) %>%
          tidyr::pivot_longer(cols = everything(),
                              names_to = "variable",
                              values_to = "IgG_After_Exposure") %>%
          dplyr::mutate(antigen = gsub("antigen","",
                                       gsub(":SinceBirthExposedSinceBirth",
                                            "",
                                            variable))) %>%
          dplyr::distinct() %>%
          dplyr::filter(antigen != "SinceBirthExposedSinceBirth") %>%
          dplyr::mutate(IgG_After_Exposure = IgG_After_Exposure + base_inheritance_and_exposure_since_birth) %>%
          dplyr::select(-variable) %>%
          dplyr::left_join(
            # Add types for antigens
            inheritance_and_exposure.df %>%
              dplyr::select(antigen,type) %>%
              dplyr::distinct(),
            by = c("antigen")
          ) %>%
          dplyr::left_join(
              # Factors for inheritance of IgG levels
              inheritance_and_exposure.coef %>%
                dplyr::select(dplyr::ends_with(':Cord_IgG')) %>%
                tidyr::pivot_longer(cols = everything(),
                                    names_to = "variable",
                                    values_to = "IgG_in_cord") %>%
                dplyr::mutate(antigen = gsub("antigen",
                                             "",
                                             gsub(":Cord_IgG",
                                                  "",
                                                  variable))) %>%
                dplyr::distinct() %>%
                dplyr::mutate(IgG_in_cord = IgG_in_cord + base_inheritance_and_exposure_cord_correlation) %>%
                dplyr::select(-variable),
            by = c("antigen")
          ) %>%
          
          dplyr::left_join(
              # Factors for age - 6 months
              inheritance_and_exposure_types.df %>%
                dplyr::rename(month_6 = type_response),
            by = c("type")
          ) %>%
          
          dplyr::left_join(
              # Factors for age - 12 months
              inheritance_and_exposure.coef %>%
                dplyr::select(dplyr::starts_with('sero_age12 months:type')) %>%
                tidyr::pivot_longer(cols = everything(),
                                    names_to = "variable",
                                    values_to = "month_12") %>%
                dplyr::mutate(type = gsub("sero_age12 months:type",
                                          "",
                                          variable)
                              ) %>%
                dplyr::distinct() %>%
                dplyr::select(-variable) %>%
                dplyr::left_join(
                  inheritance_and_exposure_types.df,
                  by = c("type")
                ) %>%
                dplyr::mutate(month_12 = month_12 + type_response) %>%
                dplyr::select(-type_response),
            by = c("type")
          ) %>%
          
          dplyr::left_join(
              # Factors for age - 18 months
              inheritance_and_exposure.coef %>%
                dplyr::select(dplyr::starts_with('sero_age18 months:type')) %>%
                tidyr::pivot_longer(cols = everything(),
                                    names_to = "variable",
                                    values_to = "month_18") %>%
                dplyr::mutate(type = gsub("sero_age18 months:type",
                                          "",
                                          variable)
                              ) %>%
                dplyr::distinct() %>%
                dplyr::select(-variable) %>%
                dplyr::left_join(
                  inheritance_and_exposure_types.df,
                  by = c("type")
                ) %>%
                dplyr::mutate(month_18 = month_18 + type_response) %>%
                dplyr::select(-type_response),
            by = c("type")
          ) %>%
          
          dplyr::left_join(
              # Factors for age - 24 months
              inheritance_and_exposure.coef %>%
                dplyr::select(dplyr::starts_with('sero_age24 months:type')) %>%
                tidyr::pivot_longer(cols = everything(),
                                    names_to = "variable",
                                    values_to = "month_24") %>%
                dplyr::mutate(type = gsub("sero_age24 months:type",
                                          "",
                                          variable)
                              ) %>%
                dplyr::distinct() %>%
                dplyr::select(-variable) %>%
                dplyr::left_join(
                  inheritance_and_exposure_types.df,
                  by = c("type")
                ) %>%
                dplyr::mutate(month_24 = month_24 + type_response) %>%
                dplyr::select(-type_response),
            by = c("type")
          ) %>%
          
          dplyr::left_join(
              # Responses by antigen
              inheritance_and_exposure.coef %>%
                dplyr::select(dplyr::starts_with('antigen')) %>%
                tidyr::pivot_longer(cols = everything(),
                                    names_to = "variable",
                                    values_to = "antigen_response") %>%
                dplyr::mutate(antigen = gsub("antigen",
                                             "",
                                             variable)
                              ) %>%
                dplyr::distinct() %>%
                dplyr::mutate(antigen_response = antigen_response) %>%
                dplyr::select(-variable),
            by = c("antigen")
          ) %>%
          tidyr::pivot_longer(cols = c(month_6,
                                       month_12,
                                       month_18,
                                       month_24,
                                       antigen_response,
                                       IgG_in_cord,
                                       IgG_After_Exposure),
                              names_to = "IgG_response",
                              values_to = "IgG") %>%
          
          # Add adhesin levels (as reference here)
          dplyr::mutate(IgG = dplyr::case_when(
                    IgG_response == "month_6" & type == "Adhesin" ~ base_inheritance_and_exposure_6_months,
                    IgG_response == "month_12" & type == "Adhesin" ~ base_inheritance_and_exposure_12_months,
                    IgG_response == "month_18" & type == "Adhesin" ~ base_inheritance_and_exposure_18_months,
                    IgG_response == "month_24" & type == "Adhesin" ~ base_inheritance_and_exposure_24_months,
                    IgG_response == "month_6" & type != "Adhesin" ~ IgG + base_inheritance_and_exposure_6_months,
                    IgG_response == "month_12" & type != "Adhesin" ~ IgG + base_inheritance_and_exposure_12_months,
                    IgG_response == "month_18" & type != "Adhesin" ~ IgG + base_inheritance_and_exposure_18_months,
                    IgG_response == "month_24" & type != "Adhesin" ~ IgG + base_inheritance_and_exposure_24_months,
                    IgG_response == "type_response" & type == "Adhesin" ~ 1,
                    TRUE ~ IgG
                  )
          ) %>%
          dplyr::mutate(Model = "Differential exposure"),
        ################################################
        # Models not including differences in exposure #
        ################################################
        inheritance.coef %>%
    
          dplyr::select(dplyr::ends_with(':Cord_IgG')) %>%
                tidyr::pivot_longer(cols = everything(),
                                    names_to = "variable",
                                    values_to = "IgG_in_cord") %>%
                dplyr::mutate(antigen = gsub("antigen",
                                             "",
                                             gsub(":Cord_IgG",
                                                  "",
                                                  variable))) %>%
                dplyr::distinct() %>%
                dplyr::mutate(IgG_in_cord = IgG_in_cord + base_inheritance_cord_correlation) %>%
                dplyr::select(-variable) %>%
          
          dplyr::left_join(
            # Add types for antigens
            inheritance_and_exposure.df %>%
              dplyr::select(antigen,type) %>%
              dplyr::distinct(),
            by = c("antigen")
          ) %>%
          
          dplyr::left_join(
              # Factors for age - 6 months
              inheritance_types.df %>%
                dplyr::rename(month_6 = type_response),
            by = c("type")
          ) %>%
          
          dplyr::left_join(
              # Factors for age - 12 months
              inheritance.coef %>%
                dplyr::select(dplyr::starts_with('sero_age12 months:type')) %>%
                tidyr::pivot_longer(cols = everything(),
                                    names_to = "variable",
                                    values_to = "month_12") %>%
                dplyr::mutate(type = gsub("sero_age12 months:type",
                                          "",
                                          variable)
                              ) %>%
                dplyr::distinct() %>%
                dplyr::select(-variable) %>%
                dplyr::left_join(
                  inheritance_types.df,
                  by = c("type")
                ) %>%
                dplyr::mutate(month_12 = month_12 + type_response) %>%
                dplyr::select(-type_response),
            by = c("type")
          ) %>%
          
          dplyr::left_join(
              # Factors for age - 18 months
              inheritance.coef %>%
                dplyr::select(dplyr::starts_with('sero_age18 months:type')) %>%
                tidyr::pivot_longer(cols = everything(),
                                    names_to = "variable",
                                    values_to = "month_18") %>%
                dplyr::mutate(type = gsub("sero_age18 months:type",
                                          "",
                                          variable)
                              ) %>%
                dplyr::distinct() %>%
                dplyr::select(-variable) %>%
                dplyr::left_join(
                  inheritance_types.df,
                  by = c("type")
                ) %>%
                dplyr::mutate(month_18 = month_18 + type_response) %>%
                dplyr::select(-type_response),
            by = c("type")
          ) %>%
          
          dplyr::left_join(
              # Factors for age - 24 months
              inheritance.coef %>%
                dplyr::select(dplyr::starts_with('sero_age24 months:type')) %>%
                tidyr::pivot_longer(cols = everything(),
                                    names_to = "variable",
                                    values_to = "month_24") %>%
                dplyr::mutate(type = gsub("sero_age24 months:type",
                                          "",
                                          variable)
                              ) %>%
                dplyr::distinct() %>%
                dplyr::select(-variable) %>%
                dplyr::left_join(
                  inheritance_types.df,
                  by = c("type")
                ) %>%
                dplyr::mutate(month_24 = month_24 + type_response) %>%
                dplyr::select(-type_response),
            by = c("type")
          ) %>%
    
          dplyr::left_join(
              # Responses by antigen
              inheritance.coef %>%
                dplyr::select(dplyr::starts_with('antigen')) %>%
                tidyr::pivot_longer(cols = everything(),
                                    names_to = "variable",
                                    values_to = "antigen_response") %>%
                dplyr::mutate(antigen = gsub("antigen",
                                             "",
                                             variable)
                              ) %>%
                dplyr::distinct() %>%
                dplyr::mutate(antigen_response = antigen_response) %>%
                dplyr::select(-variable),
            by = c("antigen")
          ) %>%
          tidyr::pivot_longer(cols = c(month_6,
                                       month_12,
                                       month_18,
                                       month_24,
                                       antigen_response,
                                       IgG_in_cord),
                              names_to = "IgG_response",
                              values_to = "IgG") %>%
          
          # Add adhesin levels (as reference here)
          dplyr::mutate(IgG = dplyr::case_when(
                    IgG_response == "month_6" & type == "Adhesin" ~ base_inheritance_6_months,
                    IgG_response == "month_12" & type == "Adhesin" ~ base_inheritance_12_months,
                    IgG_response == "month_18" & type == "Adhesin" ~ base_inheritance_18_months,
                    IgG_response == "month_24" & type == "Adhesin" ~ base_inheritance_24_months,
                    IgG_response == "month_6" & type != "Adhesin" ~ IgG + base_inheritance_6_months,
                    IgG_response == "month_12" & type != "Adhesin" ~ IgG + base_inheritance_12_months,
                    IgG_response == "month_18" & type != "Adhesin" ~ IgG + base_inheritance_18_months,
                    IgG_response == "month_24" & type != "Adhesin" ~ IgG + base_inheritance_24_months,
                    IgG_response == "type_response" & type == "Adhesin" ~ 1,
                    TRUE ~ IgG
                  )
          ) %>%
          dplyr::mutate(Model = "Universal exposure")
      ) %>%
      dplyr::mutate(IgG_response = dplyr::case_when(
                      IgG_response == "antigen_response" ~ "Binding to protein type",
                      IgG_response == "IgG_in_cord" ~ "Correlation with cord sample",
                      IgG_response == "IgG_After_Exposure" ~ "Change after exposure",
                      TRUE ~ IgG_response
                    )
      ) %>%
      dplyr::mutate(IgG_response = factor(IgG_response,
                                          levels = c("month_6",
                                                     "month_12",
                                                     "month_18",
                                                     "month_24",
                                                     "Binding to protein type",
                                                     "Correlation with cord sample",
                                                     "Change after exposure")
                                          )
                    )
    return(list(combined_output.df,individuals_model_comparisons.df))
}

combined_exposure_model_outputs <- combine_exposure_model_outputs(inheritance.antigen_model, inheritance_and_exposure_since_birth.antigen_model)
antigen_response_analysis.df <- combined_exposure_model_outputs[[1]]
individuals_model_comparisons.df <- combined_exposure_model_outputs[[2]]

exposure_outlier_labels.df <-
  antigen_response_analysis.df %>%
    dplyr::filter(IgG_response == "Change after exposure") %>%
    dplyr::filter(IgG > 1) %>%
    dplyr::left_join(antigens.df %>% dplyr::select(antigen,prot_label) %>% dplyr::distinct(),
                     by = c("antigen"))

(modelling_response_after_exposure_plot <-
  ggplot(antigen_response_analysis.df %>%
           dplyr::filter(IgG_response %in% c("Binding to protein type",
                                                   "Correlation with cord sample",
                                                   "Change after exposure")) %>%
         dplyr::mutate(Model = dplyr::if_else(Model == "Differential exposure",
                                              "Partial exposure",
                                              Model)),
         aes(x = type,
             y = IgG,
             colour = type,
             fill = type)) +
    geom_point(position = position_jitterdodge(seed = 42),
               alpha = 0.5,
               size = 1) +
    geom_violin(draw_quantiles = c(0.5),
               alpha = 0.25,
               scale = "width") +
    facet_grid(IgG_response~Model,
               scales = 'free_y',
               labeller = labeller(IgG_response = label_wrap_gen(18))) +
    ggrepel::geom_label_repel(data = exposure_outlier_labels.df %>%
         dplyr::mutate(Model = dplyr::if_else(Model == "Differential exposure",
                                              "Partial exposure",
                                              Model)),
                              aes(label = prot_label),
                              min.segment.length = 100,
                              fill = NA,
                              force_pull = 1e-6,
                              label.padding = 0.25,
                              nudge_x = 0.5,
                              nudge_y = 0.25,
                              size = 3
                              ) +
    xlab("Protein type") +
    ylab("Parameter estimate") +
    theme_bw() +
    scale_colour_manual(values = type_palette,
                        aesthetics = c("fill","colour"),
                        drop = TRUE, limits = force,
                        name = "Protein type") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          strip.background = element_blank())
)

```

```{r Plot types by age, fig.width = 8, fig.height = 8}

ggplot(antigen_response_analysis.df %>%
         dplyr::filter(IgG_response %in% c("month_6",
                                                 "month_12",
                                                 "month_18",
                                                 "month_24")) %>%
         dplyr::select(type,IgG_response,IgG,Model) %>%
         dplyr::distinct() %>%
         dplyr::mutate(Model = dplyr::if_else(Model == "Differential exposure",
                                              "Partial exposure",
                                              Model)),
       aes(x = IgG_response,
           y = IgG,
           colour = type,
           fill = type,
           shape = Model,
           linetype = Model,
           group = interaction(type,Model))) +
  geom_point(alpha = 0.5,
             size = 1) +
  geom_line() +
  scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  scale_x_discrete(labels = c("6 months","12 months","18 months","24 months")) +
  ylab("IgG binding") +
  xlab("Age") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

```

```{r Plot random effects by individual}

(individuals_comparison_plot <-
  ggplot(individuals_model_comparisons.df,
         aes(x = Inheritance,
             y = Inheritance_exposure)) +
  geom_point() +
  geom_smooth(method = "lm",formula=y~1+x) +
  xlab("Individual random effect in universal exposure model") +
  ylab("Individual random effect in partial exposure model") +
  stat_cor(method = "pearson",
             p.accuracy = 0.000001,
             label.sep='\n'
  ) +
  theme_bw()
)

```

```{r Plot prediction againt observation for inheritance exposure lmers, fig.height = 8, fig.width = 8}

inheritance_predictions <- predict(inheritance.antigen_model)
inheritance_and_exposure_predictions <- predict(inheritance_and_exposure_since_birth.antigen_model)

inheritance_and_exposure_with_predictions.df <-
  inheritance_and_exposure.df %>%
      dplyr::mutate(Exposed = dplyr::if_else(antigen %in% both_inheritance_and_exposure.antigen_list,
                                                  "Partial",
                                                  "Universal"
                                                  )
                    ) %>%
      dplyr::group_by(Exposed) %>%
      dplyr::mutate(index=row_number()) %>%
      dplyr::ungroup() %>%
      dplyr::left_join(
        dplyr::bind_rows(
          tibble(
            "predicted" = inheritance_predictions,
            "index"=1:length(inheritance_predictions),
            "Exposed" = "Universal"
          ),
          tibble(
            "predicted" = inheritance_and_exposure_predictions,
            "index"=1:length(inheritance_and_exposure_predictions),
            "Exposed" = "Partial"
          )
        ),
        by = c("Exposed","index")
      )

ggplot(inheritance_and_exposure_with_predictions.df,
       aes(x = igg,
           y = predicted,
           colour = Exposed)) +
  geom_point(alpha = 0.25,
             size = 0.25) +
  geom_smooth(method = "lm") +
  xlab("Observed IgG level") +
  ylab("Predicted IgG level") +
  stat_cor(method = "pearson",
             p.accuracy = 0.000001,
             label.sep='\n'
  ) +
  theme_bw()

```

```{r Seroconversion for antigens from LMEM, fig.height = 11, fig.width = 8}

n_exp = 0:24

lmem_exposure_summary_plot.df <-
  exposure_summary.df %>%
    dplyr::filter(antigen %in% exposure_outlier_labels.df$antigen |
                       antigen %in% outlier_labels_1.df$antigen |
                       antigen %in% outlier_labels_1b.df$antigen |
                    gene %in% candidate_vaccine_antigens) %>%
    dplyr::left_join(antigen_exposure_response.df %>%
                       dplyr::filter(sero_age %in% c("6 months","12 months","18 months","24 months")) %>%
                       dplyr::select(antigen,codenum,ExposuresSinceBirth,igg,sero_age),
                     by = c("antigen")) %>%
    dplyr::mutate(name = paste0(gene,"\n(",antigen,")"))

lmem_exposure_summary_models_plot.df <-
  lmem_exposure_summary_plot.df %>%
    dplyr::select(antigen,name,A,B,C) %>%
    dplyr::rowwise() %>%
    dplyr::mutate(n_exp = list(n_exp),
                  pred_val = list(B - C*exp(-1*A*n_exp))) %>%
    dplyr::ungroup() %>%
    tidyr::unnest(cols = c(n_exp,pred_val))

ggplot(lmem_exposure_summary_plot.df,
         aes(x = ExposuresSinceBirth,
             y = igg,
             group = antigen,
             shape = sero_age)) +
    geom_point(position = position_jitter(width = 0.05),
               alpha = 0.5,
               size = 0.5) +
    # geom_line(data = lmem_exposure_summary_models_plot.df,
    #           aes(x = n_exp,
    #               y = pred_val),
    #           inherit.aes = FALSE,
    #           colour = "red") +
    geom_smooth() +
    ylim(c(-1,8)) +
    xlab("Number of antigen-positive samples since birth") +
    ylab("IgG binding") +
    theme_bw() +
    scale_shape_manual(values = sero_age_shapes[names(sero_age_shapes) %in% c("6 months","12 months","18 months","24 months")],
                       name = "Age") +
    facet_wrap(.~name) +
    theme(legend.position = "bottom",
          strip.background = element_blank())

```

```{r Seroconversion for PclA, fig.height = 8, fig.width = 8}

n_exp = 0:24

pclA_exposure_summary_plot.df <-
  exposure_summary.df %>%
    dplyr::filter(gene == "pclA") %>%
    dplyr::left_join(antigen_exposure_response.df %>%
                       dplyr::filter(sero_age %in% c("6 months",
                                                     "12 months",
                                                     "18 months",
                                                     "24 months")) %>%
                      dplyr::select(antigen,codenum,ExposuresSinceBirth,igg,sero_age),
                     by = c("antigen")) %>%
    dplyr::mutate(name = paste0(gene,"\n(",antigen,")")) %>%
    dplyr::group_by(gene,codenum,sero_age) %>%
    dplyr::mutate(pclA_exposures = sum(ExposuresSinceBirth)) %>%
    dplyr::ungroup()

ggplot(pclA_exposure_summary_plot.df,
         aes(x = pclA_exposures,
             y = igg,
             group = antigen,
             shape = sero_age)) +
    geom_point(position = position_jitter(width = 0.05),
               alpha = 0.5,
               size = 0.5) +
    geom_smooth() +
    ylim(c(-1,8)) +
    xlab("Number of antigen-positive samples since birth") +
    ylab("IgG binding") +
    theme_bw() +
    scale_shape_manual(values = sero_age_shapes[names(sero_age_shapes) %in% c("6 months","12 months","18 months","24 months")],
                       name = "Age") +
    facet_wrap(.~name) +
    theme(legend.position = "bottom",
          strip.background = element_blank())

```

```{r Overall summary of exposure, fig.height = 8, fig.width = 8}

exposure_splines.df <-
  antigen_exposure_response.df %>%
    dplyr::filter(sero_age %in% c("6 months",
                                  "12 months",
                                  "18 months",
                                  "24 months")) %>%
    dplyr::filter(type != "non-ABT") %>%
    dplyr::select(antigen,gene,type,igg,ExposuresSinceBirth) %>%
    dplyr::group_by(antigen,gene,type,ExposuresSinceBirth) %>%
    dplyr::summarise(median = median(igg),
                     mean = mean(igg),
                     max = max(igg),
                     min = min(igg),
                     count = n()) %>%
    dplyr::ungroup() %>%
    dplyr::filter(count >= 3) %>%
    dplyr::mutate(`Effect of Exposure` = dplyr::if_else(
                      (antigen %in% exposure_outlier_labels.df$antigen),
                      "Response",
                      "No response"
                    )
                  )

exposure_splines_labels.df <-
  exposure_splines.df %>%
    dplyr::filter(`Effect of Exposure` == "Response") %>%
    dplyr::group_by(antigen) %>%
    dplyr::filter(ExposuresSinceBirth == max(ExposuresSinceBirth)) %>%
    dplyr::ungroup() %>%
    dplyr::left_join(antigens.df %>% dplyr::select(antigen,prot_label) %>% dplyr::distinct(),
                     by = c("antigen"))

(response_by_exposures_plot <-
  ggplot(exposure_splines.df,
         aes(x = ExposuresSinceBirth,
             y = mean,
             colour = type,
             groups = antigen,
             linetype = `Effect of Exposure`,
             alpha = `Effect of Exposure`)) +
    geom_line() +
    scale_linetype_manual(values = c("Response" = 1,
                                     "No response" = 1)
                          ) +
    scale_alpha_manual(values = c("Response" = 1,
                                     "No response" = 0.25)
                       ) +
    ylab("Median IgG level") +
    xlab("Number of positive samples since birth") +
    facet_wrap(.~type) +
    ggrepel::geom_label_repel(data = exposure_splines_labels.df,
                              aes(label = prot_label),
                              fill = "white",
                              nudge_x = 2,
                              force_pull = 1e-6,
                              label.padding = 0.25
                            ) +
    scale_colour_manual(values = type_palette,
                        aesthetics = c("fill","colour"),
                        drop = TRUE, limits = force,
                        name = "Protein type") +
    theme_bw() +
    theme(legend.position = "bottom", legend.justification = "center",
          strip.background = element_blank()) +
    guides(fill = guide_legend(title.position = "top", title.hjust=0.5),
           linetype = guide_legend(title.position = "top", title.hjust=0.5))
)

```

# Machine learning prediction of cluster membership from birth

```{r Functions for SVM analysis}

make_svm_input <- function(ae.df,age=NULL) {
    dplyr::left_join(
      ae.df %>%
        dplyr::select(codenum,antigen,sero_age,igg) %>%
        dplyr::filter(sero_age == age) %>%
        dplyr::select(-c(sero_age)) %>%
        tidyr::pivot_wider(id_cols = codenum,names_from = antigen,values_from = igg),
      ae.df %>%
        dplyr::select(codenum,sero_age,cluster) %>%
        dplyr::filter(sero_age == "24 months") %>%
        dplyr::select(-c(sero_age)),
      by = c("codenum")
    ) %>%
    dplyr::select(-codenum) %>%
    dplyr::mutate(FinalCluster = factor(cluster)) %>%
    dplyr::select(-cluster)
}

normalise_svm_input <- function(df) {
  i.df <-
    cbind(
      apply(
        as.matrix(
          df %>%
            dplyr::select(-FinalCluster)
        ),
        2,
        function(v){(v-mean(v)/sd(v))}
      ),
      df %>%
            dplyr::select(FinalCluster)
    )
  return(i.df)
}

plot_svm_weights <- function(svm,a.df) {
  # Inspired by https://link.springer.com/content/pdf/10.1023/A:1012487302797.pdf
  weights<- apply(svm$SV, 2, function(v){sqrt(sum(v^2))})  # weight
  
  df <-
    a.df %>%
      dplyr::select(antigen,type) %>%
      dplyr::distinct() %>%
      dplyr::left_join(
        data.frame(
          "antigen" = names(weights),
          "SVM weight" = as.numeric(weights)
        ),
        by = c("antigen")
      )
  
  ggplot(df,
         aes(x = type,
             y = SVM.weight,
             colour = type,
             fill = type)) +
    geom_point(position = position_jitter(width = 0.1),
               alpha = 1,
               size = 0.5) +
    geom_violin(alpha = 0.1,draw_quantiles = c(0.5),
               scale = "width") +
    scale_colour_manual(values = type_palette,
                        aesthetics = c("fill","colour"),
                        drop = TRUE, limits = force,
                        name = "Protein type") +
    theme_bw() +
    ylab("Support vector weights")
}

```

```{r Prediction with SVM from birth data, eval = FALSE}

svm_input <- make_svm_input(antigen_exposure_response.df,age="Birth")

tune.out<-tune(e1071::svm,
              FinalCluster ~ .,
              data=svm_input,
              kernel = "linear", 
              ranges =list(cost=c(0.001,0.01,0.1, 1,5,10,100)),
              scale = FALSE)

svmfit = e1071::svm(FinalCluster ~ .,
             data = svm_input,
             kernel = "linear",
             cost = 10,
             scale = FALSE)

save(tune.out,svm_input,svmfit,
     file="Maela_antigens_svm_data.Rdata")

```

```{r Demonstrate SVM results have no error independent of cost function}

load("Maela_antigens_svm_data.Rdata")

tune.out$performances %>%
  kableExtra::kable()  %>%
  kableExtra::kable_styling(latex_options = "scale_down")

```

```{r Plot contributions to SVM model, fig.width = 8, fig.height = 8}

plot_svm_weights(svmfit,antigens.df)

```

```{r Use cord values for SVM analysis, eval = FALSE}

svm_input_cord <- make_svm_input(antigen_exposure_response.df,age="Cord")

svmfit_cord = e1071::svm(FinalCluster ~ .,
             data = svm_input_cord,
             kernel = "linear",
             cost = 10,
             scale = FALSE)

save(svm_input_cord,svmfit_cord,
     file="Maela_antigens_svm_cord_data.Rdata")

```

```{r Plot contributions to SVM cord model, fig.width = 8, fig.height = 8}

load(file="Maela_antigens_svm_cord_data.Rdata")

plot_svm_weights(svmfit_cord,antigens.df)

```

```{r Compare weights to absolute IgG responses}

antigen_SVM_weights.vec <- apply(svmfit$SV, 2, function(v){sqrt(sum(v^2))})

antigen_SVM_weights.df <-
  data.frame(
    "antigen" = names(antigen_SVM_weights.vec),
    "SVM.weight" = as.numeric(antigen_SVM_weights.vec)
  )

compare_weights_responses.df <-
  antigen_SVM_weights.df %>%
    dplyr::left_join(antigen_exposure_response.df %>%
                       dplyr::select(sero_age,antigen,igg) %>%
                       dplyr::group_by(sero_age,antigen) %>%
                       dplyr::summarise(`Mean IgG` = mean(igg)) %>%
                       dplyr::ungroup(),
                     by = c("antigen")
    ) %>%
    dplyr::left_join(antigens.df %>%
                       dplyr::select(antigen,type) %>%
                       dplyr::distinct(),
                     by = c("antigen")
                     )

ggplot(compare_weights_responses.df,
       aes(x = SVM.weight,
           y = `Mean IgG`,
           colour = type)) +
  geom_point(size = 0.5) +
  facet_wrap(.~sero_age) +
  theme_bw() +
  theme(strip.background = element_blank())

```

```{r Normalise SVM input data}

normalised_svm_input <- normalise_svm_input(svm_input)
normalised_svm_input_cord <- normalise_svm_input(svm_input_cord)

```

```{r Run SVM on normalised birth data, eval = FALSE}

normalised.tune.out<-tune(e1071::svm,
              FinalCluster ~ .,
              data=normalised_svm_input,
              kernel = "linear", 
              ranges =list(cost=c(0.001,0.01,0.1, 1,5,10,100)),
              scale = FALSE)

normalised_svmfit = e1071::svm(FinalCluster ~ .,
             data = normalised_svm_input,
             kernel = "linear",
             cost = 10,
             scale = FALSE)

save(normalised.tune.out,normalised_svm_input,normalised_svmfit,
     file="normalised_Maela_antigens_svm_data.Rdata")

```

```{r Demonstrate normalised SVM results have no error independent of cost function}

load(file="normalised_Maela_antigens_svm_data.Rdata")

normalised.tune.out$performances %>%
  kableExtra::kable()  %>%
  kableExtra::kable_styling(latex_options = "scale_down")

```

```{r Plot SV weights from normalised birth data, fig.width = 8, fig.height = 8}

plot_svm_weights(normalised_svmfit,antigens.df)

```

```{r Run SVM on normalised cord data, eval = FALSE}

normalised_svmfit_cord = e1071::svm(FinalCluster ~ .,
             data = normalised_svm_input_cord,
             kernel = "linear",
             cost = 10,
             scale = FALSE)

save(normalised.tune.out,normalised_svm_input_cord,normalised_svmfit_cord,
     file="normalised_Maela_antigens_svm_data_cord.Rdata")

```

```{r Plot SVM on normalised cord data, eval = TRUE}

load(file="normalised_Maela_antigens_svm_data_cord.Rdata")

plot_svm_weights(normalised_svmfit_cord,antigens.df)

```

# ZmpB summary plot

```{r ZmpB summary plot, fig.width = 8, fig.height = 10}

antigen_summary.df <-
  antigen_exposure_response.df %>%
    # dplyr::left_join(antigens.df %>%
    #                    dplyr::select(antigen,type) %>%
    #                    dplyr::distinct(),
    #                  by = c("antigen")
    #                 ) %>%
    dplyr::select(type,codenum,sero_age,igg) %>%
    dplyr::group_by(type,codenum,sero_age) %>%
    dplyr::summarise(`Mean IgG` = mean(igg)) %>%
    dplyr::ungroup()

zmpB_summary.df <-
  antigen_summary.df %>%
    tidyr::pivot_wider(id_cols = c(codenum,type),names_from = sero_age,values_from = `Mean IgG`)

ggplot(zmpB_summary.df,
       aes(x = Cord,
           y = `6 months`,
           colour = type,
           fill = type)) +
  geom_point(size = 0.5, alpha = 1) +
  scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  xlab("Mean IgG in cord sample") +
  ylab("Mean IgG at 6 months") +
  geom_smooth(method = "lm") +
  stat_cor(label.y = 6.5) +
  facet_wrap(.~type, nrow = 5) +
  theme_bw() +
  theme(legend.position = "bottom",
        strip.background = element_blank())

```

```{r Compare 6 months to 24 months by type, fig.width = 8, fig.height = 10}

ggplot(zmpB_summary.df,
       aes(x = `6 months`,
           y = `24 months`,
           colour = type,
           fill = type)) +
  geom_point(size = 0.5, alpha = 1) +
  scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type") +
  ylab("Mean IgG at 24 months") +
  xlab("Mean IgG at 6 months") +
  geom_smooth(method = "lm") +
  stat_cor(label.y = 6) +
  facet_wrap(.~type, nrow = 5) +
  theme_bw() +
  theme(legend.position = "bottom",
        strip.background = element_blank())

```

## Run UMAP analyses

```{r Test UMAP analysis, fig.height = 11, fig.width = 8}

library(umap)

plot_umap <- function(mat,df,perplexity = 40,match_by=NULL,colour_by=NULL,group_by=NULL,shape_by=NULL,group_min=10,max_iter=200) {
  
  custom.config <- umap.defaults
  custom.config$n_neighbors <- perplexity
  custom.config$n_epochs <- max_iter
  umap_out<-umap::umap(mat, config=custom.config)
  
  mat_row_order<-match(rownames(mat),df[[match_by]])
  df<-df[mat_row_order,]
  df[["x"]]<-umap_out$layout[,1]
  df[["y"]]<-umap_out$layout[,2]
  
  df %<>%
    dplyr::group_by(.data[[colour_by]]) %>%
    dplyr::mutate(group_count = n()) %>% 
    dplyr::ungroup()
  
  base_plot <- NULL
  if (is_null(shape_by)) {
    base_plot <- ggplot(df,
                         aes(x = x,
                             y = y,
                             colour = .data[[colour_by]],
                             group = .data[[group_by]]))
  } else {
    base_plot <- ggplot(df,
                         aes(x = x,
                             y = y,
                             colour = .data[[colour_by]],
                             shape = .data[[shape_by]],
                             group = .data[[group_by]]))
    
  }
  
  base_plot <-
    base_plot +
      geom_point()
  
  if (group_min < 4) {
    base_plot <-
      base_plot +
      ggforce::geom_mark_ellipse(aes(x0 = x, y0 = y))
  } else {
    base_plot <-
      base_plot +
      stat_ellipse(data = df %>% dplyr::filter(group_count >= group_min),type = "t")
  }
    
  base_plot <-
    base_plot +
    xlab("UMAP dimension 1") +
    ylab("UMAP dimension 2") +
    theme_bw() +
    theme(legend.position = "bottom")

}

(single_isolate_umap <-
    plot_umap(maela_single_isolate_presence_absence.mat,
            lane.df,
            perplexity = 10,
            match_by = "lane",
            colour_by = "InferredGPSC",
            group_by = "InferredGPSC",
            group_min = 5)
) + theme(legend.position = "none")

```

```{r Variable antigen UMAP, fig.height = 11, fig.width = 8}

(single_isolate_variable_antigen_umap <-
  plot_umap(maela_variable_antigen.mat,
            lane.df,
            perplexity = 10,
            match_by = "lane",
            colour_by = "InferredGPSC",
            group_by = "InferredGPSC",
            group_min = 10,
            max_iter = 1000,
            )
)

```

```{r All antigen UMAP, fig.height = 11, fig.width = 8}

(all_antigens <-
  plot_umap(igg.data,
            antigens.df %>%
              dplyr::select(antigen,type,gene) %>%
              dplyr::distinct() %>%
              dplyr::mutate(type = dplyr::case_when(
                                antigen %in% pbp_alleles ~ "PBP",
                                type == "CellWall" ~ "CellWall (not PBP)",
                                antigen %in% pclA_alleles ~ "PclA",
                                type == "Adhesin" ~ "Adhesin (not PclA)",
                                TRUE ~ type
                              )
                            ),
            perplexity = 10,
            match_by="antigen",
            colour_by="type",
            group_by="type",
            group_min=5,
            max_iter=1000) +
            scale_colour_manual(values = type_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Protein type")
)

```

```{r All individual UMAP, fig.height = 11, fig.width = 8}

(all_individuals_umap <-
  plot_umap(t(igg.data),
            antigens.df,
            perplexity = 10,
            match_by="sero_specnum",
            colour_by="codenum",
            group_by="codenum",
            shape_by="sero_age",
            group_min=3,
            max_iter=500) +
  scale_shape_manual(values = sero_age_shapes, name = "Age") +
  guides(colour = guide_legend(title.position = "top",
                               title = "Household"),
         shape = guide_legend(title.position = "top"))
)

```

```{r All individuals early UMAP, fig.height = 11, fig.width = 8}

(all_individuals_early <-
  plot_umap(t(early_igg_by_cluster.data),
            antigens.df[antigens.df$sero_age %in% c("Cord","Birth"),],
            perplexity = 10,
            match_by="sero_specnum",
            colour_by="cluster",
            group_by="codenum",
            shape_by="sero_age",
            group_min=2,
            max_iter=500) +
  scale_shape_manual(values = sero_age_shapes, name = "Age") +
  guides(colour = guide_legend(title.position = "top",
                               title = "Cluster"),
         shape = guide_legend(title.position = "top")) +
  scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster")

)


```

```{r All individuals late UMAP, fig.height = 8, fig.width = 8}

(all_individuals_late <-
  plot_umap(t(late_igg_by_cluster.data),
            antigens.df[antigens.df$sero_specnum %in% late_specnums,],
            perplexity = 10,
            match_by="sero_specnum",
            colour_by="cluster",
            group_by="codenum",
            shape_by="sero_age",
            group_min=2,
            max_iter=500) +
  scale_shape_manual(values = sero_age_shapes, name = "Age") +
  guides(colour = guide_legend(title.position = "top",
                               title = "Cluster"),
         shape = guide_legend(title.position = "top")) +
  scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster")
)

```

# Analyse the household data

```{r Read household data}

household.df <-
  read.csv(archive::archive_read("household_data.csv.tar.gz")) %>%
  dplyr::select(codenum,mother_ethnic_group,mother_smoke,infant_gender,infant_bw,type_delivery,house_fire,house_fuel) %>%
  dplyr::mutate(house_fuel = str_to_title(house_fuel)) %>%
  dplyr::mutate(infant_gender = str_to_title(infant_gender)) %>%
  dplyr::mutate(mother_ethnic_group = dplyr::case_when(
                    mother_ethnic_group == "muslim" ~ "Muslim",
                    mother_ethnic_group == "sgaw karen" ~ "S'gaw Karen",
                    mother_ethnic_group == "pwo karen" ~ "Pwo Karen",
                    mother_ethnic_group == "burmese" ~ "Burmese",
                    TRUE ~ "Unknown"
                  )
                )

antigen_exposure_response_environment.df <-
  antigen_exposure_response.df %>%
    dplyr::left_join(household.df, by = c("codenum"))

if (!("type" %in% colnames(antigen_exposure_response_environment.df))) {
  antigen_exposure_response_environment.df %<>%
    dplyr::left_join(antigens.df %>%
                       dplyr::select(antigen,type) %>%
                       dplyr::distinct(),
                     by = c("antigen"))
}

```

```{r Plot response by ethnicity, fig.width = 8, fig.height = 11}

plot_igg_levels(antigen_exposure_response_environment.df,
                "mother_ethnic_group",
                x_lab = "Ethnic group of mother",
                wrap_len = 15,
                include_ages = c("Cord", "Birth", "6 months", "12 months", 
"18 months", "24 months"))

```

```{r Tabulate differences by ethnicity}

cluster_by_ethnicity.df <-
  antigen_exposure_response_environment.df %>%
    dplyr::select(codenum,cluster,mother_ethnic_group) %>%
    dplyr::distinct()

cluster_by_ethnicity.tab <- table(cluster_by_ethnicity.df$cluster,cluster_by_ethnicity.df$mother_ethnic_group)
cluster_by_ethnicity.tab %>%
  kableExtra::kable() %>%
  kableExtra::kable_styling(latex_options = "scale_down")

```

```{r Test for association with ethnicity}

fisher.test(cluster_by_ethnicity.tab,simulate.p.value=TRUE)

```

```{r Plot response by smoking, fig.width = 8, fig.height = 11}

plot_igg_levels(antigen_exposure_response_environment.df,
                "mother_smoke",
                x_lab = "Maternal smoking",
                wrap_len = 15,
                include_ages = c("Cord", "Birth", "6 months", "12 months", 
"18 months", "24 months"))

```

```{r Tabulate differences by smoking}

cluster_by_smoking.df <-
  antigen_exposure_response_environment.df %>%
    dplyr::select(codenum,cluster,mother_smoke) %>%
    dplyr::distinct()

cluster_by_smoking.tab <- table(cluster_by_smoking.df$cluster,cluster_by_smoking.df$mother_smoke)
cluster_by_smoking.tab %>%
  kableExtra::kable() %>%
  kableExtra::kable_styling(latex_options = "scale_down")

```

```{r Test for association with smoking}

fisher.test(cluster_by_smoking.tab)

```

```{r Plot response by infant sex, fig.width = 8, fig.height = 11}

plot_igg_levels(antigen_exposure_response_environment.df,
                "infant_gender",
                x_lab = "Sex",
                wrap_len = 15)

```

```{r Tabulate LRTI by sex}

lrti_by_sex.df <-
 antigen_exposure_response_environment.df %>%
      dplyr::select(codenum,infant_gender,sero_age,LRTI_since_last_sample) %>%
      dplyr::distinct() %>%
      dplyr::select(infant_gender,sero_age,LRTI_since_last_sample)

lrti_by_sex.df %>%
  dplyr::select(infant_gender,sero_age,LRTI_since_last_sample) %>%
  dplyr::mutate(one_pnc = dplyr::if_else(LRTI_since_last_sample > 0,
                                         1,
                                         0)) %>%
  dplyr::group_by(infant_gender,sero_age) %>%
  dplyr::summarise(pnc = sum(one_pnc)) %>%
  dplyr::ungroup() %>%
  kableExtra::kable() %>%
  kableExtra::kable_styling(latex_options = "scale_down")

```

```{r Tabulate differences by sex}

cluster_by_sex.df <-
  antigen_exposure_response_environment.df %>%
    dplyr::select(codenum,cluster,infant_gender) %>%
    dplyr::distinct()

cluster_by_sex.tab <- table(cluster_by_sex.df$cluster,cluster_by_sex.df$infant_gender)
cluster_by_sex.tab %>%
  kableExtra::kable() %>%
  kableExtra::kable_styling(latex_options = "scale_down")

```

```{r Test for association with sex}

fisher.test(cluster_by_sex.tab)

```

```{r Tabulate differences by sex - just infant clusters}

cluster_by_sex_just_infant.df <-
  antigen_exposure_response_environment.df %>%
    dplyr::filter(sero_age %in% c("6 months","12 months","18 months","24 months")) %>%
    dplyr::select(codenum,cluster,infant_gender) %>%
    dplyr::distinct() %>%
    dplyr::mutate(cluster = dplyr::if_else(cluster == "Infant low",
                                           "Infant low",
                                           "Other"))

cluster_by_sex_just_infant.tab <-
  table(cluster_by_sex_just_infant.df$cluster,cluster_by_sex_just_infant.df$infant_gender)
cluster_by_sex_just_infant.tab %>%
  kableExtra::kable() %>%
  kableExtra::kable_styling(latex_options = "scale_down")

```

```{r Test for association with sex - just infant clusters}

fisher.test(cluster_by_sex_just_infant.tab)

```

```{r Plot response by delivery type, fig.width = 8, fig.height = 11}

plot_igg_levels(antigen_exposure_response_environment.df,
                "type_delivery",
                x_lab = "Type of delivery",
                wrap_len = 15)

```

```{r Plot response by heat source, fig.width = 8, fig.height = 11}

plot_igg_levels(antigen_exposure_response_environment.df,
                "house_fire",
                x_lab = "House heated by fire",
                wrap_len = 15,
                include_ages = c("Cord", "Birth", "6 months", "12 months", 
"18 months", "24 months"))

```

```{r Plot response by heat source fuel, fig.width = 8, fig.height = 11}

plot_igg_levels(antigen_exposure_response_environment.df,
                "house_fuel",
                x_lab = "Fuel used for house fire",
                wrap_len = 15,
                include_ages = c("Cord", "Birth", "6 months", "12 months", 
"18 months", "24 months"))

```

```{r Tabulate differences by fuel}

cluster_by_fuel.df <-
  antigen_exposure_response_environment.df %>%
    dplyr::select(codenum,cluster,house_fuel) %>%
    dplyr::distinct()
  
table(cluster_by_fuel.df$cluster,cluster_by_fuel.df$house_fuel) %>%
  kableExtra::kable() %>%
  kableExtra::kable_styling(latex_options = "scale_down")

```

# SIMLR validation

```{r Quick SIMLR test}

test.igg <-
  matrix((c(0.93,1.02,0.94,1.04,1.02,0.94,0.93,1.02,0.94,1.04,1.02,0.94,
           0.92,1.07,0.91,1.03,0.92,1.07,0.92,1.07,0.91,1.03,0.92,1.07,
           0.97,1.02,0.9,1.02,0.9,1.02,0.97,1.02,0.9,1.02,0.9,1.02,
           0.93,1.02,0.94,1.04,1.02,0.94,0.93,1.02,0.94,1.04,1.02,0.94,
           0.92,1.07,0.91,1.03,0.92,1.07,0.92,1.07,0.91,1.03,0.92,1.07,
           0.97,1.02,0.9,1.02,0.9,1.02,0.97,1.02,0.9,1.02,0.9,1.02,
           2.01,2.4,10.3,10.2,2.4,10.3,2.01,2.4,10.3,10.2,2.4,10.3,
           3.1,3.3,11.2,11.8,3.3,11.2,3.1,3.3,11.2,11.8,3.3,11.2,
           1.2,1.1,20.4,20.5,1.1,20.4,1.2,1.1,20.4,20.5,1.1,20.4,
           2.01,2.4,10.3,10.2,2.4,10.3,2.01,2.4,10.3,10.2,2.4,10.3,
           3.1,3.3,11.2,11.8,3.3,11.2, 3.1,3.3,11.2,11.8,3.3,11.2,
           1.2,1.1,20.4,20.5,1.1,20.4,1.2,1.1,20.4,20.5,1.1,20.4)),
         byrow = TRUE, nrow = 12, ncol = 12)

colnames(test.igg) <- paste0("ind_",c(1:12))
rownames(test.igg) <- c(paste0("no_",c(1:6)),paste0("yes_",c(1:6)))

sim.out <- SIMLR::SIMLR(test.igg,2)

y.out <- SIMLR:: SIMLR_Feature_Ranking(sim.out$S,test.igg)

y.out$names <- rownames(test.igg)

y.out$sd <- apply(test.igg,1,sd)

ggplot(as.data.frame(y.out),
       aes(x = sd,
           y = aggR)) +
  geom_point() +
  theme_bw()

```

## Detaiil of ZmpB responses

```{r Processing of raw data to generate foreground values for ZmpB, eval=FALSE}

# Start with modified version of the code from Joe Campo
# Read in the data
back.df <- 
  read.csv(archive::archive_read("probe_background.csv.tar.gz"), stringsAsFactors = FALSE, check.names = FALSE, row.names = 1)
fore.df <- 
  read.csv(archive::archive_read("probe_foreground.csv.tar.gz"), stringsAsFactors = FALSE, check.names = FALSE, row.names = 1)
ann.df <- 
  read.csv(archive::archive_read("probe_annotation.csv.tar.gz"), stringsAsFactors = FALSE, check.names = FALSE, row.names = 1)
pheno.df <- 
  read.csv(archive::archive_read("sample_characteristics.csv.tar.gz"), stringsAsFactors = FALSE, check.names = FALSE)

# PspA_1 looks missing on RefPool_05 sample, set to NA
back.df[which(ann.df$Gene.ID == "PspA_1"), "RefPool_05"] <- NA
fore.df[which(ann.df$Gene.ID == "PspA_1"), "RefPool_05"] <- NA

# Remove local background from foreground
zmpb.data.df <- fore.df - back.df
rm(back.df) # keep fore.df for later assessment of saturation
zmpb.rawdata.df <- cbind(ann.df, zmpb.data.df)

# Make an empty wide data frame to take max of duplicated Gene.ID (one value per Gene.ID)
# Keep only the COGS from SPARC study, omit the TIGR4 geneome
# Do the same for the Foreground data frame, because will use later to assess saturation
zmpb.temp.df <- zmpb.temp2.df <- data.frame(row.names = unique(subset(ann.df, Spot.Type == "IVTT.AG" & Strain.ID != "TIGR4")$Unique.ID))
for (v in pheno.df$Sample.ID) {
  zmpb.temp.df[, v] <- zmpb.temp2.df[, v] <- NA
}
rm(v)

for (v in rownames(zmpb.temp.df)) {
  zmpb.temp.df[v, ] <- apply(zmpb.data.df[subset(ann.df, Unique.ID == v)$Unique.ID, colnames(zmpb.temp.df)], 2, max, na.rm = TRUE)
  zmpb.temp2.df[v, ] <- apply(fore.df[subset(ann.df, Unique.ID == v)$Unique.ID, colnames(zmpb.temp2.df)], 2, max, na.rm = TRUE)
}
rm(v)

# Keep the IVTT.CTRLs by making separate data frames and binding them
zmpb.ctrl.df <- zmpb.data.df[subset(ann.df, Spot.Type == "IVTT.CTRL")$Unique.ID, colnames(zmpb.temp.df)]
zmpb.newdata.df <- rbind(zmpb.ctrl.df, zmpb.temp.df)



#########

# newann.df <- subset(ann.df, !duplicated(Unique.ID) & Spot.Type == "IVTT.AG" & Strain.ID != "TIGR4")
# rownames(newann.df) <- newann.df$Unique.ID
# newann.df <- rbind(subset(ann.df, Spot.Type == "IVTT.CTRL"), newann.df)
# 
# data.df <- newdata.df[rownames(newann.df), ]
# ann.df <- newann.df
# ann.df$Unique.ID <- rownames(ann.df)
# fore.df <- temp2.df[subset(ann.df, Spot.Type == "IVTT.AG")$Gene.ID, ]
# 
# rm(ctrl.df, newann.df, newdata.df, temp.df, temp2.df)

#########

zmpb.data.df <- zmpb.newdata.df
zmpb.fore.df <- zmpb.temp2.df

#########

# Adjust negatives to 1--this will be the floor of the data
for (i in 1:ncol(zmpb.data.df)) {
  zmpb.data.df[which(zmpb.data.df[, i] < 1), i] <- 1
}
rm(i)
# Set IVTT spots with foreground MFI < 100 to NA--it is likely a missing spot
for (i in colnames(zmpb.data.df)) {
  zmpb.data.df[rownames(subset(fore.df, get(i) < 100)), i] <- NA
}
rm(i)

# Make the data long
zmpb.long.df <- reshape(zmpb.data.df, direction = "long", ids = row.names(zmpb.data.df),
                   idvar = "Unique.ID", times = names(zmpb.data.df),
                   timevar = "Sample.ID", varying = list(names(zmpb.data.df)),
                   v.names = "MFI")

#########

# Adjust negatives to 1 and log transform for log2 data frame
zmpb.log2.df <- zmpb.temp.df <- log2(zmpb.data.df)
zmpb.temp.df <- reshape(zmpb.temp.df,
                        direction = "long",
                        ids = row.names(zmpb.temp.df),
                        idvar = "Unique.ID",
                        times = names(zmpb.temp.df),
                        timevar = "Sample.ID",
                        varying = list(names(zmpb.temp.df)),
                        v.names = "Log2.MFI")
zmpb.long.df <- merge(zmpb.long.df,
                      zmpb.temp.df,
                      by = c("Unique.ID", "Sample.ID"),
                      all = TRUE,
                      sort = FALSE)

rm(zmpb.temp.df)

#########

# Make normalized data: Subtract the IVTT.CTRL from log2 data = log2(fold-over)
# Use the original normalisation values so that probe values match with the original dataset
#zmpb.ivtt.back <- apply(zmpb.log2.df, 2, median, na.rm = TRUE)
zmpb.temp.df <- zmpb.log2.df
for (i in 1:ncol(zmpb.temp.df)) {
  zmpb.temp.df[, i] <- zmpb.temp.df[, i] - ivtt.back[i]
}

# Set floor of normalized data to -2, or 1/4th of the IVTT.CTRL medians
for (i in 1:ncol(zmpb.temp.df)) {
  zmpb.temp.df[which(zmpb.temp.df[, i] < -2), i] <- -2
}
zmpb.norm.df <- zmpb.temp.df

#rm(zmpb.ivtt.back, i)

zmpb.temp.df <- reshape(zmpb.temp.df,
                        direction = "long",
                        ids = row.names(zmpb.temp.df),
                        idvar = "Unique.ID",
                        times = names(zmpb.temp.df),
                        timevar = "Sample.ID",
                        varying = list(names(zmpb.temp.df)),
                        v.names = "Norm.MFI")
zmpb.long.df <- merge(zmpb.long.df, zmpb.temp.df, by = c("Unique.ID", "Sample.ID"), all = TRUE, sort = FALSE)

rm(zmpb.temp.df)

# Remove background and raw MFI; keep foreground for use in evaluating saturation
rm(data.df)

# Now continue with processing to remove technical validation samples

# Data processing detailed in Data_Xfer/Source/SpFull_Maela_makedata_v2.R
probe.igg.data<-t(zmpb.norm.df)
zmpb.pheno.data<-pheno.df

# remove negative controls
probe.igg.data<-probe.igg.data[,!(grepl("Negative control",colnames(probe.igg.data)))]

# remove ref pools
probe.igg.data<-t(as.data.frame(probe.igg.data[!(grepl("REF",zmpb.pheno.data$Sample.Type)),]))
zmpb.pheno.data<-zmpb.pheno.data[!(grepl("REF",zmpb.pheno.data$Sample.Type)),]

# filter to immunoreactive probes
#igg.data<-igg.data[apply(igg.data,1,max) >= 1,]

# fill in missing data
for (i in 1:ncol(probe.igg.data)) {
  probe.igg.data[is.na(probe.igg.data[,i]), i] <- mean(probe.igg.data[,i], na.rm = TRUE)
}

# Remove undefined columns
probe.igg.data<-probe.igg.data[!is.na(rownames(probe.igg.data)),]
probe.igg.data<-probe.igg.data[rownames(probe.igg.data)!="CLS02728",]

# Reformat some antigen names
rownames(probe.igg.data) <-
    tibble::as_tibble(rownames(probe.igg.data)) %>%
    # Add functional information
    dplyr::mutate(value = dplyr::case_when(
                    grepl("zmp",value) ~ gsub("zmp","Zmp",value),
                    value == "psrP" ~ "PsrP",
                    value == "lytA" ~ "LytA",
                    value == "pblB" ~ "PblB",
                    TRUE ~ value
                  )
                ) %>%
    dplyr::pull()
```

```{r Plot responses to ZmpB using generic Zmp functions, fig.height = 11, fig.width = 8}

# Convert into ZmpB data frame
get_zmp_df <- function(df,search_term) {
  
  as.data.frame(df) %>%
    tibble::rownames_to_column("probe") %>%
    tidyr::pivot_longer(cols = -probe,
                        names_to = "sero_specnum",
                        values_to = "igg") %>%
    dplyr::filter(grepl(search_term,probe)) %>%
    dplyr::left_join(ann.df %>%
                       dplyr::select(Unique.ID,Gene.ID) %>%
                       dplyr::mutate(Unique.ID = gsub("zmp","Zmp",Unique.ID)),
                     by = c("probe"="Unique.ID")) %>%
    dplyr::left_join(translation.df,
                     by = c("sero_specnum")) %>%
    dplyr::mutate(probe_type = dplyr::case_when(
                      grepl("s1",probe) ~ "N terminal",
                      grepl("s2",probe) ~ "middle",
                      grepl("s3",probe) ~ "C terminal",
                      TRUE ~ "complete"
                    )
                  ) %>%
    dplyr::mutate(probe = sub('_[^_]+$', '', probe))
  
}

plot_zmp_df <- function(df) {

  ggplot(df,
       aes(x = sero_age,
           y = igg,
           group = probe,
           colour = probe_type,
           fill = probe_type)) +
  geom_point(position = position_jitter(width = 0.1),
             size = 0.5,
             alpha = 0.5) +
  geom_smooth() +
  facet_wrap(.~Gene.ID) +
  xlab("Sample") +
  ylab("IgG binding") +
  ggrepel::geom_label_repel(data = df %>%
                              dplyr::filter(sero_age == "24 months") %>%
                              dplyr::group_by(probe) %>%
                              dplyr::slice(1) %>%
                              dplyr::ungroup(),
                            aes(label = probe_type),
                            fill = "white",
                            size = 2
                            ) +
  guides(colour = guide_legend(title = "Probe type"),
         fill = guide_legend(title = "Probe type")) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "bottom",
        strip.background = element_blank())

}

zmpB_igg.df <-
  get_zmp_df(probe.igg.data,"mpB")

plot_zmp_df(zmpB_igg.df)

```

```{r Compare with ZmpA, fig.height = 11, fig.width = 8}

zmpA_igg.df <-
  get_zmp_df(probe.igg.data,"mpA")

plot_zmp_df(zmpA_igg.df)

```

```{r Compare with ZmpC, fig.height = 11, fig.width = 8}

zmpC_igg.df <-
  get_zmp_df(probe.igg.data,"CLS01991")

plot_zmp_df(zmpC_igg.df)

```

```{r Compare with ZmpD, fig.height = 11, fig.width = 8}

zmpD_igg.df <-
  get_zmp_df(probe.igg.data,"CLS02608")

plot_zmp_df(zmpD_igg.df)

```

```{r What is the location of the IgG interaction domain, fig.width = 8, fig.height = 8}

#Batch CD-search tool   NIH/NLM/NCBI
#cdsid  QM3-qcdsearch-1A38FB116054B78-DC4668AD6CC8A34
#datatype       hitsStandard Results
#status 0
#Start time     2023-04-20T16:08:10     Run time        0:00:00:49

cdd.df <-
  read.table("psp_zmp_domains.tab",
                     comment.char = "",
                     sep = "\t",
                     header = TRUE) %>%
    tidyr::separate(Query,sep = " ",into = c("Index","Dash","Query","Gene")) %>%
    dplyr::select(-c(Index,Dash)) %>%
    dplyr::mutate(Query = gsub(">","",Query)) %>%
    dplyr::filter(Hit.type != "superfamily") %>%
    dplyr::left_join(ann.df %>%
                       #tibble::rownames_to_column("probe") %>%
                       dplyr::select(probe,Gene.ID) %>%
                       dplyr::mutate(probe = sub('_[^_]+$', '', probe)),
                     by = c("Query"="probe")) %>%
    dplyr::mutate(Gene.ID = dplyr::case_when(
              grepl("7622_3#91_CLS02608",Query) ~ "CLS02608",
              grepl("CLS01991",Query) ~ "CLS01991",
              TRUE ~ Gene.ID
        )
      ) %>%
    dplyr::filter(!is.na(Gene.ID)) %>%
    dplyr::mutate(Gene = dplyr::case_when(
                          Gene.ID == "CLS02608" ~ "zmpD",
                          Gene.ID == "CLS01991" ~ "zmpC",
                          TRUE ~ Gene)
                  ) %>%
    dplyr::mutate(probe_type = dplyr::case_when(
                      Gene == "zmpA" & grepl("s1",Query) ~ "ZmpA N terminus",
                      Gene == "zmpA" & grepl("s2",Query) ~ "ZmpA middle",
                      Gene == "zmpA" & grepl("s3",Query) ~ "ZmpA C terminus",
                      Gene == "zmpB" & grepl("s1",Query) ~ "ZmpB N terminus",
                      Gene == "zmpB" & grepl("s2",Query) ~ "ZmpB middle",
                      Gene == "zmpB" & grepl("s3",Query) ~ "ZmpB C terminus",
                      Gene == "zmpD" & grepl("s1",Query) ~ "ZmpD N terminus",
                      Gene == "zmpD" & grepl("s2",Query) ~ "ZmpD middle",
                      Gene == "zmpD" & grepl("s3",Query) ~ "ZmpD C terminus",
                      Gene == "zmpC" & grepl("s1",Query) ~ "ZmpC N terminus",
                      Gene == "zmpC" & grepl("s2",Query) ~ "ZmpC middle",
                      Gene == "zmpC" & grepl("s3",Query) ~ "ZmpC C terminus",
                      Gene == "zmpA" ~ "ZmpA complete",
                      Gene == "zmpB" ~ "ZmpB complete",
                      Gene == "zmpD" ~ "ZmpD complete",
                      Gene == "zmpC" ~ "ZmpC complete",
                      Gene == "pspA" ~ "PspA",
                      Gene == "pspC" ~ "PspC",
                      TRUE ~ "Other"
                    )
                  )

cdd_summary.df <-
  cdd.df %>%
    dplyr::group_by(probe_type,Short.name) %>%
    dplyr::distinct() %>%
    dplyr::summarise(Domain_count = n()) %>%
    dplyr::ungroup() %>%
    dplyr::group_by(Short.name) %>%
    dplyr::mutate(max_count = max(Domain_count)) %>%
    dplyr::ungroup() %>%
    dplyr::filter(max_count >= 10)

ggplot(cdd_summary.df,
       aes(x = Short.name,
           colour = probe_type,
           fill = probe_type,
           y = Domain_count)) +
  geom_col() +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "bottom")

```

```{r Plot response by domain, fig.width = 8, fig.height = 8}

responses_to_domains.df <-
  cdd.df %>%
    dplyr::select(-c(From,To,E.Value,Bitscore,Incomplete,Hit.type,Accession,PSSM.ID,Superfamily)) %>%
    dplyr::distinct() %>%
    dplyr::mutate(Present = "Present") %>%
    tidyr::pivot_wider(id_cols = c(Query,Gene,Gene.ID,probe_type),
                       names_from = Short.name,
                       values_from = Present,
                       values_fill = "Absent") %>%
    dplyr::left_join(probe.igg.data %>%
                       tibble::as_tibble(rownames = "probe") %>%
                       tidyr::pivot_longer(cols = -probe,
                                           names_to = "sero_sample",
                                           values_to = "igg") %>%
                       dplyr::mutate(probe = gsub("Zmp","zmp",probe)) %>%
                       dplyr::mutate(probe = sub('_[^_]+$', '', probe)),
                      by = c("Query" = "probe")
                     )

plot_by_domain <- function(df,dom=NA) {
  ggplot(df %>%
    dplyr::filter(!grepl("mpD",probe_type)),
       aes(x = probe_type,
           y = igg,
           colour = .data[[dom]],
           fill = .data[[dom]])) +
  geom_point(alpha = 0.05,
             size = 0.25,
             position = position_jitterdodge()) +
  geom_violin(alpha = 0.1, draw_quantiles = c(0.5),
               scale = "width") +
  xlab("Probe type") +
  ylab("IgG") +
  ggtitle(paste("Domain:",dom)) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
}

peptidase_plot <- plot_by_domain(responses_to_domains.df,"Peptidase_M26_N")
pspc_related_plot <- plot_by_domain(responses_to_domains.df,"PspC_relate_1")
pspc_s1_plot <- plot_by_domain(responses_to_domains.df,"PspC_subgroup_1")
pspc_s2_plot <- plot_by_domain(responses_to_domains.df,"PspC_subgroup_2")

cowplot::plot_grid(
  plotlist = list(
    cowplot::plot_grid(
      plotlist = list(peptidase_plot + theme(legend.position = "none"),
                      pspc_related_plot + theme(legend.position = "none"),
                      pspc_s1_plot + theme(legend.position = "none"),
                      pspc_s2_plot + theme(legend.position = "none")),
      labels = "auto"
    ),
    cowplot::get_legend(peptidase_plot + guides(colour = guide_legend(title = "Domain",
                                                                      direction = "horizontal"),
                                                fill = guide_legend(title = "Domain",
                                                                      direction = "horizontal")))
  ),
  nrow = 2,
  ncol = 1,
  rel_heights = c(0.95,0.05)
)

```

```{r Summary plot of probe levels, fig.width = 8, fig.height = 10}

dodge_width = 0.75

annotated_responses_to_domains.df <-
  responses_to_domains.df %>%
           dplyr::select(Gene,sero_sample,probe_type,igg) %>%
           dplyr::filter(Gene %in% c("zmpA","zmpB","zmpC","zmpD")) %>%
           dplyr::mutate(probe_type = gsub("Zmp[ABCD] ","",probe_type)) %>%
           dplyr::mutate(probe_type = factor(probe_type,
                                             levels = c("N terminus",
                                                        "middle",
                                                        "C terminus",
                                                        "complete"))) %>%
           dplyr::left_join(sampling.df %>%
                              dplyr::select(sero_specnum,sero_age),
                            by = c("sero_sample"="sero_specnum")) %>%
           dplyr::mutate(Gene = gsub("zmp","Zmp",Gene)) %>%
           dplyr::mutate(type = dplyr::if_else(Gene == "ZmpD","Degradative",Gene)) %>%
           dplyr::filter(!is.na(sero_age))

(zmp_probe_plot <- 
  ggplot(annotated_responses_to_domains.df,
         aes(x = sero_age,
             y = igg,
             shape = sero_age,
             fill = type,
             colour = type
             )
          ) +
    geom_point(position = position_jitterdodge(dodge.width = dodge_width,
                                               jitter.width = 0.7),
               size = 0.25,
               alpha = 0.25) +
    geom_violin(draw_quantiles = c(0.5),
                position = position_dodge(width = dodge_width),
                alpha = 0.1,
               scale = "width") +
    facet_grid(probe_type~Gene) +
    ylab("IgG binding") +
    xlab("Age") +
    scale_shape_manual(values = sero_age_shapes, name = "Age") +
    scale_colour_manual(values = c(type_palette,"ZmpC"="magenta"),
                        aesthetics = c("fill","colour"),
                        drop = TRUE, limits = force,
                        name = "Protein type") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "none",
          strip.background = element_blank())
)

```

```{r Compare ZmpB probes to maximum response, fig.width = 8}

zmpB_max.df <-
  annotated_responses_to_domains.df %>%
    dplyr::select(sero_sample,sero_age,type,probe_type,igg) %>%
    dplyr::left_join(
        apply(probe.igg.data,2,max) %>%
          tibble::as_tibble(rownames = "sero_sample") %>%
          dplyr::rename(max_response = value)
    ) %>%
    dplyr::filter(type == "ZmpB" & probe_type == "N terminus")

(zmpB_max_plot <-
  ggplot(zmpB_max.df,
         aes(x = max_response,
             y = igg,
             shape = sero_age,
             colour = type)) +
    geom_point() +
    geom_abline(slope = 1, intercept = 0, col = "black", linetype = 2) +
      scale_colour_manual(values = type_palette,
                          aesthetics = c("fill","colour"),
                          drop = TRUE, limits = force,
                          name = "Protein type") +
    scale_shape_manual(values = sero_age_shapes, name = "Age") +
    ylab("IgG binding to ZmpB N terminal probe") +
    xlab("Maximum IgG binding across all probes") +
    theme_bw() +
    theme(legend.position = "none")
)

```

```{r Probe correlations over time, fig.width = 8, fig.height = 10}

get_probe_correlation <- function(test.df) {
  test.df %>%
    tidyr::pivot_wider(id_cols = c(sero_age),
                       names_from = c(codenum,probe),
                       values_from = igg,
                       names_glue = "{codenum}-{probe}") %>%
    corrr::correlate() %>%
    tidyr::pivot_longer(-term,names_to = "probe",values_to = "correlation") %>%
    dplyr::summarise(correlation = mean(correlation, na.rm = TRUE))
}

multiprobe.df <-
  as.data.frame(probe.igg.data) %>%
    tibble::rownames_to_column("probe") %>%
    tidyr::pivot_longer(cols = -probe,
                        names_to = "sero_specnum",
                        values_to = "igg") %>%
    dplyr::left_join(ann.df %>%
                       dplyr::select(Unique.ID,Gene.ID) %>%
                       dplyr::mutate(Unique.ID = gsub("zmp","Zmp",Unique.ID)),
                     by = c("probe"="Unique.ID")) %>%
    dplyr::left_join(translation.df,
                     by = c("sero_specnum")) %>%
    dplyr::group_by(Gene.ID) %>%
    dplyr::mutate(num_probes = n_distinct(probe)) %>%
    dplyr::ungroup() %>%
    dplyr::filter(num_probes > 1) %>%
    dplyr::select(probe,Gene.ID,sero_age,codenum,igg)

multiprobe_cor.df <-
  multiprobe.df %>%
    dplyr::group_by(Gene.ID) %>%
    tidyr::nest() %>%
    mutate(
      correlations = map(data, get_probe_correlation)
    ) %>%
    tidyr::unnest() %>%
    dplyr::select(-c(probe,sero_age,codenum,igg)) %>%
    dplyr::distinct() %>%
    # Add functional information
    dplyr::mutate(Gene.ID = dplyr::case_when(
                    grepl("zmp",Gene.ID) ~ gsub("zmp","Zmp",Gene.ID),
                    Gene.ID == "psrP" ~ "PsrP",
                    Gene.ID == "lytA" ~ "LytA",
                    Gene.ID == "pblB" ~ "PblB",
                    TRUE ~ Gene.ID
                  )
                ) %>%
    dplyr::left_join(antigen_classification.df,
                     by = c("Gene.ID" = "antigen"))

multiprobe_cor_plot <-
  ggplot(multiprobe_cor.df,
         aes(x = type,
             y = correlation,
             colour = type,
             fill = type)) +
    geom_violin(draw_quantiles = c(0.5),
                alpha = 0.25,
                 scale = "width") +
    geom_point(size = 0.5,
               position = position_jitter(width = 0.15)) +
              scale_colour_manual(values = type_palette,
                        aesthetics = c("fill","colour"),
                        drop = TRUE, limits = force,
                        name = "Protein type") +
    xlab("Protein type") +
    ylab("Mean correlation across probes") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom")

cowplot::plot_grid(
  plotlist = list(
    multiprobe_cor_plot,
    zmpB_max_plot
  ),
  nrow = 2,
  rel_heights = c(0.5,0.4),
  labels = "auto"
)

```

## Does the early response represent the time of first exposure?

```{r Get time of first exposure}

time_of_first_exposure.df <-
  antigen_exposure_response.df %>%
    dplyr::select(codenum,category,dob,first_exposure,sero_age,cluster) %>%
    dplyr::filter(category == "Infant") %>%
    dplyr::distinct() %>%
    dplyr::select(-category) %>%
    dplyr::group_by(codenum) %>%
    dplyr::mutate(first_spn = min(first_exposure, na.rm = TRUE)) %>%
    dplyr::mutate(time_to_first_spn = first_spn - dob) %>%
    dplyr::ungroup() %>%
    dplyr::select(-c(first_exposure,first_spn)) %>%
    dplyr::distinct() %>%
    tidyr::pivot_wider(id_cols = c(codenum,dob,time_to_first_spn),names_from = sero_age,names_prefix = "cluster_",values_from = cluster)

(first_exposure_plot <-
  ggplot(time_of_first_exposure.df,
         aes(x = `cluster_24 months`,
             y = time_to_first_spn,
             fill = `cluster_24 months`,
             colour = `cluster_24 months`)) +
    geom_point(position = position_jitter(width = 0.1)) +
    geom_violin(draw_quantiles = c(0.5), alpha = 0.1,
                 scale = "width") +
    xlab("Cluster at 24 months") +
    ylab("Age at time of first colonisation (days)") +
    scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster") +
    theme_bw()
)

```

## Assemble Figure 1

```{r Assemble multipanel Figure 1, fig.width = 8, fig.height = 12}

cowplot::plot_grid(
  plotlist = list(
    cowplot::plot_grid(
      plotlist = list(
        zmp_probe_plot + theme(legend.position = "none"),
        clustering_by_age_plot + theme(legend.position = "none"),
        all_individuals_early_by_cluster + theme(legend.position = "none"),
        all_individuals_just_late_by_cluster + theme(legend.position = "none")
      ),
      nrow = 2,
      ncol = 2,
      rel_heights = c(0.6,0.4),
      labels = "auto"
    ),
    cowplot::get_legend(all_individuals_early_by_cluster)
    ),
    nrow = 2,
    rel_heights = c(0.95,0.075)
)

```

```{r Assemble multipanel Figure 3, fig.width = 8, fig.height = 8}


cowplot::plot_grid(
  plotlist = list(
    cowplot::plot_grid(
      plotlist = list(
        early_sample_correlations_plot + theme(legend.position = "none"),
        late_sample_correlations_plot + theme(legend.position = "none")
      ),
      ncol = 2,
      labels = "auto",
      rel_widths = c(0.55,0.45)
    ),
    cowplot::get_legend(early_sample_correlations_plot +
      guides(colour = guide_legend(nrow = 2),
             fill = guide_legend(nrow = 2)))
  ),
  nrow = 2,
  rel_heights = c(0.925,0.075)
)

ggsave("Figure_4.pdf",
       width = 8,
       height = 8)

```

```{r Assemble multipanel Figure 5, fig.width = 8, fig.height = 11}


cowplot::plot_grid(
  plotlist = list(
    modelling_response_after_exposure_plot + theme(legend.position = "none"),
    response_by_exposures_plot
  ),
  nrow = 2,
  labels = "auto",
  rel_heights = c(0.5,0.55)
)

```

## What are the slowest developing responses?

```{r Compare maternal and 24 mo responses, fig.width = 8}

first_and_last.df <-
  antigen_exposure_response.df %>%
    dplyr::select(codenum,antigen,type,gene,sero_age,igg) %>%
    dplyr::filter(sero_age %in% c("Birth","24 months")) %>%
    tidyr::pivot_wider(id_cols = c(codenum,antigen,type,gene),
                       names_from = sero_age,
                       values_from = igg) %>%
    dplyr::group_by(antigen,type,gene) %>%
    dplyr::summarise(Birth = median(Birth),
                     `24 months`  = median(`24 months`)) %>%
    dplyr::ungroup()

first_and_last_outliers.df <-
  first_and_last.df %>%
    dplyr::mutate(adj_birth = Birth - min(Birth)) %>%
    dplyr::mutate(adj_24 = `24 months` - min(`24 months`)) %>%
    dplyr::filter(Birth > 1 | `24 months` > 1) %>%
    dplyr::filter(abs(Birth - `24 months`)/(0.5*Birth*`24 months`) > 1) %>%
    dplyr::inner_join(antigens.df %>%
                        dplyr::select(antigen,prot_label) %>%
                        dplyr::distinct(),
                      by = c("antigen"))
    # 
    # dplyr::mutate(name = dplyr::case_when(
    #                   type %in% c("PspA","PspC","ZmpA","ZmpB") ~ antigen,
    #                   grepl("PspA",antigen) ~ antigen,
    #                   antigen == "PsrP" ~ antigen,
    #                   gene == "-" | gene == " " | gene == "" ~ antigen,
    #                   TRUE ~ paste0(antigen,"\n(",gene,")")
    #                 )
    #               ) %>%
    # dplyr::filter(type != "non-ABT" | Birth > 2 | `24 months` > 1)

(first_last_plot <-
  ggplot(first_and_last.df,
         aes(x = Birth,
             y = `24 months`)
             ) +
    geom_point(aes(colour = type)) +
    scale_colour_manual(values = type_palette,
                        aesthetics = c("fill","colour"),
                        drop = TRUE, limits = force,
                        name = "Protein type") +
    geom_abline(intercept = 0, slope = 1, linetype = 2) +
    ylim(c(-1.5,5.5)) +
    xlim(c(-0.5,6.5)) +
    ylab("Median IgG binding at 24 mo") +
    xlab("Median IgG binding in maternal birth sample") +
    geom_smooth(colour = "black") +
    ggrepel::geom_label_repel(data = first_and_last_outliers.df %>%
                                      dplyr::filter(prot_label != "-" & !(type %in% c("PspA","non-ABT"))) %>%
                                      dplyr::filter(type %in% c("Adhesin") | prot_label %in% c("Wzg","TrpX","BlpB","AcoL")),
                              #aes(label = paste0(antigen,"\n",gene)),
                             aes(label = prot_label,
                                 colour = type),
                             size = 3,
                             nudge_y = -1.25,
                             nudge_x = 1,
                             force_pull = 0.1
                              ) +
    stat_cor(method = "pearson",
             p.accuracy = 0.000001,
             label.sep='\n'
    ) +
    theme_bw()
)

```

## Any effect of IgG on colonisation?

```{r Compare with colonisation frequency, fig.width = 8}

healthy_colonisation.df <-
  sampling.df %>%
    tidyr::fill(sero_specnum,.direction = "up") %>%
    tidyr::fill(sero_age,.direction = "up") %>%
    dplyr::filter(sero_age %in% c("6 months",
                                    "12 months",
                                    "18 months",
                                    "24 months")
                  ) %>%
    dplyr::filter(diagnosis_clinical == "Healthy") %>%
    dplyr::group_by(sero_specnum) %>%
    dplyr::summarise(num_samples = n(),
                     num_positive_samples = sum(pnc=="Pneumococcus isolated")) %>%
    dplyr::ungroup() %>%
    dplyr::mutate(proportion_positive_samples = num_positive_samples/num_samples)

(colonisation_frequency_categorisation_plot <-
  ggplot(healthy_colonisation.df,
         aes(x = proportion_positive_samples)) +
      geom_density(alpha = 0.2, fill = "blue", line = "blue") +
      geom_vline(xintercept = 0.75,
                 linetype = 2,
                 colour = "red") +
    xlab("Proportion of healthy samples positive for pneumococci") +
    theme_bw()
)

```

```{r Compare IgG with colonisation frequency, fig.width = 8, fig.height = 11}

antigen_exposure_response_colonisation.df <-
  antigen_exposure_response.df %>%
    dplyr::left_join(
      healthy_colonisation.df,
        by = c("sero_specnum")
    ) %>%
    dplyr::filter(sero_age %in% c("6 months",
                                    "12 months",
                                    "18 months",
                                    "24 months")
                  ) %>%
  dplyr::mutate(Colonisation_freq = dplyr::if_else(
                    proportion_positive_samples > 0.75,
                    "Frequently colonised",
                    "Infrequently colonised"
                  )
                )

plot_igg_levels(antigen_exposure_response_colonisation.df,
                "Colonisation_freq",
                x_lab = "Frequency of pneumococcal colonisation",
                wrap_len = 15)

# Plot for presentation
# (carriage_plot<-plot_igg_levels(antigen_exposure_response_colonisation.df,
#                 "Colonisation_freq",
#                 x_lab = "Frequency of pneumococcal colonisation",
#                 wrap_len = 15) + theme(legend.position = "right",legend.key.size = unit(1.5, 'cm'),legend.text = element_text(size=15),axis.text.x = element_text(size=10)) + scale_y_continuous(expand = expansion(mult = c(0.05, 0.15))))

```

## Effects of colonisation level on clustering

```{r Effects of early colonisation on clustering, fig.height = 11.5, fig.width = 9}

early_colonisation_late_clustering.df <-
  antigen_exposure_response_colonisation.df %>%
    dplyr::select(codenum,cluster,sero_age,Colonisation_freq) %>%
    dplyr::distinct() %>%
    dplyr::group_by(codenum) %>%
    dplyr::mutate(ClusterAt24Months = dplyr::if_else(
                            sero_age == "24 months",
                            cluster,
                            NA_character_
                        )
                    ) %>%
    tidyr::fill(ClusterAt24Months,.direction = "updown") %>%
    dplyr::rename(`Colonisation frequency` = Colonisation_freq)

early_colonisation_late_clustering <-
  ggplot(early_colonisation_late_clustering.df,
         aes(x = `Colonisation frequency`,
             colour = ClusterAt24Months,
             fill = ClusterAt24Months)) +
    geom_bar() +
    ylab("Number of individuals") +
    xlab("Colonisation frequency") +
    facet_grid(.~sero_age) +
    guides(fill = guide_legend(title = "Immune cluster at 24 mo"),
           colour = guide_legend(title = "Immune cluster at 24 mo")) +
    scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster") +
    theme_bw() +
    theme(axis.text.x  = element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom",
          strip.background = element_blank())

cowplot::plot_grid(
  plotlist = list(
    first_exposure_plot + theme(legend.position = "none"),
    colonisation_frequency_categorisation_plot,
    early_colonisation_late_clustering
  ),
  labels = "auto",
  nrow = 3,
  ncol = 1,
  rel_heights = c(0.3,0.3,0.4)
)

```

## Assemble Figure 2

```{r Assemble multipanel Figure 2, fig.width = 8, fig.height = 11}

cowplot::plot_grid(
  plotlist = list(
    cowplot::plot_grid(
      plotlist = list(
        model_comp_raw_data_plot + theme(legend.position = "none"),
        cowplot::plot_grid(
          plotlist = list(
            first_last_plot + theme(legend.position = "none"),
            simple_mef_type_plot + theme(legend.position = "none")
          ),
          ncol = 2,
          nrow = 1,
          labels = c("B","C")
        )
      ),
      nrow = 2,
      ncol = 1,
      rel_heights = c(0.5,0.5),
      labels = c("A","")
    ),
    cowplot::get_legend(simple_mef_type_plot +
                          guides(colour = guide_legend(direction = "horizontal")))
    ),
    nrow = 2,
    ncol = 1,
    rel_heights = c(0.95,0.075)
)

```

## Table S1

```{r Build Table S1}

epi_summary.df <-
  antigen_exposure_response_environment.df %>%
    dplyr::select(sero_specnum,
                    codenum,
                    sero_age,
                    LRTI_since_last_sample,
                    severe_LRTI_since_last_sample,
                    very_severe_LRTI_since_last_sample,
                    infant_gender,
                    mother_ethnic_group,
                    mother_smoke,
                    house_fuel,
                    infant_bw
                    ) %>%
    dplyr::distinct() %>%
    dplyr::arrange(codenum,sero_age) %>%
    dplyr::rename(
      `Sample` = sero_specnum,
      `Individual` = codenum,
      `Age at sampling` = sero_age,
      `Pneumonia diagnoses since previous sample` = LRTI_since_last_sample,
      `Severe pneumonia diagnoses since previous sample` = severe_LRTI_since_last_sample,
      `Very severe pneumonia diagnoses since previous sample` = very_severe_LRTI_since_last_sample,
      `Infant sex` = infant_gender,
      `Maternal ethnicity` = mother_ethnic_group,
      `Maternal smoking` = mother_smoke,
      `House fire fuel type` = house_fuel,
      `Infant birth weight` = infant_bw
      )

(table_s1.df <-
  epi_summary.df %>%
    dplyr::select(-c(`Pneumonia diagnoses since previous sample`,
                     `Severe pneumonia diagnoses since previous sample`,
                     `Very severe pneumonia diagnoses since previous sample`)) %>%
    tidyr::pivot_wider(id_cols = c(`Individual`,
                                   `Infant sex`,
                                   `Infant birth weight`,
                                   `Maternal ethnicity`,
                                   `Maternal smoking`,
                                   `House fire fuel type`),
                       names_from = `Age at sampling`,
                       values_from = Sample)
)

write.csv(table_s1.df,
          file = "table_s1.csv",
          quote = F,
          row.names = F)

table_s1.df %>%
    kableExtra::kable() %>%
    kableExtra::kable_styling(latex_options = "scale_down")

```

```{r Build Table S6}

(table_s6.df <-
  epi_summary.df %>%
    dplyr::select(c(Individual,
                    `Age at sampling`,
                    `Pneumonia diagnoses since previous sample`,
                     `Severe pneumonia diagnoses since previous sample`,
                     `Very severe pneumonia diagnoses since previous sample`)) %>%
    dplyr::distinct() %>%
    dplyr::filter(`Age at sampling` %in% c("6 months","12 months","18 months","24 months"))# %>%
    # dplyr::filter(`Pneumonia diagnoses since previous sample` > 0 | 
    #               `Severe pneumonia diagnoses since previous sample` > 0 | 
    #               `Very severe pneumonia diagnoses since previous sample` > 0)
)

write.csv(table_s6.df,
          file = "table_s6.csv",
          quote = F,
          row.names = F)

table_s6.df %>%
    kableExtra::kable() %>%
    kableExtra::kable_styling(latex_options = "scale_down")

```

```{r Plot responses acquired later by cluster}


acquired_responses.df <-
  antigen_exposure_response.df %>%
    dplyr::filter(sero_age == "24 months") %>%
    dplyr::select(antigen,cluster,igg) %>%
    dplyr::filter(antigen %in% c("CLS02608","CLS02871","PspA_17"))

ggplot(acquired_responses.df,
       aes(x = antigen,
           y = igg,
           colour = cluster)) +
  geom_violin(draw_quantiles = c(0.5),
              position = position_dodge()) +
  geom_point(position = position_jitterdodge()) +
  theme_bw() +
  scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster")

```

## Use modelling to test for protective effects of antibodies

```{r Modelling for protective effects, fig.height = 8, fig.width = 8}

colonisation_by_sero_code.df <-
  antigen_exposure_response_colonisation.df %>%
    dplyr::select(sero_specnum,codenum,sero_age,Colonisation_freq,AnyLRTISinceLastSample,AnySevereLRTISinceLastSample,AnyVerySevereLRTISinceLastSample) %>%
    dplyr::distinct() %>%
    dplyr::mutate(Colonisation_freq = factor(dplyr::if_else(Colonisation_freq == "Frequently colonised",
                                                            "Frequently_colonised",
                                                            "Infrequently_colonised"),
                                             levels = c("Infrequently_colonised","Frequently_colonised")
                                             )
                  )

just_inheritance.df %<>%
  dplyr::left_join(colonisation_by_sero_code.df %>%
                     dplyr::select(sero_specnum,Colonisation_freq),
                   by = c("sero_specnum"))

both_inheritance_and_exposure.df %<>%
  dplyr::left_join(colonisation_by_sero_code.df %>%
                     dplyr::select(sero_specnum,Colonisation_freq),
                   by = c("sero_specnum"))

inheritance.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + Cord_IgG*antigen + (1|codenum) + 0,
      data = just_inheritance.df,
      REML = FALSE)

inheritance.protection.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + Cord_IgG*antigen + sero_age*AnyLRTISinceLastSample + (1|codenum) + 0,
      data = just_inheritance.df,
      REML = FALSE)

inheritance.type_protection.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + Cord_IgG*antigen + sero_age*AnyLRTISinceLastSample*type + (1|codenum) + 0,
      data = just_inheritance.df,
      REML = FALSE)

inheritance_and_exposure_since_birth.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + SinceBirth*antigen + Cord_IgG*antigen + (1|codenum) + 0,
       data = both_inheritance_and_exposure.df,
       REML = FALSE)

inheritance_and_exposure_since_birth.protection.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + SinceBirth*antigen + Cord_IgG*antigen + sero_age*AnyLRTISinceLastSample + (1|codenum) + 0,
       data = both_inheritance_and_exposure.df,
       REML = FALSE)

inheritance_and_exposure_since_birth.type_protection.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + SinceBirth*antigen + Cord_IgG*antigen + sero_age*AnyLRTISinceLastSample*type + (1|codenum) + 0,
       data = both_inheritance_and_exposure.df,
       REML = FALSE)

antigen_exposure_protection_models <- 
  c(inheritance.protection.antigen_model,
       inheritance.protection.antigen_model,
       inheritance.type_protection.antigen_model,
       inheritance_and_exposure_since_birth.antigen_model,
       inheritance_and_exposure_since_birth.protection.antigen_model,
       inheritance_and_exposure_since_birth.type_protection.antigen_model)

antigen_response_protection_model_comparisons.df <-
  pairwise_model_comparisons(antigen_exposure_protection_models)

ggplot(antigen_response_protection_model_comparisons.df,
       aes(x = i,
           y = j,
           colour = delta_aic,
           fill = delta_aic)) +
  geom_tile() +
  scale_colour_distiller(type="div",aesthetics = c("colour","fill")) +
  xlab("Reference model index") +
  ylab("Comparison model index") +
  theme_bw()+
  guides(colour = guide_colourbar(title = expression("AIC"[ref]-"AIC"[comp])),
         fill = guide_colourbar(title = expression("AIC"[ref]-"AIC"[comp])))

```

Then extract the parameter estimates:

```{r Extract the parameter estimates from protection models, fig.width = 8}

antigen_response_combined_exposure_model_outputs <- combine_exposure_model_outputs(inheritance.protection.antigen_model, inheritance_and_exposure_since_birth.protection.antigen_model)
antigen_response_protection_analysis.df <- antigen_response_combined_exposure_model_outputs[[1]]
antigen_response_individuals_model_comparisons.df <- antigen_response_combined_exposure_model_outputs[[2]]

exposure_protection_outlier_labels.df <-
  antigen_response_protection_analysis.df %>%
    dplyr::filter(IgG_response == "Change after exposure") %>%
    dplyr::filter(IgG > 1) %>%
    dplyr::left_join(antigens.df %>% dplyr::select(antigen,prot_label) %>% dplyr::distinct(),
                     by = c("antigen"))

(modelling_response_protection_after_exposure_plot <-
  ggplot(antigen_response_protection_analysis.df %>%
           dplyr::filter(IgG_response %in% c("Binding to protein type",
                                                   "Correlation with cord sample",
                                                   "Change after exposure")) %>%
         dplyr::mutate(Model = dplyr::if_else(Model == "Differential exposure",
                                              "Partial exposure",
                                              Model)),
         aes(x = type,
             y = IgG,
             colour = type,
             fill = type)) +
    geom_point(position = position_jitterdodge(seed = 42),
               alpha = 0.5,
               size = 1) +
    geom_violin(draw_quantiles = c(0.5),
               alpha = 0.25,
               scale = "width") +
    facet_grid(IgG_response~Model,
               scales = 'free_y',
               labeller = labeller(IgG_response = label_wrap_gen(18))) +
    ggrepel::geom_label_repel(data = exposure_protection_outlier_labels.df %>%
         dplyr::mutate(Model = dplyr::if_else(Model == "Differential exposure",
                                              "Partial exposure",
                                              Model)),
                              aes(label = prot_label),
                              min.segment.length = 100,
                              fill = NA,
                              force_pull = 1e-6,
                              label.padding = 0.25,
                              nudge_x = 0.5,
                              nudge_y = 0.25,
                              size = 3
                              ) +
    xlab("Protein type") +
    ylab("Parameter estimate") +
    theme_bw() +
    scale_colour_manual(values = type_palette,
                        aesthetics = c("fill","colour"),
                        drop = TRUE, limits = force,
                        name = "Protein type") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          strip.background = element_blank())
)

```

```{r Fit equivalent models for differing pneumonia severities}

inheritance.severe_protection.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + Cord_IgG*antigen + sero_age*AnySevereLRTISinceLastSample + (1|codenum) + 0,
      data = just_inheritance.df,
      REML = FALSE)

inheritance.very_severe_protection.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + Cord_IgG*antigen + sero_age*AnyVerySevereLRTISinceLastSample + (1|codenum) + 0,
      data = just_inheritance.df,
      REML = FALSE)

inheritance_and_exposure_since_birth.severe_protection.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + SinceBirth*antigen + Cord_IgG*antigen + sero_age*AnySevereLRTISinceLastSample + (1|codenum) + 0,
       data = both_inheritance_and_exposure.df,
       REML = FALSE)

inheritance_and_exposure_since_birth.very_severe_protection.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + SinceBirth*antigen + Cord_IgG*antigen + sero_age*AnyVerySevereLRTISinceLastSample + (1|codenum) + 0,
       data = both_inheritance_and_exposure.df,
       REML = FALSE)

inheritance_and_exposure_since_birth.colonisation_protection.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + SinceBirth*antigen + Cord_IgG*antigen + sero_age*Colonisation_freq + (1|codenum) + 0,
       data = both_inheritance_and_exposure.df,
       REML = FALSE)

inheritance.colonisation_protection.antigen_model <-
  lmer(igg ~ sero_age*type + antigen + SinceBirth*antigen + Cord_IgG*antigen + sero_age*Colonisation_freq + (1|codenum) + 0,
       data = just_inheritance.df,
       REML = FALSE)

```

```{r Plot change in protection by age}

library(multcomp)

# combine CIs using https://stats.stackexchange.com/questions/391040/reporting-models-add-intercept-from-fixed-effect-including-confidence-intervals
get_estimate_cis <- function(lmer_mod,col_name,base_term) {
  Age <- c(6,12,18,24)
  if (any(grepl("18 months:Any",names(coef(lmer_mod)[[1]])))) {
    factor_vector <- c(
      paste(paste0(col_name,base_term),"= 0"),
      paste0(col_name,base_term," + `sero_age12 months:",col_name,base_term,"` = 0"),
      paste0(col_name,base_term," + `sero_age18 months:",col_name,base_term,"` = 0"),
      paste0(col_name,base_term," + `sero_age24 months:",col_name,base_term,"` = 0")
    )
  } else {
    factor_vector <- c(
      paste(paste0(col_name,base_term),"= 0"),
      paste0(col_name,base_term," + `sero_age12 months:",col_name,base_term,"` = 0"),
      paste0(col_name,base_term," + `sero_age24 months:",col_name,base_term,"` = 0")
    )
    Age <- c(6,12,24)
  }
  as_tibble(
    cbind(
      Age,
      confint(
        glht(lmer_mod,
         linfct = factor_vector
         )
      )$confint
    )
  )
}

protection_estimates_by_age.df <-
  dplyr::bind_rows(
    get_estimate_cis(inheritance.protection.antigen_model,
                     "AnyLRTISinceLastSample",
                     "LRTISinceLastSample") %>%
      dplyr::mutate(Model = "Universal exposure") %>%
      dplyr::mutate(Diagnosis = "Pneumonia"),
    get_estimate_cis(inheritance_and_exposure_since_birth.protection.antigen_model,
                     "AnyLRTISinceLastSample",
                     "LRTISinceLastSample") %>%
      dplyr::mutate(Model = "Partial exposure") %>%
      dplyr::mutate(Diagnosis = "Pneumonia"),
    get_estimate_cis(inheritance.severe_protection.antigen_model,
                     "AnySevereLRTISinceLastSample",
                     "SevereLRTISinceLastSample") %>%
      dplyr::mutate(Model = "Universal exposure") %>%
      dplyr::mutate(Diagnosis = "Severe pneumonia"),
    get_estimate_cis(inheritance.very_severe_protection.antigen_model,
                     "AnyVerySevereLRTISinceLastSample",
                     "VerySevereLRTISinceLastSample") %>%
      dplyr::mutate(Model = "Universal exposure") %>%
      dplyr::mutate(Diagnosis = "Very severe pneumonia"),
    get_estimate_cis(inheritance_and_exposure_since_birth.severe_protection.antigen_model,
                     "AnySevereLRTISinceLastSample",
                     "SevereLRTISinceLastSample") %>%
      dplyr::mutate(Model = "Partial exposure") %>%
      dplyr::mutate(Diagnosis = "Severe pneumonia"),
    get_estimate_cis(inheritance_and_exposure_since_birth.very_severe_protection.antigen_model,
                     "AnyVerySevereLRTISinceLastSample",
                     "VerySevereLRTISinceLastSample") %>%
      dplyr::mutate(Model = "Partial exposure") %>%
      dplyr::mutate(Diagnosis = "Very severe pneumonia"),
    get_estimate_cis(inheritance.colonisation_protection.antigen_model,
                     "Colonisation_freq",
                     "Frequently_colonised") %>%
      dplyr::mutate(Model = "Universal exposure") %>%
      dplyr::mutate(Diagnosis = "Frequent colonisation"),
    get_estimate_cis(inheritance_and_exposure_since_birth.colonisation_protection.antigen_model,
                     "Colonisation_freq",
                     "Frequently_colonised") %>%
      dplyr::mutate(Model = "Partial exposure") %>%
      dplyr::mutate(Diagnosis = "Frequent colonisation")
  )

dodge_width = 0.2

(diagnosis_lmm_by_age_plot <-
  ggplot(protection_estimates_by_age.df,
         aes(x = Age,
             y = Estimate,
             colour = Diagnosis,
             ymin = lwr,
             ymax = upr)) +
    geom_point(position = position_dodge(width = dodge_width)) +
    geom_errorbar(width = 0.25, position = position_dodge(width = dodge_width)) +
    geom_line() +
    geom_hline(yintercept = 0, linetype = 2) +
    xlab("Age (months)") +
    ylab("Relative IgG binding in affected children\nin the previous 6 months") +
    theme_bw() +
    facet_grid(.~Model) +
    theme(legend.position = "bottom",
          strip.background = element_blank())
)

```

```{r Model effects of maternal antibodies, fig.width = 8}

early_diagnosis.df <-
  antigen_exposure_response.df %>%
    dplyr::filter(sero_age %in% c("Cord") & type != "non-ABT") %>%
    dplyr::select(sero_specnum,sero_age,codenum,antigen,type,igg) %>%
    dplyr::distinct() %>%
    dplyr::left_join(colonisation_by_sero_code.df %>%
                       dplyr::filter(sero_age == "6 months") %>%
                       dplyr::select(codenum,Colonisation_freq,AnyLRTISinceLastSample,AnySevereLRTISinceLastSample,AnyVerySevereLRTISinceLastSample),
                     by = c("codenum")) %>%
    dplyr::mutate(AnyLRTISinceLastSample = factor(AnyLRTISinceLastSample, levels = c("NoLRTISinceLastSample","LRTISinceLastSample"))) %>%
    dplyr::mutate(AnySevereLRTISinceLastSample = factor(AnySevereLRTISinceLastSample, levels = c("NoSevereLRTISinceLastSample","SevereLRTISinceLastSample"))) %>%
    dplyr::mutate(AnyVerySevereLRTISinceLastSample = factor(AnyVerySevereLRTISinceLastSample, levels = c("NoVerySevereLRTISinceLastSample","VerySevereLRTISinceLastSample")))
    
get_simple_estimate <- function(lmer_mod,col_name) {
  linfac_name <- paste(col_name," == 0")
  confint(
          glht(lmer_mod,
           linfct = linfac_name
           )
        )$confint
}

early_diagnosis_carriage.antigen_model <-
  lmer(igg ~ antigen + Colonisation_freq + (1|codenum) + 0,
       data = early_diagnosis.df,
       REML = FALSE)

protection_against_carriage.df <-
  get_simple_estimate(early_diagnosis_carriage.antigen_model,
                      "Colonisation_freqFrequently_colonised")

early_diagnosis_pneumonia.antigen_model <-
  lmer(igg ~ antigen + AnyLRTISinceLastSample*type + (1|codenum) + 0,
       data = early_diagnosis.df,
       REML = FALSE)

protection_against_pneumonia.df <-
  get_simple_estimate(early_diagnosis_pneumonia.antigen_model,
                      "AnyLRTISinceLastSampleLRTISinceLastSample")

early_diagnosis_severe_pneumonia.antigen_model <-
  lmer(igg ~ antigen + AnySevereLRTISinceLastSample + (1|codenum) + 0,
       data = early_diagnosis.df,
       REML = FALSE)

protection_against_severe_pneumonia.df <-
  get_simple_estimate(early_diagnosis_severe_pneumonia.antigen_model,
                      "AnySevereLRTISinceLastSampleSevereLRTISinceLastSample")

early_diagnosis_very_severe_pneumonia.antigen_model <-
  lmer(igg ~ antigen + AnyVerySevereLRTISinceLastSample + (1|codenum) + 0,
       data = early_diagnosis.df,
       REML = FALSE)

protection_against_very_severe_pneumonia.df <-
  get_simple_estimate(early_diagnosis_very_severe_pneumonia.antigen_model,
                      "AnyVerySevereLRTISinceLastSampleVerySevereLRTISinceLastSample")

syndrome_list <- c(
  "Frequent colonisation",
  "Pneumonia",
  "Severe pneumonia",
  "Very severe pneumonia"
)
maternal_protection.df <-
  cbind(
    syndrome_list,
    dplyr::bind_rows(
      lapply(list(protection_against_carriage.df,
                protection_against_pneumonia.df,
                protection_against_severe_pneumonia.df,
                protection_against_very_severe_pneumonia.df),
           as_tibble)
    )
  )

ggplot(maternal_protection.df,
       aes(x = syndrome_list,
           y = Estimate,
           ymin = lwr,
           ymax = upr)) +
  geom_point() +
  geom_errorbar(width = 0.05) +
  xlab("Diagnosis in first six months of child's life") +
  ylab("Relative IgG binding in the umbilical cord sample") +
  geom_hline(yintercept = 0, linetype = 2) +
  theme_bw()

```

```{r Plot prediction againt observation for inheritance exposure and disease lmers, fig.height = 8, fig.width = 8}

inheritance_pneumonia_predictions <- 
  predict(inheritance.protection.antigen_model)
inheritance_and_exposure_pneumonia_predictions <-
  predict(inheritance_and_exposure_since_birth.protection.antigen_model)

inheritance_and_exposure_pneumonia_with_predictions.df <-
  inheritance_and_exposure.df %>%
      dplyr::mutate(Exposed = dplyr::if_else(antigen %in% both_inheritance_and_exposure.antigen_list,
                                                  "Partial",
                                                  "Universal"
                                                  )
                    ) %>%
      dplyr::group_by(Exposed) %>%
      dplyr::mutate(index=row_number()) %>%
      dplyr::ungroup() %>%
      dplyr::left_join(
        dplyr::bind_rows(
          tibble(
            "predicted" = inheritance_pneumonia_predictions,
            "index"=1:length(inheritance_pneumonia_predictions),
            "Exposed" = "Universal"
          ),
          tibble(
            "predicted" = inheritance_and_exposure_pneumonia_predictions,
            "index"=1:length(inheritance_and_exposure_pneumonia_predictions),
            "Exposed" = "Partial"
          )
        ),
        by = c("Exposed","index")
      )

ggplot(inheritance_and_exposure_pneumonia_with_predictions.df,
       aes(x = igg,
           y = predicted,
           colour = Exposed)) +
  geom_point(alpha = 0.25,
             size = 0.25) +
  geom_smooth(method = "lm") +
  xlab("Observed IgG level") +
  ylab("Predicted IgG level") +
  stat_cor(method = "pearson",
             p.accuracy = 0.000001,
             label.sep='\n'
  ) +
  theme_bw()

```

```{r Plot against birth weight, fig.height = 8, fig.width = 8}

linked.df <-
  antigen_response_individuals_model_comparisons.df %>%
    dplyr::left_join(antigen_exposure_response_environment.df %>% dplyr::select(codenum,infant_bw),
                     by = c("Individual"="codenum"))

cowplot::plot_grid(
  plotlist = list(
    ggplot(linked.df,
         aes(x = infant_bw,
             y = Inheritance)
         ) +
    geom_point() +
    geom_smooth(method = "lm") +
    ylab("Individual random effect in\nuniversal exposure model") +
    xlab("Infant birth weight (g)") +
    stat_cor(method = "pearson",
             p.accuracy = 0.000001,
             label.sep='\n'
    ) +
    theme_bw(),
    ggplot(linked.df,
         aes(x = infant_bw,
             y = Inheritance_exposure)
         ) +
    geom_point() +
    geom_smooth(method = "lm") +
    ylab("Individual random effect in\npartial exposure model") +
    xlab("Infant birth weight (g)") +
    stat_cor(method = "pearson",
             p.accuracy = 0.000001,
             label.sep='\n'
    ) +
    theme_bw()
  ),
  nrow = 2,
  labels ="auto"
)

```

```{r Write out data file}

save(antigen_exposure_response_environment.df,
     file="Maela_panproteome_array_data.Rdata")

write.csv(antigen_exposure_response_environment.df %>%
            dplyr::rowwise() %>%
            dplyr::mutate(
              first_serotypes = paste(unlist(first_serotypes),collapse=';'),
              first_GPSCs =paste(unlist(first_GPSCs),collapse=';'),
              serotypesSinceLastSample = paste(unlist(serotypesSinceLastSample),collapse=';'),
              GPSCsSinceLastSample = paste(unlist(GPSCsSinceLastSample),collapse=';'),
              accessionsSinceLastSample = paste(unlist(accessionsSinceLastSample),collapse=';'),
              lanesSinceLastSample = paste(unlist(lanesSinceLastSample),collapse=';')
            ) %>%
            dplyr::ungroup(),
          file = "Maela_panproteome_array_data.csv",
          quote = F,
          row.names = F)

```

```{r Make accession codes table}

write.csv(antigen_exposure_response_environment.df %>%
           dplyr::select(sero_specnum,sero_age,codenum,accessionsSinceLastSample) %>%
           dplyr::distinct() %>%
           dplyr::rowwise() %>%
           dplyr::mutate(
                  `Genomic datasets since previous serological assay` = paste(unlist(accessionsSinceLastSample),collapse=';')
           ) %>%
           dplyr::ungroup() %>%
           dplyr::select(-accessionsSinceLastSample) %>%
           dplyr::arrange(codenum,sero_age),
          file = "Maela_panproteome_array_accessions.csv",
          quote = F,
          row.names = F)


```

# Revision

## Compare with anti-capsular IgG

``` {r Kinetics of cps responses}

cps_raw.df <-
  read.csv(archive::archive_read("cps_IgG.csv.tar.gz"),
           header = T) %>%
        dplyr::rename(sero_specnum = Specimen.Number) %>%
  dplyr::group_by(Study.code) %>%
  dplyr::mutate(Age = row_number()) %>%
  dplyr::ungroup() %>%
  tidyr::pivot_longer(dplyr::starts_with("Serotype_"),
                      names_to = "Serotype",
                      values_to = "igg") %>%
  dplyr::mutate(Serotype = gsub("Serotype_","",Serotype))

ggplot(cps_raw.df,
       aes(x = Age,
           y = igg,
           group = Study.code)) +
  geom_line() +
  scale_y_continuous(trans = "log10") +
  facet_wrap(.~Serotype) +
  xlab("Age (months)") +
  ylab(expression('Geometric mean IgG titre (mg L'^{-1}*')')) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "bottom")

```

```{r Compare with cps IgG}

cps_igg.df <-
  antigen_exposure_response_environment.df %>%
    dplyr::select(sero_specnum,codenum,cluster,sero_age) %>%
    dplyr::distinct() %>%
    dplyr::inner_join(
      cps_raw.df %>%
        dplyr::select(-c(Study.code,Age)),
      by = c("sero_specnum")
    )

ggplot(cps_igg.df %>% dplyr::filter(cluster %in% c("Infant high","Infant low")),
       aes(x = sero_age,
           y = igg,
           colour = cluster,
           fill = cluster)) +
  geom_violin(draw_quantiles = c(0.5), alpha = 0.25, scale = "width", width = 0.5, position = position_dodge(width = 0.5)) +
  geom_point(position = position_jitterdodge(dodge.width = 0.5, jitter.width = 0.25)) +
  scale_y_continuous(trans = "log10") +
  facet_wrap(.~Serotype) +
  scale_colour_manual(values = cluster_palette, aesthetics = c("fill","colour")) +
  xlab("Age") +
  ylab(expression('Geometric mean IgG titre (mg L'^{-1}*')')) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "bottom")

```

# Multivariate environmental model

``` {r Multivariate environmental model}

me_env.df <-
  mef_input_with_clustering.df %>%
    dplyr::left_join(antigen_exposure_response_environment.df %>%
                       dplyr::select(codenum,
                                     sero_age,
                                     mother_ethnic_group,
                                     mother_smoke,
                                     infant_gender,
                                     infant_bw,
                                     house_fuel) %>%
                       dplyr::distinct(),
                     by = c("codenum","sero_age"))

igg.int_clustering_model_4_rerun <-
  lmer(igg ~ sero_age*type*cluster + (1|codenum) + 0,
       data = me_env.df,
       REML = FALSE)

igg.with_ethnicity <-
  lmer(igg ~ sero_age*type*cluster + (1|codenum) + mother_ethnic_group + 0,
       data = me_env.df,
       REML = FALSE)

igg.with_smoking <-
  lmer(igg ~ sero_age*type*cluster + (1|codenum) + mother_smoke + 0,
       data = me_env.df,
       REML = FALSE)

igg.with_sex <-
  lmer(igg ~ sero_age*type*cluster + (1|codenum) + infant_gender + 0,
       data = me_env.df,
       REML = FALSE)

igg.with_bw <-
  lmer(igg ~ sero_age*type*cluster + (1|codenum) + infant_bw + 0,
       data = me_env.df,
       REML = FALSE)

igg.with_fuel <-
  lmer(igg ~ sero_age*type*cluster + (1|codenum) + house_fuel + 0,
       data = me_env.df,
       REML = FALSE)

igg.with_all_env <-
  lmer(igg ~ sero_age*type*cluster + (1|codenum) + mother_ethnic_group + mother_smoke + infant_gender + infant_bw  + house_fuel + 0,
       data = me_env.df,
       REML = FALSE)


env_model_list <- c(igg.int_clustering_model_4_rerun,
                    igg.with_ethnicity,
                    igg.with_smoking,
                    igg.with_sex,
                    igg.with_bw,
                    igg.with_fuel,
                    igg.with_all_env)

env_model_comparison.df <-
  pairwise_model_comparisons(env_model_list)

ggplot(env_model_comparison.df,
       aes(x = i,
           y = j,
           colour = delta_aic,
           fill = delta_aic)) +
  geom_tile() +
  scale_colour_distiller(type="div",aesthetics = c("colour","fill")) +
  xlab("Reference model index") +
  ylab("Comparison model index") +
  theme_bw() +
  guides(colour = guide_colourbar(title = expression("AIC"[ref]-"AIC"[comp])),
         fill = guide_colourbar(title = expression("AIC"[ref]-"AIC"[comp])))

```

```{r Tabulate env model comparisons}

env_model_comparison.df %>%
  kableExtra::kable() %>%
  kableExtra::kable_styling(latex_options = "scale_down")

```

# Correlation with protein ELISA

``` {r Correlation with ELISA, fig.height = 8, fig.width = 8}

elisa.df <- 
  read.csv(archive::archive_read("protein_elisa_data.csv.tar.gz")) %>%
    dplyr::select(codenum,
                  sero_specnum,
                  dplyr::starts_with("PP")) %>%
    tidyr::pivot_longer(dplyr::starts_with("PP"), names_to = "Elisa_protein", values_to = "Elisa_value") %>%
    dplyr::left_join(
      read.csv(archive::archive_read("protein_elisa_names.csv.tar.gz")) %>%
        dplyr::rename(antigen = Array_protein),
      by = c("Elisa_protein")
    ) %>%
    dplyr::inner_join(
      antigen_exposure_response.df %>%
        dplyr::select(sero_specnum,sero_age,cluster,antigen,gene,igg),
      by = c("sero_specnum","antigen")
    ) %>%
  dplyr::mutate(antigen = factor(antigen))

ggplot(elisa.df,
       aes(
        x = Elisa_value,
        y = igg)
      ) +
  geom_point() +
  scale_x_continuous(trans = "log10") +
  geom_smooth(method = "lm") +
  stat_cor(method = "pearson",
             p.accuracy = 0.000001,
             label.sep='\n'
  ) +
  xlab(expression('ELISA IgG titre (mg L'^{-1}*')')) +
  ylab("Normalised array IgG measurement") +
  facet_wrap(~antigen) +
  theme_bw() +
  theme(strip.background = element_blank())

```

``` {r Plot ELISA data by cluster, fig.height = 11, fig.width = 8}

elisa_cluster_test.df <-
  elisa.df %>%
    dplyr::group_by(antigen,sero_age) %>%
        rstatix::wilcox_test(Elisa_value ~ cluster,
                             ref.group = "Infant high",
                             paired = FALSE,
                             p.adjust.method = "holm") %>%
        dplyr::ungroup() %>%
        rstatix::add_xy_position(x = "cluster", dodge = 1, scales = "free_y") %>%
        dplyr::filter(p.adj.signif != "ns")

ggplot(elisa.df,
       aes(x = cluster,
           y = Elisa_value,
           colour = cluster,
           fill = cluster)) +
  geom_point(size = 0.025, alpha = 0.5, position = position_jitter()) +
  geom_violin(draw_quantiles = c(0.5), alpha = 0.25, linewidth = 0.1) +
  scale_y_continuous(trans = "log10",
                     expand = expansion(mult = c(0, 0.35), 
                                         add = c(0, 0))
                     ) + #, limits = c(10,75000)) +
  scale_colour_manual(values = cluster_palette,
                      aesthetics = c("fill","colour"),
                      drop = TRUE, limits = force,
                      name = "Immune cluster") +
  facet_grid(antigen~sero_age, scales = 'free_y') +
  theme_bw() +
  xlab("") +
  ylab(expression('ELISA IgG titre (mg L'^{-1}*')')) +
  theme(strip.background = element_blank(), strip.text.y = element_text(angle = 0),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 7),
        axis.text.y = element_text(size = 8),
        legend.position = "bottom") +
  ggpubr::stat_pvalue_manual(elisa_cluster_test.df %>% 
                               dplyr::mutate(y.position = log(1.5*y.position,10)),
                              xmin = "group1", xmax = "group2", #y.position = "y.position",xmin="group1",xmax = "group2",
                              label = "p.adj.signif",
                              tip.length = 0,
                              inherit.aes = FALSE)

```


